// svix/src/api/dist/authentication.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.Authentication = void 0;
var apiTokenOut_1 = require("../models/apiTokenOut");
var appPortalAccessIn_1 = require("../models/appPortalAccessIn");
var appPortalAccessOut_1 = require("../models/appPortalAccessOut");
var applicationTokenExpireIn_1 = require("../models/applicationTokenExpireIn");
var rotatePollerTokenIn_1 = require("../models/rotatePollerTokenIn");
var streamPortalAccessIn_1 = require("../models/streamPortalAccessIn");
var streamTokenExpireIn_1 = require("../models/streamTokenExpireIn");
var dashboardAccessOut_1 = require("../models/dashboardAccessOut");
var request_1 = require("../request");
var Authentication = /** @class */ (function () {
    function Authentication(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /** Use this function to get magic links (and authentication codes) for connecting your users to the Consumer Application Portal. */
    Authentication.prototype.appPortalAccess = function (appId, appPortalAccessIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/app-portal-access/{app_id}");
        request.setPathParam("app_id", appId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(appPortalAccessIn_1.AppPortalAccessInSerializer._toJsonObject(appPortalAccessIn));
        return request.send(this.requestCtx, appPortalAccessOut_1.AppPortalAccessOutSerializer._fromJsonObject);
    };
    /** Expire all of the tokens associated with a specific application. */
    Authentication.prototype.expireAll = function (appId, applicationTokenExpireIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/app/{app_id}/expire-all");
        request.setPathParam("app_id", appId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(applicationTokenExpireIn_1.ApplicationTokenExpireInSerializer._toJsonObject(applicationTokenExpireIn));
        return request.sendNoResponseBody(this.requestCtx);
    };
    /** @deprecated Please use `appPortalAccess` instead. */
    Authentication.prototype.dashboardAccess = function (appId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/dashboard-access/{app_id}");
        request.setPathParam("app_id", appId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        return request.send(this.requestCtx, dashboardAccessOut_1.DashboardAccessOutSerializer._fromJsonObject);
    };
    /**
     * Logout an app token.
     *
     * Trying to log out other tokens will fail.
     */
    Authentication.prototype.logout = function (options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/logout");
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        return request.sendNoResponseBody(this.requestCtx);
    };
    /**
     * Logout a stream token.
     *
     * Trying to log out other tokens will fail.
     */
    Authentication.prototype.streamLogout = function (options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/stream-logout");
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        return request.sendNoResponseBody(this.requestCtx);
    };
    /** Use this function to get magic links (and authentication codes) for connecting your users to the Stream Consumer Portal. */
    Authentication.prototype.streamPortalAccess = function (streamId, streamPortalAccessIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/stream-portal-access/{stream_id}");
        request.setPathParam("stream_id", streamId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(streamPortalAccessIn_1.StreamPortalAccessInSerializer._toJsonObject(streamPortalAccessIn));
        return request.send(this.requestCtx, appPortalAccessOut_1.AppPortalAccessOutSerializer._fromJsonObject);
    };
    /** Expire all of the tokens associated with a specific stream. */
    Authentication.prototype.streamExpireAll = function (streamId, streamTokenExpireIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/stream/{stream_id}/expire-all");
        request.setPathParam("stream_id", streamId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(streamTokenExpireIn_1.StreamTokenExpireInSerializer._toJsonObject(streamTokenExpireIn));
        return request.sendNoResponseBody(this.requestCtx);
    };
    /** Get the current auth token for the stream poller. */
    Authentication.prototype.getStreamPollerToken = function (streamId, sinkId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/auth/stream/{stream_id}/sink/{sink_id}/poller/token");
        request.setPathParam("stream_id", streamId);
        request.setPathParam("sink_id", sinkId);
        return request.send(this.requestCtx, apiTokenOut_1.ApiTokenOutSerializer._fromJsonObject);
    };
    /** Create a new auth token for the stream poller API. */
    Authentication.prototype.rotateStreamPollerToken = function (streamId, sinkId, rotatePollerTokenIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/stream/{stream_id}/sink/{sink_id}/poller/token/rotate");
        request.setPathParam("stream_id", streamId);
        request.setPathParam("sink_id", sinkId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(rotatePollerTokenIn_1.RotatePollerTokenInSerializer._toJsonObject(rotatePollerTokenIn));
        return request.send(this.requestCtx, apiTokenOut_1.ApiTokenOutSerializer._fromJsonObject);
    };
    return Authentication;
}());
exports.Authentication = Authentication;
