// svix/src/api/dist/backgroundTask.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.BackgroundTask = void 0;
var backgroundTaskOut_1 = require("../models/backgroundTaskOut");
var listResponseBackgroundTaskOut_1 = require("../models/listResponseBackgroundTaskOut");
var request_1 = require("../request");
var BackgroundTask = /** @class */ (function () {
    function BackgroundTask(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /** List background tasks executed in the past 90 days. */
    BackgroundTask.prototype.list = function (options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/background-task");
        request.setQueryParams({
            status: options === null || options === void 0 ? void 0 : options.status,
            task: options === null || options === void 0 ? void 0 : options.task,
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            order: options === null || options === void 0 ? void 0 : options.order
        });
        return request.send(this.requestCtx, listResponseBackgroundTaskOut_1.ListResponseBackgroundTaskOutSerializer._fromJsonObject);
    };
    /**
     * List background tasks executed in the past 90 days.
     *
     * @deprecated Use list instead.
     * */
    BackgroundTask.prototype.listByEndpoint = function (options) {
        return this.list(options);
    };
    /** Get a background task by ID. */
    BackgroundTask.prototype.get = function (taskId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/background-task/{task_id}");
        request.setPathParam("task_id", taskId);
        return request.send(this.requestCtx, backgroundTaskOut_1.BackgroundTaskOutSerializer._fromJsonObject);
    };
    return BackgroundTask;
}());
exports.BackgroundTask = BackgroundTask;
