// svix/src/api/dist/connector.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.Connector = void 0;
var connectorIn_1 = require("../models/connectorIn");
var connectorOut_1 = require("../models/connectorOut");
var connectorPatch_1 = require("../models/connectorPatch");
var connectorUpdate_1 = require("../models/connectorUpdate");
var listResponseConnectorOut_1 = require("../models/listResponseConnectorOut");
var request_1 = require("../request");
var Connector = /** @class */ (function () {
    function Connector(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /** List all connectors for an application. */
    Connector.prototype.list = function (options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/connector");
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            order: options === null || options === void 0 ? void 0 : options.order,
            product_type: options === null || options === void 0 ? void 0 : options.productType
        });
        return request.send(this.requestCtx, listResponseConnectorOut_1.ListResponseConnectorOutSerializer._fromJsonObject);
    };
    /** Create a new connector. */
    Connector.prototype.create = function (connectorIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/connector");
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(connectorIn_1.ConnectorInSerializer._toJsonObject(connectorIn));
        return request.send(this.requestCtx, connectorOut_1.ConnectorOutSerializer._fromJsonObject);
    };
    /** Get a connector. */
    Connector.prototype.get = function (connectorId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/connector/{connector_id}");
        request.setPathParam("connector_id", connectorId);
        return request.send(this.requestCtx, connectorOut_1.ConnectorOutSerializer._fromJsonObject);
    };
    /** Update a connector. */
    Connector.prototype.update = function (connectorId, connectorUpdate) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/connector/{connector_id}");
        request.setPathParam("connector_id", connectorId);
        request.setBody(connectorUpdate_1.ConnectorUpdateSerializer._toJsonObject(connectorUpdate));
        return request.send(this.requestCtx, connectorOut_1.ConnectorOutSerializer._fromJsonObject);
    };
    /** Delete a connector. */
    Connector.prototype["delete"] = function (connectorId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/connector/{connector_id}");
        request.setPathParam("connector_id", connectorId);
        return request.sendNoResponseBody(this.requestCtx);
    };
    /** Partially update a connector. */
    Connector.prototype.patch = function (connectorId, connectorPatch) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/connector/{connector_id}");
        request.setPathParam("connector_id", connectorId);
        request.setBody(connectorPatch_1.ConnectorPatchSerializer._toJsonObject(connectorPatch));
        return request.send(this.requestCtx, connectorOut_1.ConnectorOutSerializer._fromJsonObject);
    };
    return Connector;
}());
exports.Connector = Connector;
