// svix/src/api/dist/endpoint.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.Endpoint = void 0;
var bulkReplayIn_1 = require("../models/bulkReplayIn");
var endpointHeadersIn_1 = require("../models/endpointHeadersIn");
var endpointHeadersOut_1 = require("../models/endpointHeadersOut");
var endpointHeadersPatchIn_1 = require("../models/endpointHeadersPatchIn");
var endpointIn_1 = require("../models/endpointIn");
var endpointOut_1 = require("../models/endpointOut");
var endpointPatch_1 = require("../models/endpointPatch");
var endpointSecretOut_1 = require("../models/endpointSecretOut");
var endpointSecretRotateIn_1 = require("../models/endpointSecretRotateIn");
var endpointStats_1 = require("../models/endpointStats");
var endpointTransformationIn_1 = require("../models/endpointTransformationIn");
var endpointTransformationOut_1 = require("../models/endpointTransformationOut");
var endpointTransformationPatch_1 = require("../models/endpointTransformationPatch");
var endpointUpdate_1 = require("../models/endpointUpdate");
var eventExampleIn_1 = require("../models/eventExampleIn");
var listResponseEndpointOut_1 = require("../models/listResponseEndpointOut");
var messageOut_1 = require("../models/messageOut");
var recoverIn_1 = require("../models/recoverIn");
var recoverOut_1 = require("../models/recoverOut");
var replayIn_1 = require("../models/replayIn");
var replayOut_1 = require("../models/replayOut");
var request_1 = require("../request");
var Endpoint = /** @class */ (function () {
    function Endpoint(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /** List the application's endpoints. */
    Endpoint.prototype.list = function (appId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint");
        request.setPathParam("app_id", appId);
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            order: options === null || options === void 0 ? void 0 : options.order
        });
        return request.send(this.requestCtx, listResponseEndpointOut_1.ListResponseEndpointOutSerializer._fromJsonObject);
    };
    /**
     * Create a new endpoint for the application.
     *
     * When `secret` is `null` the secret is automatically generated (recommended).
     */
    Endpoint.prototype.create = function (appId, endpointIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint");
        request.setPathParam("app_id", appId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(endpointIn_1.EndpointInSerializer._toJsonObject(endpointIn));
        return request.send(this.requestCtx, endpointOut_1.EndpointOutSerializer._fromJsonObject);
    };
    /** Get an endpoint. */
    Endpoint.prototype.get = function (appId, endpointId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        return request.send(this.requestCtx, endpointOut_1.EndpointOutSerializer._fromJsonObject);
    };
    /** Update an endpoint. */
    Endpoint.prototype.update = function (appId, endpointId, endpointUpdate) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/app/{app_id}/endpoint/{endpoint_id}");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        request.setBody(endpointUpdate_1.EndpointUpdateSerializer._toJsonObject(endpointUpdate));
        return request.send(this.requestCtx, endpointOut_1.EndpointOutSerializer._fromJsonObject);
    };
    /** Delete an endpoint. */
    Endpoint.prototype["delete"] = function (appId, endpointId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/app/{app_id}/endpoint/{endpoint_id}");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        return request.sendNoResponseBody(this.requestCtx);
    };
    /** Partially update an endpoint. */
    Endpoint.prototype.patch = function (appId, endpointId, endpointPatch) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/app/{app_id}/endpoint/{endpoint_id}");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        request.setBody(endpointPatch_1.EndpointPatchSerializer._toJsonObject(endpointPatch));
        return request.send(this.requestCtx, endpointOut_1.EndpointOutSerializer._fromJsonObject);
    };
    /**
     * Bulk replay messages sent to the endpoint.
     *
     * Only messages that were created after `since` will be sent.
     * This will replay both successful, and failed messages
     *
     * A completed task will return a payload like the following:
     * ```json
     * {
     *   "id": "qtask_33qen93MNuelBAq1T9G7eHLJRsF",
     *   "status": "finished",
     *   "task": "endpoint.bulk-replay",
     *   "data": {
     *     "messagesSent": 2
     *   }
     * }
     * ```
     */
    Endpoint.prototype.bulkReplay = function (appId, endpointId, bulkReplayIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/bulk-replay");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(bulkReplayIn_1.BulkReplayInSerializer._toJsonObject(bulkReplayIn));
        return request.send(this.requestCtx, replayOut_1.ReplayOutSerializer._fromJsonObject);
    };
    /** Get the additional headers to be sent with the webhook. */
    Endpoint.prototype.getHeaders = function (appId, endpointId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        return request.send(this.requestCtx, endpointHeadersOut_1.EndpointHeadersOutSerializer._fromJsonObject);
    };
    /** Set the additional headers to be sent with the webhook. */
    Endpoint.prototype.updateHeaders = function (appId, endpointId, endpointHeadersIn) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        request.setBody(endpointHeadersIn_1.EndpointHeadersInSerializer._toJsonObject(endpointHeadersIn));
        return request.sendNoResponseBody(this.requestCtx);
    };
    Endpoint.prototype.headersUpdate = function (appId, endpointId, endpointHeadersIn) {
        return this.updateHeaders(appId, endpointId, endpointHeadersIn);
    };
    /** Partially set the additional headers to be sent with the webhook. */
    Endpoint.prototype.patchHeaders = function (appId, endpointId, endpointHeadersPatchIn) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        request.setBody(endpointHeadersPatchIn_1.EndpointHeadersPatchInSerializer._toJsonObject(endpointHeadersPatchIn));
        return request.sendNoResponseBody(this.requestCtx);
    };
    Endpoint.prototype.headersPatch = function (appId, endpointId, endpointHeadersPatchIn) {
        return this.patchHeaders(appId, endpointId, endpointHeadersPatchIn);
    };
    /**
     * Resend all failed messages since a given time.
     *
     * Messages that were sent successfully, even if failed initially, are not resent.
     *
     * A completed task will return a payload like the following:
     * ```json
     * {
     *   "id": "qtask_33qen93MNuelBAq1T9G7eHLJRsF",
     *   "status": "finished",
     *   "task": "endpoint.recover",
     *   "data": {
     *     "messagesSent": 2
     *   }
     * }
     * ```
     */
    Endpoint.prototype.recover = function (appId, endpointId, recoverIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/recover");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(recoverIn_1.RecoverInSerializer._toJsonObject(recoverIn));
        return request.send(this.requestCtx, recoverOut_1.RecoverOutSerializer._fromJsonObject);
    };
    /**
     * Replays messages to the endpoint.
     *
     * Only messages that were created after `since` will be sent.
     * Messages that were previously sent to the endpoint are not resent.
     *
     * A completed task will return a payload like the following:
     * ```json
     * {
     *   "id": "qtask_33qen93MNuelBAq1T9G7eHLJRsF",
     *   "status": "finished",
     *   "task": "endpoint.replay",
     *   "data": {
     *     "messagesSent": 2
     *   }
     * }
     * ```
     */
    Endpoint.prototype.replayMissing = function (appId, endpointId, replayIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/replay-missing");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(replayIn_1.ReplayInSerializer._toJsonObject(replayIn));
        return request.send(this.requestCtx, replayOut_1.ReplayOutSerializer._fromJsonObject);
    };
    /**
     * Get the endpoint's signing secret.
     *
     * This is used to verify the authenticity of the webhook.
     * For more information please refer to [the consuming webhooks docs](https://docs.svix.com/consuming-webhooks/).
     */
    Endpoint.prototype.getSecret = function (appId, endpointId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/secret");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        return request.send(this.requestCtx, endpointSecretOut_1.EndpointSecretOutSerializer._fromJsonObject);
    };
    /**
     * Rotates the endpoint's signing secret.
     *
     * The previous secret will remain valid for the next 24 hours.
     */
    Endpoint.prototype.rotateSecret = function (appId, endpointId, endpointSecretRotateIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/secret/rotate");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(endpointSecretRotateIn_1.EndpointSecretRotateInSerializer._toJsonObject(endpointSecretRotateIn));
        return request.sendNoResponseBody(this.requestCtx);
    };
    /** Send an example message for an event. */
    Endpoint.prototype.sendExample = function (appId, endpointId, eventExampleIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/send-example");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(eventExampleIn_1.EventExampleInSerializer._toJsonObject(eventExampleIn));
        return request.send(this.requestCtx, messageOut_1.MessageOutSerializer._fromJsonObject);
    };
    /** Get basic statistics for the endpoint. */
    Endpoint.prototype.getStats = function (appId, endpointId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/stats");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        request.setQueryParams({
            since: options === null || options === void 0 ? void 0 : options.since,
            until: options === null || options === void 0 ? void 0 : options.until
        });
        return request.send(this.requestCtx, endpointStats_1.EndpointStatsSerializer._fromJsonObject);
    };
    /** Get the transformation code associated with this endpoint. */
    Endpoint.prototype.transformationGet = function (appId, endpointId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/transformation");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        return request.send(this.requestCtx, endpointTransformationOut_1.EndpointTransformationOutSerializer._fromJsonObject);
    };
    /** Set or unset the transformation code associated with this endpoint. */
    Endpoint.prototype.patchTransformation = function (appId, endpointId, endpointTransformationPatch) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/transformation");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        request.setBody(endpointTransformationPatch_1.EndpointTransformationPatchSerializer._toJsonObject(endpointTransformationPatch));
        return request.sendNoResponseBody(this.requestCtx);
    };
    /**
     * This operation was renamed to `set-transformation`.
     *
     * @deprecated
     */
    Endpoint.prototype.transformationPartialUpdate = function (appId, endpointId, endpointTransformationIn) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/transformation");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        request.setBody(endpointTransformationIn_1.EndpointTransformationInSerializer._toJsonObject(endpointTransformationIn));
        return request.sendNoResponseBody(this.requestCtx);
    };
    return Endpoint;
}());
exports.Endpoint = Endpoint;
