// svix/src/api/dist/eventType.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.EventType = void 0;
var eventTypeImportOpenApiIn_1 = require("../models/eventTypeImportOpenApiIn");
var eventTypeImportOpenApiOut_1 = require("../models/eventTypeImportOpenApiOut");
var eventTypeIn_1 = require("../models/eventTypeIn");
var eventTypeOut_1 = require("../models/eventTypeOut");
var eventTypePatch_1 = require("../models/eventTypePatch");
var eventTypeUpdate_1 = require("../models/eventTypeUpdate");
var listResponseEventTypeOut_1 = require("../models/listResponseEventTypeOut");
var request_1 = require("../request");
var EventType = /** @class */ (function () {
    function EventType(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /** Return the list of event types. */
    EventType.prototype.list = function (options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/event-type");
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            order: options === null || options === void 0 ? void 0 : options.order,
            include_archived: options === null || options === void 0 ? void 0 : options.includeArchived,
            with_content: options === null || options === void 0 ? void 0 : options.withContent
        });
        return request.send(this.requestCtx, listResponseEventTypeOut_1.ListResponseEventTypeOutSerializer._fromJsonObject);
    };
    /**
     * Create new or unarchive existing event type.
     *
     * Unarchiving an event type will allow endpoints to filter on it and messages to be sent with it.
     * Endpoints filtering on the event type before archival will continue to filter on it.
     * This operation does not preserve the description and schemas.
     */
    EventType.prototype.create = function (eventTypeIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/event-type");
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(eventTypeIn_1.EventTypeInSerializer._toJsonObject(eventTypeIn));
        return request.send(this.requestCtx, eventTypeOut_1.EventTypeOutSerializer._fromJsonObject);
    };
    /**
     * Given an OpenAPI spec, create new or update existing event types.
     *
     * If an existing `archived` event type is updated, it will be unarchived.
     * The importer will convert all webhooks found in the either the `webhooks` or `x-webhooks`
     * top-level.
     */
    EventType.prototype.importOpenapi = function (eventTypeImportOpenApiIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/event-type/import/openapi");
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(eventTypeImportOpenApiIn_1.EventTypeImportOpenApiInSerializer._toJsonObject(eventTypeImportOpenApiIn));
        return request.send(this.requestCtx, eventTypeImportOpenApiOut_1.EventTypeImportOpenApiOutSerializer._fromJsonObject);
    };
    /** Get an event type. */
    EventType.prototype.get = function (eventTypeName) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/event-type/{event_type_name}");
        request.setPathParam("event_type_name", eventTypeName);
        return request.send(this.requestCtx, eventTypeOut_1.EventTypeOutSerializer._fromJsonObject);
    };
    /** Update an event type. */
    EventType.prototype.update = function (eventTypeName, eventTypeUpdate) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/event-type/{event_type_name}");
        request.setPathParam("event_type_name", eventTypeName);
        request.setBody(eventTypeUpdate_1.EventTypeUpdateSerializer._toJsonObject(eventTypeUpdate));
        return request.send(this.requestCtx, eventTypeOut_1.EventTypeOutSerializer._fromJsonObject);
    };
    /**
     * Archive an event type.
     *
     * Endpoints already configured to filter on an event type will continue to do so after archival.
     * However, new messages can not be sent with it and endpoints can not filter on it.
     * An event type can be unarchived with the
     * [create operation](#operation/create_event_type_api_v1_event_type__post).
     */
    EventType.prototype["delete"] = function (eventTypeName, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/event-type/{event_type_name}");
        request.setPathParam("event_type_name", eventTypeName);
        request.setQueryParams({
            expunge: options === null || options === void 0 ? void 0 : options.expunge
        });
        return request.sendNoResponseBody(this.requestCtx);
    };
    /** Partially update an event type. */
    EventType.prototype.patch = function (eventTypeName, eventTypePatch) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/event-type/{event_type_name}");
        request.setPathParam("event_type_name", eventTypeName);
        request.setBody(eventTypePatch_1.EventTypePatchSerializer._toJsonObject(eventTypePatch));
        return request.send(this.requestCtx, eventTypeOut_1.EventTypeOutSerializer._fromJsonObject);
    };
    return EventType;
}());
exports.EventType = EventType;
