// svix/src/api/dist/ingestSource.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.IngestSource = void 0;
var ingestSourceIn_1 = require("../models/ingestSourceIn");
var ingestSourceOut_1 = require("../models/ingestSourceOut");
var listResponseIngestSourceOut_1 = require("../models/listResponseIngestSourceOut");
var rotateTokenOut_1 = require("../models/rotateTokenOut");
var request_1 = require("../request");
var IngestSource = /** @class */ (function () {
    function IngestSource(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /** List of all the organization's Ingest Sources. */
    IngestSource.prototype.list = function (options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source");
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            order: options === null || options === void 0 ? void 0 : options.order
        });
        return request.send(this.requestCtx, listResponseIngestSourceOut_1.ListResponseIngestSourceOutSerializer._fromJsonObject);
    };
    /** Create Ingest Source. */
    IngestSource.prototype.create = function (ingestSourceIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/ingest/api/v1/source");
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(ingestSourceIn_1.IngestSourceInSerializer._toJsonObject(ingestSourceIn));
        return request.send(this.requestCtx, ingestSourceOut_1.IngestSourceOutSerializer._fromJsonObject);
    };
    /** Get an Ingest Source by id or uid. */
    IngestSource.prototype.get = function (sourceId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}");
        request.setPathParam("source_id", sourceId);
        return request.send(this.requestCtx, ingestSourceOut_1.IngestSourceOutSerializer._fromJsonObject);
    };
    /** Update an Ingest Source. */
    IngestSource.prototype.update = function (sourceId, ingestSourceIn) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/ingest/api/v1/source/{source_id}");
        request.setPathParam("source_id", sourceId);
        request.setBody(ingestSourceIn_1.IngestSourceInSerializer._toJsonObject(ingestSourceIn));
        return request.send(this.requestCtx, ingestSourceOut_1.IngestSourceOutSerializer._fromJsonObject);
    };
    /** Delete an Ingest Source. */
    IngestSource.prototype["delete"] = function (sourceId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/ingest/api/v1/source/{source_id}");
        request.setPathParam("source_id", sourceId);
        return request.sendNoResponseBody(this.requestCtx);
    };
    /**
     * Rotate the Ingest Source's Url Token.
     *
     * This will rotate the ingest source's token, which is used to
     * construct the unique `ingestUrl` for the source. Previous tokens
     * will remain valid for 48 hours after rotation. The token can be
     * rotated a maximum of three times within the 48-hour period.
     */
    IngestSource.prototype.rotateToken = function (sourceId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/ingest/api/v1/source/{source_id}/token/rotate");
        request.setPathParam("source_id", sourceId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        return request.send(this.requestCtx, rotateTokenOut_1.RotateTokenOutSerializer._fromJsonObject);
    };
    return IngestSource;
}());
exports.IngestSource = IngestSource;
