// svix/src/api/dist/ingestEndpoint.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.IngestEndpoint = void 0;
var ingestEndpointHeadersIn_1 = require("../models/ingestEndpointHeadersIn");
var ingestEndpointHeadersOut_1 = require("../models/ingestEndpointHeadersOut");
var ingestEndpointIn_1 = require("../models/ingestEndpointIn");
var ingestEndpointOut_1 = require("../models/ingestEndpointOut");
var ingestEndpointSecretIn_1 = require("../models/ingestEndpointSecretIn");
var ingestEndpointSecretOut_1 = require("../models/ingestEndpointSecretOut");
var ingestEndpointTransformationOut_1 = require("../models/ingestEndpointTransformationOut");
var ingestEndpointTransformationPatch_1 = require("../models/ingestEndpointTransformationPatch");
var ingestEndpointUpdate_1 = require("../models/ingestEndpointUpdate");
var listResponseIngestEndpointOut_1 = require("../models/listResponseIngestEndpointOut");
var request_1 = require("../request");
var IngestEndpoint = /** @class */ (function () {
    function IngestEndpoint(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /** List ingest endpoints. */
    IngestEndpoint.prototype.list = function (sourceId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint");
        request.setPathParam("source_id", sourceId);
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            order: options === null || options === void 0 ? void 0 : options.order
        });
        return request.send(this.requestCtx, listResponseIngestEndpointOut_1.ListResponseIngestEndpointOutSerializer._fromJsonObject);
    };
    /** Create an ingest endpoint. */
    IngestEndpoint.prototype.create = function (sourceId, ingestEndpointIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/ingest/api/v1/source/{source_id}/endpoint");
        request.setPathParam("source_id", sourceId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(ingestEndpointIn_1.IngestEndpointInSerializer._toJsonObject(ingestEndpointIn));
        return request.send(this.requestCtx, ingestEndpointOut_1.IngestEndpointOutSerializer._fromJsonObject);
    };
    /** Get an ingest endpoint. */
    IngestEndpoint.prototype.get = function (sourceId, endpointId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}");
        request.setPathParam("source_id", sourceId);
        request.setPathParam("endpoint_id", endpointId);
        return request.send(this.requestCtx, ingestEndpointOut_1.IngestEndpointOutSerializer._fromJsonObject);
    };
    /** Update an ingest endpoint. */
    IngestEndpoint.prototype.update = function (sourceId, endpointId, ingestEndpointUpdate) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}");
        request.setPathParam("source_id", sourceId);
        request.setPathParam("endpoint_id", endpointId);
        request.setBody(ingestEndpointUpdate_1.IngestEndpointUpdateSerializer._toJsonObject(ingestEndpointUpdate));
        return request.send(this.requestCtx, ingestEndpointOut_1.IngestEndpointOutSerializer._fromJsonObject);
    };
    /** Delete an ingest endpoint. */
    IngestEndpoint.prototype["delete"] = function (sourceId, endpointId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}");
        request.setPathParam("source_id", sourceId);
        request.setPathParam("endpoint_id", endpointId);
        return request.sendNoResponseBody(this.requestCtx);
    };
    /** Get the additional headers to be sent with the ingest. */
    IngestEndpoint.prototype.getHeaders = function (sourceId, endpointId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/headers");
        request.setPathParam("source_id", sourceId);
        request.setPathParam("endpoint_id", endpointId);
        return request.send(this.requestCtx, ingestEndpointHeadersOut_1.IngestEndpointHeadersOutSerializer._fromJsonObject);
    };
    /** Set the additional headers to be sent to the endpoint. */
    IngestEndpoint.prototype.updateHeaders = function (sourceId, endpointId, ingestEndpointHeadersIn) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/headers");
        request.setPathParam("source_id", sourceId);
        request.setPathParam("endpoint_id", endpointId);
        request.setBody(ingestEndpointHeadersIn_1.IngestEndpointHeadersInSerializer._toJsonObject(ingestEndpointHeadersIn));
        return request.sendNoResponseBody(this.requestCtx);
    };
    /**
     * Get an ingest endpoint's signing secret.
     *
     * This is used to verify the authenticity of the webhook.
     * For more information please refer to [the consuming webhooks docs](https://docs.svix.com/consuming-webhooks/).
     */
    IngestEndpoint.prototype.getSecret = function (sourceId, endpointId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/secret");
        request.setPathParam("source_id", sourceId);
        request.setPathParam("endpoint_id", endpointId);
        return request.send(this.requestCtx, ingestEndpointSecretOut_1.IngestEndpointSecretOutSerializer._fromJsonObject);
    };
    /**
     * Rotates an ingest endpoint's signing secret.
     *
     * The previous secret will remain valid for the next 24 hours.
     */
    IngestEndpoint.prototype.rotateSecret = function (sourceId, endpointId, ingestEndpointSecretIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/secret/rotate");
        request.setPathParam("source_id", sourceId);
        request.setPathParam("endpoint_id", endpointId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(ingestEndpointSecretIn_1.IngestEndpointSecretInSerializer._toJsonObject(ingestEndpointSecretIn));
        return request.sendNoResponseBody(this.requestCtx);
    };
    /** Get the transformation code associated with this ingest endpoint. */
    IngestEndpoint.prototype.getTransformation = function (sourceId, endpointId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/transformation");
        request.setPathParam("source_id", sourceId);
        request.setPathParam("endpoint_id", endpointId);
        return request.send(this.requestCtx, ingestEndpointTransformationOut_1.IngestEndpointTransformationOutSerializer._fromJsonObject);
    };
    /** Set or unset the transformation code associated with this ingest endpoint. */
    IngestEndpoint.prototype.setTransformation = function (sourceId, endpointId, ingestEndpointTransformationPatch) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/transformation");
        request.setPathParam("source_id", sourceId);
        request.setPathParam("endpoint_id", endpointId);
        request.setBody(ingestEndpointTransformationPatch_1.IngestEndpointTransformationPatchSerializer._toJsonObject(ingestEndpointTransformationPatch));
        return request.sendNoResponseBody(this.requestCtx);
    };
    return IngestEndpoint;
}());
exports.IngestEndpoint = IngestEndpoint;
