// svix/src/api/dist/integration.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.Integration = void 0;
var integrationIn_1 = require("../models/integrationIn");
var integrationKeyOut_1 = require("../models/integrationKeyOut");
var integrationOut_1 = require("../models/integrationOut");
var integrationUpdate_1 = require("../models/integrationUpdate");
var listResponseIntegrationOut_1 = require("../models/listResponseIntegrationOut");
var request_1 = require("../request");
var Integration = /** @class */ (function () {
    function Integration(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /** List the application's integrations. */
    Integration.prototype.list = function (appId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/integration");
        request.setPathParam("app_id", appId);
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            order: options === null || options === void 0 ? void 0 : options.order
        });
        return request.send(this.requestCtx, listResponseIntegrationOut_1.ListResponseIntegrationOutSerializer._fromJsonObject);
    };
    /** Create an integration. */
    Integration.prototype.create = function (appId, integrationIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/integration");
        request.setPathParam("app_id", appId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(integrationIn_1.IntegrationInSerializer._toJsonObject(integrationIn));
        return request.send(this.requestCtx, integrationOut_1.IntegrationOutSerializer._fromJsonObject);
    };
    /** Get an integration. */
    Integration.prototype.get = function (appId, integId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/integration/{integ_id}");
        request.setPathParam("app_id", appId);
        request.setPathParam("integ_id", integId);
        return request.send(this.requestCtx, integrationOut_1.IntegrationOutSerializer._fromJsonObject);
    };
    /** Update an integration. */
    Integration.prototype.update = function (appId, integId, integrationUpdate) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/app/{app_id}/integration/{integ_id}");
        request.setPathParam("app_id", appId);
        request.setPathParam("integ_id", integId);
        request.setBody(integrationUpdate_1.IntegrationUpdateSerializer._toJsonObject(integrationUpdate));
        return request.send(this.requestCtx, integrationOut_1.IntegrationOutSerializer._fromJsonObject);
    };
    /** Delete an integration. */
    Integration.prototype["delete"] = function (appId, integId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/app/{app_id}/integration/{integ_id}");
        request.setPathParam("app_id", appId);
        request.setPathParam("integ_id", integId);
        return request.sendNoResponseBody(this.requestCtx);
    };
    /**
     * Get an integration's key.
     *
     * @deprecated
     */
    Integration.prototype.getKey = function (appId, integId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/integration/{integ_id}/key");
        request.setPathParam("app_id", appId);
        request.setPathParam("integ_id", integId);
        return request.send(this.requestCtx, integrationKeyOut_1.IntegrationKeyOutSerializer._fromJsonObject);
    };
    /** Rotate the integration's key. The previous key will be immediately revoked. */
    Integration.prototype.rotateKey = function (appId, integId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/integration/{integ_id}/key/rotate");
        request.setPathParam("app_id", appId);
        request.setPathParam("integ_id", integId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        return request.send(this.requestCtx, integrationKeyOut_1.IntegrationKeyOutSerializer._fromJsonObject);
    };
    return Integration;
}());
exports.Integration = Integration;
