// svix/src/api/dist/messageAttempt.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.MessageAttempt = void 0;
var emptyResponse_1 = require("../models/emptyResponse");
var listResponseEndpointMessageOut_1 = require("../models/listResponseEndpointMessageOut");
var listResponseMessageAttemptOut_1 = require("../models/listResponseMessageAttemptOut");
var listResponseMessageEndpointOut_1 = require("../models/listResponseMessageEndpointOut");
var messageAttemptOut_1 = require("../models/messageAttemptOut");
var request_1 = require("../request");
var MessageAttempt = /** @class */ (function () {
    function MessageAttempt(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /**
     * List attempts by endpoint id
     *
     * Note that by default this endpoint is limited to retrieving 90 days' worth of data
     * relative to now or, if an iterator is provided, 90 days before/after the time indicated
     * by the iterator ID. If you require data beyond those time ranges, you will need to explicitly
     * set the `before` or `after` parameter as appropriate.
     */
    MessageAttempt.prototype.listByEndpoint = function (appId, endpointId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/attempt/endpoint/{endpoint_id}");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            status: options === null || options === void 0 ? void 0 : options.status,
            status_code_class: options === null || options === void 0 ? void 0 : options.statusCodeClass,
            channel: options === null || options === void 0 ? void 0 : options.channel,
            tag: options === null || options === void 0 ? void 0 : options.tag,
            before: options === null || options === void 0 ? void 0 : options.before,
            after: options === null || options === void 0 ? void 0 : options.after,
            with_content: options === null || options === void 0 ? void 0 : options.withContent,
            with_msg: options === null || options === void 0 ? void 0 : options.withMsg,
            event_types: options === null || options === void 0 ? void 0 : options.eventTypes
        });
        return request.send(this.requestCtx, listResponseMessageAttemptOut_1.ListResponseMessageAttemptOutSerializer._fromJsonObject);
    };
    /**
     * List attempts by message ID.
     *
     * Note that by default this endpoint is limited to retrieving 90 days' worth of data
     * relative to now or, if an iterator is provided, 90 days before/after the time indicated
     * by the iterator ID. If you require data beyond those time ranges, you will need to explicitly
     * set the `before` or `after` parameter as appropriate.
     */
    MessageAttempt.prototype.listByMsg = function (appId, msgId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/attempt/msg/{msg_id}");
        request.setPathParam("app_id", appId);
        request.setPathParam("msg_id", msgId);
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            status: options === null || options === void 0 ? void 0 : options.status,
            status_code_class: options === null || options === void 0 ? void 0 : options.statusCodeClass,
            channel: options === null || options === void 0 ? void 0 : options.channel,
            tag: options === null || options === void 0 ? void 0 : options.tag,
            endpoint_id: options === null || options === void 0 ? void 0 : options.endpointId,
            before: options === null || options === void 0 ? void 0 : options.before,
            after: options === null || options === void 0 ? void 0 : options.after,
            with_content: options === null || options === void 0 ? void 0 : options.withContent,
            event_types: options === null || options === void 0 ? void 0 : options.eventTypes
        });
        return request.send(this.requestCtx, listResponseMessageAttemptOut_1.ListResponseMessageAttemptOutSerializer._fromJsonObject);
    };
    /**
     * List messages for a particular endpoint. Additionally includes metadata about the latest message attempt.
     *
     * The `before` parameter lets you filter all items created before a certain date and is ignored if an iterator is passed.
     *
     * Note that by default this endpoint is limited to retrieving 90 days' worth of data
     * relative to now or, if an iterator is provided, 90 days before/after the time indicated
     * by the iterator ID. If you require data beyond those time ranges, you will need to explicitly
     * set the `before` or `after` parameter as appropriate.
     */
    MessageAttempt.prototype.listAttemptedMessages = function (appId, endpointId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/msg");
        request.setPathParam("app_id", appId);
        request.setPathParam("endpoint_id", endpointId);
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            channel: options === null || options === void 0 ? void 0 : options.channel,
            tag: options === null || options === void 0 ? void 0 : options.tag,
            status: options === null || options === void 0 ? void 0 : options.status,
            before: options === null || options === void 0 ? void 0 : options.before,
            after: options === null || options === void 0 ? void 0 : options.after,
            with_content: options === null || options === void 0 ? void 0 : options.withContent,
            event_types: options === null || options === void 0 ? void 0 : options.eventTypes
        });
        return request.send(this.requestCtx, listResponseEndpointMessageOut_1.ListResponseEndpointMessageOutSerializer._fromJsonObject);
    };
    /** `msg_id`: Use a message id or a message `eventId` */
    MessageAttempt.prototype.get = function (appId, msgId, attemptId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/msg/{msg_id}/attempt/{attempt_id}");
        request.setPathParam("app_id", appId);
        request.setPathParam("msg_id", msgId);
        request.setPathParam("attempt_id", attemptId);
        return request.send(this.requestCtx, messageAttemptOut_1.MessageAttemptOutSerializer._fromJsonObject);
    };
    /**
     * Deletes the given attempt's response body.
     *
     * Useful when an endpoint accidentally returned sensitive content.
     * The message can't be replayed or resent once its payload has been deleted or expired.
     */
    MessageAttempt.prototype.expungeContent = function (appId, msgId, attemptId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/app/{app_id}/msg/{msg_id}/attempt/{attempt_id}/content");
        request.setPathParam("app_id", appId);
        request.setPathParam("msg_id", msgId);
        request.setPathParam("attempt_id", attemptId);
        return request.sendNoResponseBody(this.requestCtx);
    };
    /**
     * List endpoints attempted by a given message.
     *
     * Additionally includes metadata about the latest message attempt.
     * By default, endpoints are listed in ascending order by ID.
     */
    MessageAttempt.prototype.listAttemptedDestinations = function (appId, msgId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/msg/{msg_id}/endpoint");
        request.setPathParam("app_id", appId);
        request.setPathParam("msg_id", msgId);
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator
        });
        return request.send(this.requestCtx, listResponseMessageEndpointOut_1.ListResponseMessageEndpointOutSerializer._fromJsonObject);
    };
    /** Resend a message to the specified endpoint. */
    MessageAttempt.prototype.resend = function (appId, msgId, endpointId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/msg/{msg_id}/endpoint/{endpoint_id}/resend");
        request.setPathParam("app_id", appId);
        request.setPathParam("msg_id", msgId);
        request.setPathParam("endpoint_id", endpointId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        return request.send(this.requestCtx, emptyResponse_1.EmptyResponseSerializer._fromJsonObject);
    };
    return MessageAttempt;
}());
exports.MessageAttempt = MessageAttempt;
