// svix/src/api/dist/message.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.messageInRaw = exports.Message = void 0;
var expungeAllContentsOut_1 = require("../models/expungeAllContentsOut");
var listResponseMessageOut_1 = require("../models/listResponseMessageOut");
var messageOut_1 = require("../models/messageOut");
var messagePrecheckIn_1 = require("../models/messagePrecheckIn");
var messagePrecheckOut_1 = require("../models/messagePrecheckOut");
var messagePoller_1 = require("./messagePoller");
var request_1 = require("../request");
var messageIn_1 = require("../models/messageIn");
var Message = /** @class */ (function () {
    function Message(requestCtx) {
        this.requestCtx = requestCtx;
    }
    Object.defineProperty(Message.prototype, "poller", {
        get: function () {
            return new messagePoller_1.MessagePoller(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    /**
     * List all of the application's messages.
     *
     * The `before` and `after` parameters let you filter all items created before or after a certain date. These can be
     * used alongside an iterator to paginate over results within a certain window.
     *
     * Note that by default this endpoint is limited to retrieving 90 days' worth of data
     * relative to now or, if an iterator is provided, 90 days before/after the time indicated
     * by the iterator ID. If you require data beyond those time ranges, you will need to explicitly
     * set the `before` or `after` parameter as appropriate.
     */
    Message.prototype.list = function (appId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/msg");
        request.setPathParam("app_id", appId);
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            channel: options === null || options === void 0 ? void 0 : options.channel,
            before: options === null || options === void 0 ? void 0 : options.before,
            after: options === null || options === void 0 ? void 0 : options.after,
            with_content: options === null || options === void 0 ? void 0 : options.withContent,
            tag: options === null || options === void 0 ? void 0 : options.tag,
            event_types: options === null || options === void 0 ? void 0 : options.eventTypes
        });
        return request.send(this.requestCtx, listResponseMessageOut_1.ListResponseMessageOutSerializer._fromJsonObject);
    };
    /**
     * Creates a new message and dispatches it to all of the application's endpoints.
     *
     * The `eventId` is an optional custom unique ID. It's verified to be unique only up to a day, after that no verification will be made.
     * If a message with the same `eventId` already exists for the application, a 409 conflict error will be returned.
     *
     * The `eventType` indicates the type and schema of the event. All messages of a certain `eventType` are expected to have the same schema. Endpoints can choose to only listen to specific event types.
     * Messages can also have `channels`, which similar to event types let endpoints filter by them. Unlike event types, messages can have multiple channels, and channels don't imply a specific message content or schema.
     *
     * The `payload` property is the webhook's body (the actual webhook message). Svix supports payload sizes of up to 1MiB, though it's generally a good idea to keep webhook payloads small, probably no larger than 40kb.
     */
    Message.prototype.create = function (appId, messageIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/msg");
        request.setPathParam("app_id", appId);
        request.setQueryParams({
            with_content: options === null || options === void 0 ? void 0 : options.withContent
        });
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(messageIn_1.MessageInSerializer._toJsonObject(messageIn));
        return request.send(this.requestCtx, messageOut_1.MessageOutSerializer._fromJsonObject);
    };
    /**
     * Delete all message payloads for the application.
     *
     * This operation is only available in the <a href="https://svix.com/pricing" target="_blank">Enterprise</a> plan.
     *
     * A completed task will return a payload like the following:
     * ```json
     * {
     *   "id": "qtask_33qen93MNuelBAq1T9G7eHLJRsF",
     *   "status": "finished",
     *   "task": "application.purge_content",
     *   "data": {
     *     "messagesPurged": 150
     *   }
     * }
     * ```
     */
    Message.prototype.expungeAllContents = function (appId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/msg/expunge-all-contents");
        request.setPathParam("app_id", appId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        return request.send(this.requestCtx, expungeAllContentsOut_1.ExpungeAllContentsOutSerializer._fromJsonObject);
    };
    /**
     * A pre-check call for `message.create` that checks whether any active endpoints are
     * listening to this message.
     *
     * Note: most people shouldn't be using this API. Svix doesn't bill you for
     * messages not actually sent, so using this API doesn't save money.
     * If unsure, please ask Svix support before using this API.
     */
    Message.prototype.precheck = function (appId, messagePrecheckIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/msg/precheck/active");
        request.setPathParam("app_id", appId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(messagePrecheckIn_1.MessagePrecheckInSerializer._toJsonObject(messagePrecheckIn));
        return request.send(this.requestCtx, messagePrecheckOut_1.MessagePrecheckOutSerializer._fromJsonObject);
    };
    /** Get a message by its ID or eventID. */
    Message.prototype.get = function (appId, msgId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/msg/{msg_id}");
        request.setPathParam("app_id", appId);
        request.setPathParam("msg_id", msgId);
        request.setQueryParams({
            with_content: options === null || options === void 0 ? void 0 : options.withContent
        });
        return request.send(this.requestCtx, messageOut_1.MessageOutSerializer._fromJsonObject);
    };
    /**
     * Delete the given message's payload.
     *
     * Useful in cases when a message was accidentally sent with sensitive content.
     * The message can't be replayed or resent once its payload has been deleted or expired.
     */
    Message.prototype.expungeContent = function (appId, msgId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/app/{app_id}/msg/{msg_id}/content");
        request.setPathParam("app_id", appId);
        request.setPathParam("msg_id", msgId);
        return request.sendNoResponseBody(this.requestCtx);
    };
    return Message;
}());
exports.Message = Message;
/**
 * Creates a `MessageIn` with a raw string payload.
 *
 * The payload is not normalized on the server. Normally, payloads are
 * required to be JSON, and Svix will minify the payload before sending the
 * webhooks (for example, by removing extraneous whitespace or unnecessarily
 * escaped characters in strings). With this function, the payload will be
 * sent "as is", without any minification or other processing.
 *
 * @param payload Serialized message payload
 * @param contentType The value to use for the Content-Type header of the webhook sent by Svix, overwriting the default of `application/json` if specified
 *
 * See the class documentation for details about the other parameters.
 */
function messageInRaw(eventType, payload, contentType) {
    var headers = contentType ? { "content-type": contentType } : undefined;
    return {
        eventType: eventType,
        payload: {},
        transformationsParams: {
            rawPayload: payload,
            headers: headers
        }
    };
}
exports.messageInRaw = messageInRaw;
