// svix/src/api/dist/messagePoller.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.MessagePoller = void 0;
var pollingEndpointConsumerSeekIn_1 = require("../models/pollingEndpointConsumerSeekIn");
var pollingEndpointConsumerSeekOut_1 = require("../models/pollingEndpointConsumerSeekOut");
var pollingEndpointOut_1 = require("../models/pollingEndpointOut");
var request_1 = require("../request");
var MessagePoller = /** @class */ (function () {
    function MessagePoller(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /** Reads the stream of created messages for an application, filtered on the Sink's event types and Channels. */
    MessagePoller.prototype.poll = function (appId, sinkId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/poller/{sink_id}");
        request.setPathParam("app_id", appId);
        request.setPathParam("sink_id", sinkId);
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            event_type: options === null || options === void 0 ? void 0 : options.eventType,
            channel: options === null || options === void 0 ? void 0 : options.channel,
            after: options === null || options === void 0 ? void 0 : options.after
        });
        return request.send(this.requestCtx, pollingEndpointOut_1.PollingEndpointOutSerializer._fromJsonObject);
    };
    /**
     * Reads the stream of created messages for an application, filtered on the Sink's event types and
     * Channels, using server-managed iterator tracking.
     */
    MessagePoller.prototype.consumerPoll = function (appId, sinkId, consumerId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/poller/{sink_id}/consumer/{consumer_id}");
        request.setPathParam("app_id", appId);
        request.setPathParam("sink_id", sinkId);
        request.setPathParam("consumer_id", consumerId);
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator
        });
        return request.send(this.requestCtx, pollingEndpointOut_1.PollingEndpointOutSerializer._fromJsonObject);
    };
    /** Sets the starting offset for the consumer of a polling endpoint. */
    MessagePoller.prototype.consumerSeek = function (appId, sinkId, consumerId, pollingEndpointConsumerSeekIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/poller/{sink_id}/consumer/{consumer_id}/seek");
        request.setPathParam("app_id", appId);
        request.setPathParam("sink_id", sinkId);
        request.setPathParam("consumer_id", consumerId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(pollingEndpointConsumerSeekIn_1.PollingEndpointConsumerSeekInSerializer._toJsonObject(pollingEndpointConsumerSeekIn));
        return request.send(this.requestCtx, pollingEndpointConsumerSeekOut_1.PollingEndpointConsumerSeekOutSerializer._fromJsonObject);
    };
    return MessagePoller;
}());
exports.MessagePoller = MessagePoller;
