// svix/src/api/dist/operationalWebhookEndpoint.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.OperationalWebhookEndpoint = void 0;
var listResponseOperationalWebhookEndpointOut_1 = require("../models/listResponseOperationalWebhookEndpointOut");
var operationalWebhookEndpointHeadersIn_1 = require("../models/operationalWebhookEndpointHeadersIn");
var operationalWebhookEndpointHeadersOut_1 = require("../models/operationalWebhookEndpointHeadersOut");
var operationalWebhookEndpointIn_1 = require("../models/operationalWebhookEndpointIn");
var operationalWebhookEndpointOut_1 = require("../models/operationalWebhookEndpointOut");
var operationalWebhookEndpointSecretIn_1 = require("../models/operationalWebhookEndpointSecretIn");
var operationalWebhookEndpointSecretOut_1 = require("../models/operationalWebhookEndpointSecretOut");
var operationalWebhookEndpointUpdate_1 = require("../models/operationalWebhookEndpointUpdate");
var request_1 = require("../request");
var OperationalWebhookEndpoint = /** @class */ (function () {
    function OperationalWebhookEndpoint(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /** List operational webhook endpoints. */
    OperationalWebhookEndpoint.prototype.list = function (options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/operational-webhook/endpoint");
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            order: options === null || options === void 0 ? void 0 : options.order
        });
        return request.send(this.requestCtx, listResponseOperationalWebhookEndpointOut_1.ListResponseOperationalWebhookEndpointOutSerializer._fromJsonObject);
    };
    /** Create an operational webhook endpoint. */
    OperationalWebhookEndpoint.prototype.create = function (operationalWebhookEndpointIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/operational-webhook/endpoint");
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(operationalWebhookEndpointIn_1.OperationalWebhookEndpointInSerializer._toJsonObject(operationalWebhookEndpointIn));
        return request.send(this.requestCtx, operationalWebhookEndpointOut_1.OperationalWebhookEndpointOutSerializer._fromJsonObject);
    };
    /** Get an operational webhook endpoint. */
    OperationalWebhookEndpoint.prototype.get = function (endpointId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/operational-webhook/endpoint/{endpoint_id}");
        request.setPathParam("endpoint_id", endpointId);
        return request.send(this.requestCtx, operationalWebhookEndpointOut_1.OperationalWebhookEndpointOutSerializer._fromJsonObject);
    };
    /** Update an operational webhook endpoint. */
    OperationalWebhookEndpoint.prototype.update = function (endpointId, operationalWebhookEndpointUpdate) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/operational-webhook/endpoint/{endpoint_id}");
        request.setPathParam("endpoint_id", endpointId);
        request.setBody(operationalWebhookEndpointUpdate_1.OperationalWebhookEndpointUpdateSerializer._toJsonObject(operationalWebhookEndpointUpdate));
        return request.send(this.requestCtx, operationalWebhookEndpointOut_1.OperationalWebhookEndpointOutSerializer._fromJsonObject);
    };
    /** Delete an operational webhook endpoint. */
    OperationalWebhookEndpoint.prototype["delete"] = function (endpointId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/operational-webhook/endpoint/{endpoint_id}");
        request.setPathParam("endpoint_id", endpointId);
        return request.sendNoResponseBody(this.requestCtx);
    };
    /** Get the additional headers to be sent with the operational webhook. */
    OperationalWebhookEndpoint.prototype.getHeaders = function (endpointId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/operational-webhook/endpoint/{endpoint_id}/headers");
        request.setPathParam("endpoint_id", endpointId);
        return request.send(this.requestCtx, operationalWebhookEndpointHeadersOut_1.OperationalWebhookEndpointHeadersOutSerializer._fromJsonObject);
    };
    /** Set the additional headers to be sent with the operational webhook. */
    OperationalWebhookEndpoint.prototype.updateHeaders = function (endpointId, operationalWebhookEndpointHeadersIn) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/operational-webhook/endpoint/{endpoint_id}/headers");
        request.setPathParam("endpoint_id", endpointId);
        request.setBody(operationalWebhookEndpointHeadersIn_1.OperationalWebhookEndpointHeadersInSerializer._toJsonObject(operationalWebhookEndpointHeadersIn));
        return request.sendNoResponseBody(this.requestCtx);
    };
    /**
     * Get an operational webhook endpoint's signing secret.
     *
     * This is used to verify the authenticity of the webhook.
     * For more information please refer to [the consuming webhooks docs](https://docs.svix.com/consuming-webhooks/).
     */
    OperationalWebhookEndpoint.prototype.getSecret = function (endpointId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/operational-webhook/endpoint/{endpoint_id}/secret");
        request.setPathParam("endpoint_id", endpointId);
        return request.send(this.requestCtx, operationalWebhookEndpointSecretOut_1.OperationalWebhookEndpointSecretOutSerializer._fromJsonObject);
    };
    /**
     * Rotates an operational webhook endpoint's signing secret.
     *
     * The previous secret will remain valid for the next 24 hours.
     */
    OperationalWebhookEndpoint.prototype.rotateSecret = function (endpointId, operationalWebhookEndpointSecretIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/operational-webhook/endpoint/{endpoint_id}/secret/rotate");
        request.setPathParam("endpoint_id", endpointId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(operationalWebhookEndpointSecretIn_1.OperationalWebhookEndpointSecretInSerializer._toJsonObject(operationalWebhookEndpointSecretIn));
        return request.sendNoResponseBody(this.requestCtx);
    };
    return OperationalWebhookEndpoint;
}());
exports.OperationalWebhookEndpoint = OperationalWebhookEndpoint;
