// svix/src/api/dist/streamingSink.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.StreamingSink = void 0;
var emptyResponse_1 = require("../models/emptyResponse");
var endpointSecretRotateIn_1 = require("../models/endpointSecretRotateIn");
var listResponseStreamSinkOut_1 = require("../models/listResponseStreamSinkOut");
var sinkSecretOut_1 = require("../models/sinkSecretOut");
var sinkTransformIn_1 = require("../models/sinkTransformIn");
var streamSinkIn_1 = require("../models/streamSinkIn");
var streamSinkOut_1 = require("../models/streamSinkOut");
var streamSinkPatch_1 = require("../models/streamSinkPatch");
var request_1 = require("../request");
var StreamingSink = /** @class */ (function () {
    function StreamingSink(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /** List of all the stream's sinks. */
    StreamingSink.prototype.list = function (streamId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}/sink");
        request.setPathParam("stream_id", streamId);
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            order: options === null || options === void 0 ? void 0 : options.order
        });
        return request.send(this.requestCtx, listResponseStreamSinkOut_1.ListResponseStreamSinkOutSerializer._fromJsonObject);
    };
    /** Creates a new sink. */
    StreamingSink.prototype.create = function (streamId, streamSinkIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/stream/{stream_id}/sink");
        request.setPathParam("stream_id", streamId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(streamSinkIn_1.StreamSinkInSerializer._toJsonObject(streamSinkIn));
        return request.send(this.requestCtx, streamSinkOut_1.StreamSinkOutSerializer._fromJsonObject);
    };
    /** Get a sink by id or uid. */
    StreamingSink.prototype.get = function (streamId, sinkId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}/sink/{sink_id}");
        request.setPathParam("stream_id", streamId);
        request.setPathParam("sink_id", sinkId);
        return request.send(this.requestCtx, streamSinkOut_1.StreamSinkOutSerializer._fromJsonObject);
    };
    /** Update a sink. */
    StreamingSink.prototype.update = function (streamId, sinkId, streamSinkIn) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/stream/{stream_id}/sink/{sink_id}");
        request.setPathParam("stream_id", streamId);
        request.setPathParam("sink_id", sinkId);
        request.setBody(streamSinkIn_1.StreamSinkInSerializer._toJsonObject(streamSinkIn));
        return request.send(this.requestCtx, streamSinkOut_1.StreamSinkOutSerializer._fromJsonObject);
    };
    /** Delete a sink. */
    StreamingSink.prototype["delete"] = function (streamId, sinkId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/stream/{stream_id}/sink/{sink_id}");
        request.setPathParam("stream_id", streamId);
        request.setPathParam("sink_id", sinkId);
        return request.sendNoResponseBody(this.requestCtx);
    };
    /** Partially update a sink. */
    StreamingSink.prototype.patch = function (streamId, sinkId, streamSinkPatch) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/stream/{stream_id}/sink/{sink_id}");
        request.setPathParam("stream_id", streamId);
        request.setPathParam("sink_id", sinkId);
        request.setBody(streamSinkPatch_1.StreamSinkPatchSerializer._toJsonObject(streamSinkPatch));
        return request.send(this.requestCtx, streamSinkOut_1.StreamSinkOutSerializer._fromJsonObject);
    };
    /**
     * Get the sink's signing secret (only supported for http sinks)
     *
     * This is used to verify the authenticity of the delivery.
     *
     * For more information please refer to [the consuming webhooks docs](https://docs.svix.com/consuming-webhooks/).
     */
    StreamingSink.prototype.getSecret = function (streamId, sinkId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}/sink/{sink_id}/secret");
        request.setPathParam("stream_id", streamId);
        request.setPathParam("sink_id", sinkId);
        return request.send(this.requestCtx, sinkSecretOut_1.SinkSecretOutSerializer._fromJsonObject);
    };
    /** Rotates the signing secret (only supported for http sinks). */
    StreamingSink.prototype.rotateSecret = function (streamId, sinkId, endpointSecretRotateIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/stream/{stream_id}/sink/{sink_id}/secret/rotate");
        request.setPathParam("stream_id", streamId);
        request.setPathParam("sink_id", sinkId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(endpointSecretRotateIn_1.EndpointSecretRotateInSerializer._toJsonObject(endpointSecretRotateIn));
        return request.send(this.requestCtx, emptyResponse_1.EmptyResponseSerializer._fromJsonObject);
    };
    /** Set or unset the transformation code associated with this sink. */
    StreamingSink.prototype.transformationPartialUpdate = function (streamId, sinkId, sinkTransformIn) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/stream/{stream_id}/sink/{sink_id}/transformation");
        request.setPathParam("stream_id", streamId);
        request.setPathParam("sink_id", sinkId);
        request.setBody(sinkTransformIn_1.SinkTransformInSerializer._toJsonObject(sinkTransformIn));
        return request.send(this.requestCtx, emptyResponse_1.EmptyResponseSerializer._fromJsonObject);
    };
    return StreamingSink;
}());
exports.StreamingSink = StreamingSink;
