// svix/src/api/dist/streamingStream.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.StreamingStream = void 0;
var listResponseStreamOut_1 = require("../models/listResponseStreamOut");
var streamIn_1 = require("../models/streamIn");
var streamOut_1 = require("../models/streamOut");
var streamPatch_1 = require("../models/streamPatch");
var request_1 = require("../request");
var StreamingStream = /** @class */ (function () {
    function StreamingStream(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /** List of all the organization's streams. */
    StreamingStream.prototype.list = function (options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream");
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            order: options === null || options === void 0 ? void 0 : options.order
        });
        return request.send(this.requestCtx, listResponseStreamOut_1.ListResponseStreamOutSerializer._fromJsonObject);
    };
    /** Creates a new stream. */
    StreamingStream.prototype.create = function (streamIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/stream");
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(streamIn_1.StreamInSerializer._toJsonObject(streamIn));
        return request.send(this.requestCtx, streamOut_1.StreamOutSerializer._fromJsonObject);
    };
    /** Get a stream by id or uid. */
    StreamingStream.prototype.get = function (streamId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}");
        request.setPathParam("stream_id", streamId);
        return request.send(this.requestCtx, streamOut_1.StreamOutSerializer._fromJsonObject);
    };
    /** Update a stream. */
    StreamingStream.prototype.update = function (streamId, streamIn) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/stream/{stream_id}");
        request.setPathParam("stream_id", streamId);
        request.setBody(streamIn_1.StreamInSerializer._toJsonObject(streamIn));
        return request.send(this.requestCtx, streamOut_1.StreamOutSerializer._fromJsonObject);
    };
    /** Delete a stream. */
    StreamingStream.prototype["delete"] = function (streamId) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/stream/{stream_id}");
        request.setPathParam("stream_id", streamId);
        return request.sendNoResponseBody(this.requestCtx);
    };
    /** Partially update a stream. */
    StreamingStream.prototype.patch = function (streamId, streamPatch) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/stream/{stream_id}");
        request.setPathParam("stream_id", streamId);
        request.setBody(streamPatch_1.StreamPatchSerializer._toJsonObject(streamPatch));
        return request.send(this.requestCtx, streamOut_1.StreamOutSerializer._fromJsonObject);
    };
    return StreamingStream;
}());
exports.StreamingStream = StreamingStream;
