// svix/src/api/dist/streamingEvents.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.StreamingEvents = void 0;
var createStreamEventsIn_1 = require("../models/createStreamEventsIn");
var createStreamEventsOut_1 = require("../models/createStreamEventsOut");
var eventStreamOut_1 = require("../models/eventStreamOut");
var request_1 = require("../request");
var StreamingEvents = /** @class */ (function () {
    function StreamingEvents(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /** Creates events on the Stream. */
    StreamingEvents.prototype.create = function (streamId, createStreamEventsIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/stream/{stream_id}/events");
        request.setPathParam("stream_id", streamId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(createStreamEventsIn_1.CreateStreamEventsInSerializer._toJsonObject(createStreamEventsIn));
        return request.send(this.requestCtx, createStreamEventsOut_1.CreateStreamEventsOutSerializer._fromJsonObject);
    };
    /**
     * Iterate over a stream of events.
     *
     * The sink must be of type `poller` to use the poller endpoint.
     */
    StreamingEvents.prototype.get = function (streamId, sinkId, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}/sink/{sink_id}/events");
        request.setPathParam("stream_id", streamId);
        request.setPathParam("sink_id", sinkId);
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            after: options === null || options === void 0 ? void 0 : options.after
        });
        return request.send(this.requestCtx, eventStreamOut_1.EventStreamOutSerializer._fromJsonObject);
    };
    return StreamingEvents;
}());
exports.StreamingEvents = StreamingEvents;
