// svix/src/api/dist/streamingEventType.js
"use strict";
// this file is @generated
exports.__esModule = true;
exports.StreamingEventType = void 0;
var listResponseStreamEventTypeOut_1 = require("../models/listResponseStreamEventTypeOut");
var streamEventTypeIn_1 = require("../models/streamEventTypeIn");
var streamEventTypeOut_1 = require("../models/streamEventTypeOut");
var streamEventTypePatch_1 = require("../models/streamEventTypePatch");
var request_1 = require("../request");
var StreamingEventType = /** @class */ (function () {
    function StreamingEventType(requestCtx) {
        this.requestCtx = requestCtx;
    }
    /** List of all the organization's event types for streaming. */
    StreamingEventType.prototype.list = function (options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/event-type");
        request.setQueryParams({
            limit: options === null || options === void 0 ? void 0 : options.limit,
            iterator: options === null || options === void 0 ? void 0 : options.iterator,
            order: options === null || options === void 0 ? void 0 : options.order,
            include_archived: options === null || options === void 0 ? void 0 : options.includeArchived
        });
        return request.send(this.requestCtx, listResponseStreamEventTypeOut_1.ListResponseStreamEventTypeOutSerializer._fromJsonObject);
    };
    /** Create an event type for Streams. */
    StreamingEventType.prototype.create = function (streamEventTypeIn, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/stream/event-type");
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(streamEventTypeIn_1.StreamEventTypeInSerializer._toJsonObject(streamEventTypeIn));
        return request.send(this.requestCtx, streamEventTypeOut_1.StreamEventTypeOutSerializer._fromJsonObject);
    };
    /** Get an event type. */
    StreamingEventType.prototype.get = function (name) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/event-type/{name}");
        request.setPathParam("name", name);
        return request.send(this.requestCtx, streamEventTypeOut_1.StreamEventTypeOutSerializer._fromJsonObject);
    };
    /** Update or create a event type for Streams. */
    StreamingEventType.prototype.update = function (name, streamEventTypeIn) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/stream/event-type/{name}");
        request.setPathParam("name", name);
        request.setBody(streamEventTypeIn_1.StreamEventTypeInSerializer._toJsonObject(streamEventTypeIn));
        return request.send(this.requestCtx, streamEventTypeOut_1.StreamEventTypeOutSerializer._fromJsonObject);
    };
    /** Delete an event type. */
    StreamingEventType.prototype["delete"] = function (name, options) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/stream/event-type/{name}");
        request.setPathParam("name", name);
        request.setQueryParams({
            expunge: options === null || options === void 0 ? void 0 : options.expunge
        });
        return request.sendNoResponseBody(this.requestCtx);
    };
    /** Patch an event type for Streams. */
    StreamingEventType.prototype.patch = function (name, streamEventTypePatch) {
        var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/stream/event-type/{name}");
        request.setPathParam("name", name);
        request.setBody(streamEventTypePatch_1.StreamEventTypePatchSerializer._toJsonObject(streamEventTypePatch));
        return request.send(this.requestCtx, streamEventTypeOut_1.StreamEventTypeOutSerializer._fromJsonObject);
    };
    return StreamingEventType;
}());
exports.StreamingEventType = StreamingEventType;
