// svix/dist/models/dist/environmentOut.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EnvironmentOutSerializer = void 0;

var connectorOut_1 = require("./connectorOut");

var eventTypeOut_1 = require("./eventTypeOut");

exports.EnvironmentOutSerializer = {
  _fromJsonObject: function _fromJsonObject(object) {
    return {
      connectors: object["connectors"].map(function (item) {
        return connectorOut_1.ConnectorOutSerializer._fromJsonObject(item);
      }),
      createdAt: new Date(object["createdAt"]),
      eventTypes: object["eventTypes"].map(function (item) {
        return eventTypeOut_1.EventTypeOutSerializer._fromJsonObject(item);
      }),
      settings: object["settings"],
      version: object["version"]
    };
  },
  _toJsonObject: function _toJsonObject(self) {
    return {
      connectors: self.connectors.map(function (item) {
        return connectorOut_1.ConnectorOutSerializer._toJsonObject(item);
      }),
      createdAt: self.createdAt,
      eventTypes: self.eventTypes.map(function (item) {
        return eventTypeOut_1.EventTypeOutSerializer._toJsonObject(item);
      }),
      settings: self.settings,
      version: self.version
    };
  }
};
