// svix/dist/models/dist/ingestSourceOut.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IngestSourceOutSerializer = void 0;

var adobeSignConfigOut_1 = require("./adobeSignConfigOut");

var airwallexConfigOut_1 = require("./airwallexConfigOut");

var checkbookConfigOut_1 = require("./checkbookConfigOut");

var cronConfig_1 = require("./cronConfig");

var docusignConfigOut_1 = require("./docusignConfigOut");

var easypostConfigOut_1 = require("./easypostConfigOut");

var githubConfigOut_1 = require("./githubConfigOut");

var hubspotConfigOut_1 = require("./hubspotConfigOut");

var metaConfigOut_1 = require("./metaConfigOut");

var orumIoConfigOut_1 = require("./orumIoConfigOut");

var pandaDocConfigOut_1 = require("./pandaDocConfigOut");

var portIoConfigOut_1 = require("./portIoConfigOut");

var rutterConfigOut_1 = require("./rutterConfigOut");

var segmentConfigOut_1 = require("./segmentConfigOut");

var shopifyConfigOut_1 = require("./shopifyConfigOut");

var slackConfigOut_1 = require("./slackConfigOut");

var stripeConfigOut_1 = require("./stripeConfigOut");

var svixConfigOut_1 = require("./svixConfigOut");

var telnyxConfigOut_1 = require("./telnyxConfigOut");

var vapiConfigOut_1 = require("./vapiConfigOut");

var veriffConfigOut_1 = require("./veriffConfigOut");

var zoomConfigOut_1 = require("./zoomConfigOut");

exports.IngestSourceOutSerializer = {
  _fromJsonObject: function _fromJsonObject(object) {
    var type = object["type"];

    function getConfig(type) {
      switch (type) {
        case "generic-webhook":
          return {};

        case "cron":
          return cronConfig_1.CronConfigSerializer._fromJsonObject(object["config"]);

        case "adobe-sign":
          return adobeSignConfigOut_1.AdobeSignConfigOutSerializer._fromJsonObject(object["config"]);

        case "beehiiv":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "brex":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "checkbook":
          return checkbookConfigOut_1.CheckbookConfigOutSerializer._fromJsonObject(object["config"]);

        case "clerk":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "docusign":
          return docusignConfigOut_1.DocusignConfigOutSerializer._fromJsonObject(object["config"]);

        case "easypost":
          return easypostConfigOut_1.EasypostConfigOutSerializer._fromJsonObject(object["config"]);

        case "github":
          return githubConfigOut_1.GithubConfigOutSerializer._fromJsonObject(object["config"]);

        case "guesty":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "hubspot":
          return hubspotConfigOut_1.HubspotConfigOutSerializer._fromJsonObject(object["config"]);

        case "incident-io":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "lithic":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "meta":
          return metaConfigOut_1.MetaConfigOutSerializer._fromJsonObject(object["config"]);

        case "nash":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "orum-io":
          return orumIoConfigOut_1.OrumIoConfigOutSerializer._fromJsonObject(object["config"]);

        case "panda-doc":
          return pandaDocConfigOut_1.PandaDocConfigOutSerializer._fromJsonObject(object["config"]);

        case "port-io":
          return portIoConfigOut_1.PortIoConfigOutSerializer._fromJsonObject(object["config"]);

        case "psi-fi":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "pleo":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "replicate":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "resend":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "rutter":
          return rutterConfigOut_1.RutterConfigOutSerializer._fromJsonObject(object["config"]);

        case "safebase":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "sardine":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "segment":
          return segmentConfigOut_1.SegmentConfigOutSerializer._fromJsonObject(object["config"]);

        case "shopify":
          return shopifyConfigOut_1.ShopifyConfigOutSerializer._fromJsonObject(object["config"]);

        case "slack":
          return slackConfigOut_1.SlackConfigOutSerializer._fromJsonObject(object["config"]);

        case "stripe":
          return stripeConfigOut_1.StripeConfigOutSerializer._fromJsonObject(object["config"]);

        case "stych":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "svix":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "zoom":
          return zoomConfigOut_1.ZoomConfigOutSerializer._fromJsonObject(object["config"]);

        case "telnyx":
          return telnyxConfigOut_1.TelnyxConfigOutSerializer._fromJsonObject(object["config"]);

        case "vapi":
          return vapiConfigOut_1.VapiConfigOutSerializer._fromJsonObject(object["config"]);

        case "open-ai":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "render":
          return svixConfigOut_1.SvixConfigOutSerializer._fromJsonObject(object["config"]);

        case "veriff":
          return veriffConfigOut_1.VeriffConfigOutSerializer._fromJsonObject(object["config"]);

        case "airwallex":
          return airwallexConfigOut_1.AirwallexConfigOutSerializer._fromJsonObject(object["config"]);

        default:
          throw new Error("Unexpected type: ".concat(type));
      }
    }

    return {
      type: type,
      config: getConfig(type),
      createdAt: new Date(object["createdAt"]),
      id: object["id"],
      ingestUrl: object["ingestUrl"],
      metadata: object["metadata"],
      name: object["name"],
      uid: object["uid"],
      updatedAt: new Date(object["updatedAt"])
    };
  },
  _toJsonObject: function _toJsonObject(self) {
    var config;

    switch (self.type) {
      case "generic-webhook":
        config = {};
        break;

      case "cron":
        config = cronConfig_1.CronConfigSerializer._toJsonObject(self.config);
        break;

      case "adobe-sign":
        config = adobeSignConfigOut_1.AdobeSignConfigOutSerializer._toJsonObject(self.config);
        break;

      case "beehiiv":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "brex":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "checkbook":
        config = checkbookConfigOut_1.CheckbookConfigOutSerializer._toJsonObject(self.config);
        break;

      case "clerk":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "docusign":
        config = docusignConfigOut_1.DocusignConfigOutSerializer._toJsonObject(self.config);
        break;

      case "easypost":
        config = easypostConfigOut_1.EasypostConfigOutSerializer._toJsonObject(self.config);
        break;

      case "github":
        config = githubConfigOut_1.GithubConfigOutSerializer._toJsonObject(self.config);
        break;

      case "guesty":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "hubspot":
        config = hubspotConfigOut_1.HubspotConfigOutSerializer._toJsonObject(self.config);
        break;

      case "incident-io":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "lithic":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "meta":
        config = metaConfigOut_1.MetaConfigOutSerializer._toJsonObject(self.config);
        break;

      case "nash":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "orum-io":
        config = orumIoConfigOut_1.OrumIoConfigOutSerializer._toJsonObject(self.config);
        break;

      case "panda-doc":
        config = pandaDocConfigOut_1.PandaDocConfigOutSerializer._toJsonObject(self.config);
        break;

      case "port-io":
        config = portIoConfigOut_1.PortIoConfigOutSerializer._toJsonObject(self.config);
        break;

      case "psi-fi":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "pleo":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "replicate":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "resend":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "rutter":
        config = rutterConfigOut_1.RutterConfigOutSerializer._toJsonObject(self.config);
        break;

      case "safebase":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "sardine":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "segment":
        config = segmentConfigOut_1.SegmentConfigOutSerializer._toJsonObject(self.config);
        break;

      case "shopify":
        config = shopifyConfigOut_1.ShopifyConfigOutSerializer._toJsonObject(self.config);
        break;

      case "slack":
        config = slackConfigOut_1.SlackConfigOutSerializer._toJsonObject(self.config);
        break;

      case "stripe":
        config = stripeConfigOut_1.StripeConfigOutSerializer._toJsonObject(self.config);
        break;

      case "stych":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "svix":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "zoom":
        config = zoomConfigOut_1.ZoomConfigOutSerializer._toJsonObject(self.config);
        break;

      case "telnyx":
        config = telnyxConfigOut_1.TelnyxConfigOutSerializer._toJsonObject(self.config);
        break;

      case "vapi":
        config = vapiConfigOut_1.VapiConfigOutSerializer._toJsonObject(self.config);
        break;

      case "open-ai":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "render":
        config = svixConfigOut_1.SvixConfigOutSerializer._toJsonObject(self.config);
        break;

      case "veriff":
        config = veriffConfigOut_1.VeriffConfigOutSerializer._toJsonObject(self.config);
        break;

      case "airwallex":
        config = airwallexConfigOut_1.AirwallexConfigOutSerializer._toJsonObject(self.config);
        break;
    }

    return {
      type: self.type,
      config: config,
      createdAt: self.createdAt,
      id: self.id,
      ingestUrl: self.ingestUrl,
      metadata: self.metadata,
      name: self.name,
      uid: self.uid,
      updatedAt: self.updatedAt
    };
  }
};
