// mime-types/dist/mimeScore.dev.js
"use strict";

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

// 'mime-score' back-ported to CommonJS
// Score RFC facets (see https://tools.ietf.org/html/rfc6838#section-3)
var FACET_SCORES = {
  'prs.': 100,
  'x-': 200,
  'x.': 300,
  'vnd.': 400,
  "default": 900
}; // Score mime source (Logic originally from `jshttp/mime-types` module)

var SOURCE_SCORES = {
  nginx: 10,
  apache: 20,
  iana: 40,
  "default": 30 // definitions added by `jshttp/mime-db` project?

};
var TYPE_SCORES = {
  // prefer application/xml over text/xml
  // prefer application/rtf over text/rtf
  application: 1,
  // prefer font/woff over application/font-woff
  font: 2,
  // prefer video/mp4 over audio/mp4 over application/mp4
  // See https://www.rfc-editor.org/rfc/rfc4337.html#section-2
  audio: 2,
  video: 3,
  "default": 0
};
/**
 * Get each component of the score for a mime type.  The sum of these is the
 * total score.  The higher the score, the more "official" the type.
 */

module.exports = function mimeScore(mimeType) {
  var source = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'default';

  if (mimeType === 'application/octet-stream') {
    return 0;
  }

  var _mimeType$split = mimeType.split('/'),
      _mimeType$split2 = _slicedToArray(_mimeType$split, 2),
      type = _mimeType$split2[0],
      subtype = _mimeType$split2[1];

  var facet = subtype.replace(/(\.|x-).*/, '$1');
  var facetScore = FACET_SCORES[facet] || FACET_SCORES["default"];
  var sourceScore = SOURCE_SCORES[source] || SOURCE_SCORES["default"];
  var typeScore = TYPE_SCORES[type] || TYPE_SCORES["default"]; // All else being equal prefer shorter types

  var lengthScore = 1 - mimeType.length / 100;
  return facetScore + sourceScore + typeScore + lengthScore;
};
