// svix/dist/api/dist/application.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Application = void 0;

var applicationIn_1 = require("../models/applicationIn");

var applicationOut_1 = require("../models/applicationOut");

var applicationPatch_1 = require("../models/applicationPatch");

var listResponseApplicationOut_1 = require("../models/listResponseApplicationOut");

var request_1 = require("../request");

var Application =
/*#__PURE__*/
function () {
  function Application(requestCtx) {
    _classCallCheck(this, Application);

    this.requestCtx = requestCtx;
  }

  _createClass(Application, [{
    key: "list",
    value: function list(options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app");
      request.setQueryParams({
        exclude_apps_with_no_endpoints: options === null || options === void 0 ? void 0 : options.excludeAppsWithNoEndpoints,
        exclude_apps_with_disabled_endpoints: options === null || options === void 0 ? void 0 : options.excludeAppsWithDisabledEndpoints,
        exclude_apps_with_svix_play_endpoints: options === null || options === void 0 ? void 0 : options.excludeAppsWithSvixPlayEndpoints,
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        order: options === null || options === void 0 ? void 0 : options.order
      });
      return request.send(this.requestCtx, listResponseApplicationOut_1.ListResponseApplicationOutSerializer._fromJsonObject);
    }
  }, {
    key: "create",
    value: function create(applicationIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app");
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(applicationIn_1.ApplicationInSerializer._toJsonObject(applicationIn));
      return request.send(this.requestCtx, applicationOut_1.ApplicationOutSerializer._fromJsonObject);
    }
  }, {
    key: "getOrCreate",
    value: function getOrCreate(applicationIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app");
      request.setQueryParam("get_if_exists", true);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(applicationIn_1.ApplicationInSerializer._toJsonObject(applicationIn));
      return request.send(this.requestCtx, applicationOut_1.ApplicationOutSerializer._fromJsonObject);
    }
  }, {
    key: "get",
    value: function get(appId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}");
      request.setPathParam("app_id", appId);
      return request.send(this.requestCtx, applicationOut_1.ApplicationOutSerializer._fromJsonObject);
    }
  }, {
    key: "update",
    value: function update(appId, applicationIn) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/app/{app_id}");
      request.setPathParam("app_id", appId);
      request.setBody(applicationIn_1.ApplicationInSerializer._toJsonObject(applicationIn));
      return request.send(this.requestCtx, applicationOut_1.ApplicationOutSerializer._fromJsonObject);
    }
  }, {
    key: "delete",
    value: function _delete(appId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/app/{app_id}");
      request.setPathParam("app_id", appId);
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "patch",
    value: function patch(appId, applicationPatch) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/app/{app_id}");
      request.setPathParam("app_id", appId);
      request.setBody(applicationPatch_1.ApplicationPatchSerializer._toJsonObject(applicationPatch));
      return request.send(this.requestCtx, applicationOut_1.ApplicationOutSerializer._fromJsonObject);
    }
  }]);

  return Application;
}();

exports.Application = Application;
