// svix/dist/api/dist/authentication.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Authentication = void 0;

var apiTokenOut_1 = require("../models/apiTokenOut");

var appPortalAccessIn_1 = require("../models/appPortalAccessIn");

var appPortalAccessOut_1 = require("../models/appPortalAccessOut");

var applicationTokenExpireIn_1 = require("../models/applicationTokenExpireIn");

var rotatePollerTokenIn_1 = require("../models/rotatePollerTokenIn");

var streamPortalAccessIn_1 = require("../models/streamPortalAccessIn");

var streamTokenExpireIn_1 = require("../models/streamTokenExpireIn");

var dashboardAccessOut_1 = require("../models/dashboardAccessOut");

var request_1 = require("../request");

var Authentication =
/*#__PURE__*/
function () {
  function Authentication(requestCtx) {
    _classCallCheck(this, Authentication);

    this.requestCtx = requestCtx;
  }

  _createClass(Authentication, [{
    key: "appPortalAccess",
    value: function appPortalAccess(appId, appPortalAccessIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/app-portal-access/{app_id}");
      request.setPathParam("app_id", appId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(appPortalAccessIn_1.AppPortalAccessInSerializer._toJsonObject(appPortalAccessIn));
      return request.send(this.requestCtx, appPortalAccessOut_1.AppPortalAccessOutSerializer._fromJsonObject);
    }
  }, {
    key: "expireAll",
    value: function expireAll(appId, applicationTokenExpireIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/app/{app_id}/expire-all");
      request.setPathParam("app_id", appId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(applicationTokenExpireIn_1.ApplicationTokenExpireInSerializer._toJsonObject(applicationTokenExpireIn));
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "dashboardAccess",
    value: function dashboardAccess(appId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/dashboard-access/{app_id}");
      request.setPathParam("app_id", appId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      return request.send(this.requestCtx, dashboardAccessOut_1.DashboardAccessOutSerializer._fromJsonObject);
    }
  }, {
    key: "logout",
    value: function logout(options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/logout");
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "streamLogout",
    value: function streamLogout(options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/stream-logout");
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "streamPortalAccess",
    value: function streamPortalAccess(streamId, streamPortalAccessIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/stream-portal-access/{stream_id}");
      request.setPathParam("stream_id", streamId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(streamPortalAccessIn_1.StreamPortalAccessInSerializer._toJsonObject(streamPortalAccessIn));
      return request.send(this.requestCtx, appPortalAccessOut_1.AppPortalAccessOutSerializer._fromJsonObject);
    }
  }, {
    key: "streamExpireAll",
    value: function streamExpireAll(streamId, streamTokenExpireIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/stream/{stream_id}/expire-all");
      request.setPathParam("stream_id", streamId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(streamTokenExpireIn_1.StreamTokenExpireInSerializer._toJsonObject(streamTokenExpireIn));
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "getStreamPollerToken",
    value: function getStreamPollerToken(streamId, sinkId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/auth/stream/{stream_id}/sink/{sink_id}/poller/token");
      request.setPathParam("stream_id", streamId);
      request.setPathParam("sink_id", sinkId);
      return request.send(this.requestCtx, apiTokenOut_1.ApiTokenOutSerializer._fromJsonObject);
    }
  }, {
    key: "rotateStreamPollerToken",
    value: function rotateStreamPollerToken(streamId, sinkId, rotatePollerTokenIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/stream/{stream_id}/sink/{sink_id}/poller/token/rotate");
      request.setPathParam("stream_id", streamId);
      request.setPathParam("sink_id", sinkId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(rotatePollerTokenIn_1.RotatePollerTokenInSerializer._toJsonObject(rotatePollerTokenIn));
      return request.send(this.requestCtx, apiTokenOut_1.ApiTokenOutSerializer._fromJsonObject);
    }
  }]);

  return Authentication;
}();

exports.Authentication = Authentication;
