// svix/dist/api/dist/backgroundTask.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BackgroundTask = void 0;

var backgroundTaskOut_1 = require("../models/backgroundTaskOut");

var listResponseBackgroundTaskOut_1 = require("../models/listResponseBackgroundTaskOut");

var request_1 = require("../request");

var BackgroundTask =
/*#__PURE__*/
function () {
  function BackgroundTask(requestCtx) {
    _classCallCheck(this, BackgroundTask);

    this.requestCtx = requestCtx;
  }

  _createClass(BackgroundTask, [{
    key: "list",
    value: function list(options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/background-task");
      request.setQueryParams({
        status: options === null || options === void 0 ? void 0 : options.status,
        task: options === null || options === void 0 ? void 0 : options.task,
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        order: options === null || options === void 0 ? void 0 : options.order
      });
      return request.send(this.requestCtx, listResponseBackgroundTaskOut_1.ListResponseBackgroundTaskOutSerializer._fromJsonObject);
    }
  }, {
    key: "listByEndpoint",
    value: function listByEndpoint(options) {
      return this.list(options);
    }
  }, {
    key: "get",
    value: function get(taskId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/background-task/{task_id}");
      request.setPathParam("task_id", taskId);
      return request.send(this.requestCtx, backgroundTaskOut_1.BackgroundTaskOutSerializer._fromJsonObject);
    }
  }]);

  return BackgroundTask;
}();

exports.BackgroundTask = BackgroundTask;
