// svix/dist/api/dist/connector.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Connector = void 0;

var connectorIn_1 = require("../models/connectorIn");

var connectorOut_1 = require("../models/connectorOut");

var connectorPatch_1 = require("../models/connectorPatch");

var connectorUpdate_1 = require("../models/connectorUpdate");

var listResponseConnectorOut_1 = require("../models/listResponseConnectorOut");

var request_1 = require("../request");

var Connector =
/*#__PURE__*/
function () {
  function Connector(requestCtx) {
    _classCallCheck(this, Connector);

    this.requestCtx = requestCtx;
  }

  _createClass(Connector, [{
    key: "list",
    value: function list(options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/connector");
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        order: options === null || options === void 0 ? void 0 : options.order,
        product_type: options === null || options === void 0 ? void 0 : options.productType
      });
      return request.send(this.requestCtx, listResponseConnectorOut_1.ListResponseConnectorOutSerializer._fromJsonObject);
    }
  }, {
    key: "create",
    value: function create(connectorIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/connector");
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(connectorIn_1.ConnectorInSerializer._toJsonObject(connectorIn));
      return request.send(this.requestCtx, connectorOut_1.ConnectorOutSerializer._fromJsonObject);
    }
  }, {
    key: "get",
    value: function get(connectorId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/connector/{connector_id}");
      request.setPathParam("connector_id", connectorId);
      return request.send(this.requestCtx, connectorOut_1.ConnectorOutSerializer._fromJsonObject);
    }
  }, {
    key: "update",
    value: function update(connectorId, connectorUpdate) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/connector/{connector_id}");
      request.setPathParam("connector_id", connectorId);
      request.setBody(connectorUpdate_1.ConnectorUpdateSerializer._toJsonObject(connectorUpdate));
      return request.send(this.requestCtx, connectorOut_1.ConnectorOutSerializer._fromJsonObject);
    }
  }, {
    key: "delete",
    value: function _delete(connectorId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/connector/{connector_id}");
      request.setPathParam("connector_id", connectorId);
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "patch",
    value: function patch(connectorId, connectorPatch) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/connector/{connector_id}");
      request.setPathParam("connector_id", connectorId);
      request.setBody(connectorPatch_1.ConnectorPatchSerializer._toJsonObject(connectorPatch));
      return request.send(this.requestCtx, connectorOut_1.ConnectorOutSerializer._fromJsonObject);
    }
  }]);

  return Connector;
}();

exports.Connector = Connector;
