// svix/dist/api/dist/endpoint.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Endpoint = void 0;

var bulkReplayIn_1 = require("../models/bulkReplayIn");

var endpointHeadersIn_1 = require("../models/endpointHeadersIn");

var endpointHeadersOut_1 = require("../models/endpointHeadersOut");

var endpointHeadersPatchIn_1 = require("../models/endpointHeadersPatchIn");

var endpointIn_1 = require("../models/endpointIn");

var endpointOut_1 = require("../models/endpointOut");

var endpointPatch_1 = require("../models/endpointPatch");

var endpointSecretOut_1 = require("../models/endpointSecretOut");

var endpointSecretRotateIn_1 = require("../models/endpointSecretRotateIn");

var endpointStats_1 = require("../models/endpointStats");

var endpointTransformationIn_1 = require("../models/endpointTransformationIn");

var endpointTransformationOut_1 = require("../models/endpointTransformationOut");

var endpointTransformationPatch_1 = require("../models/endpointTransformationPatch");

var endpointUpdate_1 = require("../models/endpointUpdate");

var eventExampleIn_1 = require("../models/eventExampleIn");

var listResponseEndpointOut_1 = require("../models/listResponseEndpointOut");

var messageOut_1 = require("../models/messageOut");

var recoverIn_1 = require("../models/recoverIn");

var recoverOut_1 = require("../models/recoverOut");

var replayIn_1 = require("../models/replayIn");

var replayOut_1 = require("../models/replayOut");

var request_1 = require("../request");

var Endpoint =
/*#__PURE__*/
function () {
  function Endpoint(requestCtx) {
    _classCallCheck(this, Endpoint);

    this.requestCtx = requestCtx;
  }

  _createClass(Endpoint, [{
    key: "list",
    value: function list(appId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint");
      request.setPathParam("app_id", appId);
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        order: options === null || options === void 0 ? void 0 : options.order
      });
      return request.send(this.requestCtx, listResponseEndpointOut_1.ListResponseEndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "create",
    value: function create(appId, endpointIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint");
      request.setPathParam("app_id", appId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(endpointIn_1.EndpointInSerializer._toJsonObject(endpointIn));
      return request.send(this.requestCtx, endpointOut_1.EndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "get",
    value: function get(appId, endpointId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      return request.send(this.requestCtx, endpointOut_1.EndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "update",
    value: function update(appId, endpointId, endpointUpdate) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/app/{app_id}/endpoint/{endpoint_id}");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      request.setBody(endpointUpdate_1.EndpointUpdateSerializer._toJsonObject(endpointUpdate));
      return request.send(this.requestCtx, endpointOut_1.EndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "delete",
    value: function _delete(appId, endpointId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/app/{app_id}/endpoint/{endpoint_id}");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "patch",
    value: function patch(appId, endpointId, endpointPatch) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/app/{app_id}/endpoint/{endpoint_id}");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      request.setBody(endpointPatch_1.EndpointPatchSerializer._toJsonObject(endpointPatch));
      return request.send(this.requestCtx, endpointOut_1.EndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "bulkReplay",
    value: function bulkReplay(appId, endpointId, bulkReplayIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/bulk-replay");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(bulkReplayIn_1.BulkReplayInSerializer._toJsonObject(bulkReplayIn));
      return request.send(this.requestCtx, replayOut_1.ReplayOutSerializer._fromJsonObject);
    }
  }, {
    key: "getHeaders",
    value: function getHeaders(appId, endpointId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      return request.send(this.requestCtx, endpointHeadersOut_1.EndpointHeadersOutSerializer._fromJsonObject);
    }
  }, {
    key: "updateHeaders",
    value: function updateHeaders(appId, endpointId, endpointHeadersIn) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      request.setBody(endpointHeadersIn_1.EndpointHeadersInSerializer._toJsonObject(endpointHeadersIn));
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "headersUpdate",
    value: function headersUpdate(appId, endpointId, endpointHeadersIn) {
      return this.updateHeaders(appId, endpointId, endpointHeadersIn);
    }
  }, {
    key: "patchHeaders",
    value: function patchHeaders(appId, endpointId, endpointHeadersPatchIn) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      request.setBody(endpointHeadersPatchIn_1.EndpointHeadersPatchInSerializer._toJsonObject(endpointHeadersPatchIn));
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "headersPatch",
    value: function headersPatch(appId, endpointId, endpointHeadersPatchIn) {
      return this.patchHeaders(appId, endpointId, endpointHeadersPatchIn);
    }
  }, {
    key: "recover",
    value: function recover(appId, endpointId, recoverIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/recover");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(recoverIn_1.RecoverInSerializer._toJsonObject(recoverIn));
      return request.send(this.requestCtx, recoverOut_1.RecoverOutSerializer._fromJsonObject);
    }
  }, {
    key: "replayMissing",
    value: function replayMissing(appId, endpointId, replayIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/replay-missing");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(replayIn_1.ReplayInSerializer._toJsonObject(replayIn));
      return request.send(this.requestCtx, replayOut_1.ReplayOutSerializer._fromJsonObject);
    }
  }, {
    key: "getSecret",
    value: function getSecret(appId, endpointId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/secret");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      return request.send(this.requestCtx, endpointSecretOut_1.EndpointSecretOutSerializer._fromJsonObject);
    }
  }, {
    key: "rotateSecret",
    value: function rotateSecret(appId, endpointId, endpointSecretRotateIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/secret/rotate");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(endpointSecretRotateIn_1.EndpointSecretRotateInSerializer._toJsonObject(endpointSecretRotateIn));
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "sendExample",
    value: function sendExample(appId, endpointId, eventExampleIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/send-example");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(eventExampleIn_1.EventExampleInSerializer._toJsonObject(eventExampleIn));
      return request.send(this.requestCtx, messageOut_1.MessageOutSerializer._fromJsonObject);
    }
  }, {
    key: "getStats",
    value: function getStats(appId, endpointId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/stats");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      request.setQueryParams({
        since: options === null || options === void 0 ? void 0 : options.since,
        until: options === null || options === void 0 ? void 0 : options.until
      });
      return request.send(this.requestCtx, endpointStats_1.EndpointStatsSerializer._fromJsonObject);
    }
  }, {
    key: "transformationGet",
    value: function transformationGet(appId, endpointId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/transformation");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      return request.send(this.requestCtx, endpointTransformationOut_1.EndpointTransformationOutSerializer._fromJsonObject);
    }
  }, {
    key: "patchTransformation",
    value: function patchTransformation(appId, endpointId, endpointTransformationPatch) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/transformation");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      request.setBody(endpointTransformationPatch_1.EndpointTransformationPatchSerializer._toJsonObject(endpointTransformationPatch));
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "transformationPartialUpdate",
    value: function transformationPartialUpdate(appId, endpointId, endpointTransformationIn) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/transformation");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      request.setBody(endpointTransformationIn_1.EndpointTransformationInSerializer._toJsonObject(endpointTransformationIn));
      return request.sendNoResponseBody(this.requestCtx);
    }
  }]);

  return Endpoint;
}();

exports.Endpoint = Endpoint;
