// svix/dist/api/dist/eventType.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EventType = void 0;

var eventTypeImportOpenApiIn_1 = require("../models/eventTypeImportOpenApiIn");

var eventTypeImportOpenApiOut_1 = require("../models/eventTypeImportOpenApiOut");

var eventTypeIn_1 = require("../models/eventTypeIn");

var eventTypeOut_1 = require("../models/eventTypeOut");

var eventTypePatch_1 = require("../models/eventTypePatch");

var eventTypeUpdate_1 = require("../models/eventTypeUpdate");

var listResponseEventTypeOut_1 = require("../models/listResponseEventTypeOut");

var request_1 = require("../request");

var EventType =
/*#__PURE__*/
function () {
  function EventType(requestCtx) {
    _classCallCheck(this, EventType);

    this.requestCtx = requestCtx;
  }

  _createClass(EventType, [{
    key: "list",
    value: function list(options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/event-type");
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        order: options === null || options === void 0 ? void 0 : options.order,
        include_archived: options === null || options === void 0 ? void 0 : options.includeArchived,
        with_content: options === null || options === void 0 ? void 0 : options.withContent
      });
      return request.send(this.requestCtx, listResponseEventTypeOut_1.ListResponseEventTypeOutSerializer._fromJsonObject);
    }
  }, {
    key: "create",
    value: function create(eventTypeIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/event-type");
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(eventTypeIn_1.EventTypeInSerializer._toJsonObject(eventTypeIn));
      return request.send(this.requestCtx, eventTypeOut_1.EventTypeOutSerializer._fromJsonObject);
    }
  }, {
    key: "importOpenapi",
    value: function importOpenapi(eventTypeImportOpenApiIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/event-type/import/openapi");
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(eventTypeImportOpenApiIn_1.EventTypeImportOpenApiInSerializer._toJsonObject(eventTypeImportOpenApiIn));
      return request.send(this.requestCtx, eventTypeImportOpenApiOut_1.EventTypeImportOpenApiOutSerializer._fromJsonObject);
    }
  }, {
    key: "get",
    value: function get(eventTypeName) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/event-type/{event_type_name}");
      request.setPathParam("event_type_name", eventTypeName);
      return request.send(this.requestCtx, eventTypeOut_1.EventTypeOutSerializer._fromJsonObject);
    }
  }, {
    key: "update",
    value: function update(eventTypeName, eventTypeUpdate) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/event-type/{event_type_name}");
      request.setPathParam("event_type_name", eventTypeName);
      request.setBody(eventTypeUpdate_1.EventTypeUpdateSerializer._toJsonObject(eventTypeUpdate));
      return request.send(this.requestCtx, eventTypeOut_1.EventTypeOutSerializer._fromJsonObject);
    }
  }, {
    key: "delete",
    value: function _delete(eventTypeName, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/event-type/{event_type_name}");
      request.setPathParam("event_type_name", eventTypeName);
      request.setQueryParams({
        expunge: options === null || options === void 0 ? void 0 : options.expunge
      });
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "patch",
    value: function patch(eventTypeName, eventTypePatch) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/event-type/{event_type_name}");
      request.setPathParam("event_type_name", eventTypeName);
      request.setBody(eventTypePatch_1.EventTypePatchSerializer._toJsonObject(eventTypePatch));
      return request.send(this.requestCtx, eventTypeOut_1.EventTypeOutSerializer._fromJsonObject);
    }
  }]);

  return EventType;
}();

exports.EventType = EventType;
