// svix/dist/api/dist/ingestEndpoint.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IngestEndpoint = void 0;

var ingestEndpointHeadersIn_1 = require("../models/ingestEndpointHeadersIn");

var ingestEndpointHeadersOut_1 = require("../models/ingestEndpointHeadersOut");

var ingestEndpointIn_1 = require("../models/ingestEndpointIn");

var ingestEndpointOut_1 = require("../models/ingestEndpointOut");

var ingestEndpointSecretIn_1 = require("../models/ingestEndpointSecretIn");

var ingestEndpointSecretOut_1 = require("../models/ingestEndpointSecretOut");

var ingestEndpointTransformationOut_1 = require("../models/ingestEndpointTransformationOut");

var ingestEndpointTransformationPatch_1 = require("../models/ingestEndpointTransformationPatch");

var ingestEndpointUpdate_1 = require("../models/ingestEndpointUpdate");

var listResponseIngestEndpointOut_1 = require("../models/listResponseIngestEndpointOut");

var request_1 = require("../request");

var IngestEndpoint =
/*#__PURE__*/
function () {
  function IngestEndpoint(requestCtx) {
    _classCallCheck(this, IngestEndpoint);

    this.requestCtx = requestCtx;
  }

  _createClass(IngestEndpoint, [{
    key: "list",
    value: function list(sourceId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint");
      request.setPathParam("source_id", sourceId);
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        order: options === null || options === void 0 ? void 0 : options.order
      });
      return request.send(this.requestCtx, listResponseIngestEndpointOut_1.ListResponseIngestEndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "create",
    value: function create(sourceId, ingestEndpointIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/ingest/api/v1/source/{source_id}/endpoint");
      request.setPathParam("source_id", sourceId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(ingestEndpointIn_1.IngestEndpointInSerializer._toJsonObject(ingestEndpointIn));
      return request.send(this.requestCtx, ingestEndpointOut_1.IngestEndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "get",
    value: function get(sourceId, endpointId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}");
      request.setPathParam("source_id", sourceId);
      request.setPathParam("endpoint_id", endpointId);
      return request.send(this.requestCtx, ingestEndpointOut_1.IngestEndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "update",
    value: function update(sourceId, endpointId, ingestEndpointUpdate) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}");
      request.setPathParam("source_id", sourceId);
      request.setPathParam("endpoint_id", endpointId);
      request.setBody(ingestEndpointUpdate_1.IngestEndpointUpdateSerializer._toJsonObject(ingestEndpointUpdate));
      return request.send(this.requestCtx, ingestEndpointOut_1.IngestEndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "delete",
    value: function _delete(sourceId, endpointId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}");
      request.setPathParam("source_id", sourceId);
      request.setPathParam("endpoint_id", endpointId);
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "getHeaders",
    value: function getHeaders(sourceId, endpointId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/headers");
      request.setPathParam("source_id", sourceId);
      request.setPathParam("endpoint_id", endpointId);
      return request.send(this.requestCtx, ingestEndpointHeadersOut_1.IngestEndpointHeadersOutSerializer._fromJsonObject);
    }
  }, {
    key: "updateHeaders",
    value: function updateHeaders(sourceId, endpointId, ingestEndpointHeadersIn) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/headers");
      request.setPathParam("source_id", sourceId);
      request.setPathParam("endpoint_id", endpointId);
      request.setBody(ingestEndpointHeadersIn_1.IngestEndpointHeadersInSerializer._toJsonObject(ingestEndpointHeadersIn));
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "getSecret",
    value: function getSecret(sourceId, endpointId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/secret");
      request.setPathParam("source_id", sourceId);
      request.setPathParam("endpoint_id", endpointId);
      return request.send(this.requestCtx, ingestEndpointSecretOut_1.IngestEndpointSecretOutSerializer._fromJsonObject);
    }
  }, {
    key: "rotateSecret",
    value: function rotateSecret(sourceId, endpointId, ingestEndpointSecretIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/secret/rotate");
      request.setPathParam("source_id", sourceId);
      request.setPathParam("endpoint_id", endpointId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(ingestEndpointSecretIn_1.IngestEndpointSecretInSerializer._toJsonObject(ingestEndpointSecretIn));
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "getTransformation",
    value: function getTransformation(sourceId, endpointId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/transformation");
      request.setPathParam("source_id", sourceId);
      request.setPathParam("endpoint_id", endpointId);
      return request.send(this.requestCtx, ingestEndpointTransformationOut_1.IngestEndpointTransformationOutSerializer._fromJsonObject);
    }
  }, {
    key: "setTransformation",
    value: function setTransformation(sourceId, endpointId, ingestEndpointTransformationPatch) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/transformation");
      request.setPathParam("source_id", sourceId);
      request.setPathParam("endpoint_id", endpointId);
      request.setBody(ingestEndpointTransformationPatch_1.IngestEndpointTransformationPatchSerializer._toJsonObject(ingestEndpointTransformationPatch));
      return request.sendNoResponseBody(this.requestCtx);
    }
  }]);

  return IngestEndpoint;
}();

exports.IngestEndpoint = IngestEndpoint;
