// svix/dist/api/dist/ingestSource.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IngestSource = void 0;

var ingestSourceIn_1 = require("../models/ingestSourceIn");

var ingestSourceOut_1 = require("../models/ingestSourceOut");

var listResponseIngestSourceOut_1 = require("../models/listResponseIngestSourceOut");

var rotateTokenOut_1 = require("../models/rotateTokenOut");

var request_1 = require("../request");

var IngestSource =
/*#__PURE__*/
function () {
  function IngestSource(requestCtx) {
    _classCallCheck(this, IngestSource);

    this.requestCtx = requestCtx;
  }

  _createClass(IngestSource, [{
    key: "list",
    value: function list(options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source");
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        order: options === null || options === void 0 ? void 0 : options.order
      });
      return request.send(this.requestCtx, listResponseIngestSourceOut_1.ListResponseIngestSourceOutSerializer._fromJsonObject);
    }
  }, {
    key: "create",
    value: function create(ingestSourceIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/ingest/api/v1/source");
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(ingestSourceIn_1.IngestSourceInSerializer._toJsonObject(ingestSourceIn));
      return request.send(this.requestCtx, ingestSourceOut_1.IngestSourceOutSerializer._fromJsonObject);
    }
  }, {
    key: "get",
    value: function get(sourceId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}");
      request.setPathParam("source_id", sourceId);
      return request.send(this.requestCtx, ingestSourceOut_1.IngestSourceOutSerializer._fromJsonObject);
    }
  }, {
    key: "update",
    value: function update(sourceId, ingestSourceIn) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/ingest/api/v1/source/{source_id}");
      request.setPathParam("source_id", sourceId);
      request.setBody(ingestSourceIn_1.IngestSourceInSerializer._toJsonObject(ingestSourceIn));
      return request.send(this.requestCtx, ingestSourceOut_1.IngestSourceOutSerializer._fromJsonObject);
    }
  }, {
    key: "delete",
    value: function _delete(sourceId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/ingest/api/v1/source/{source_id}");
      request.setPathParam("source_id", sourceId);
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "rotateToken",
    value: function rotateToken(sourceId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/ingest/api/v1/source/{source_id}/token/rotate");
      request.setPathParam("source_id", sourceId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      return request.send(this.requestCtx, rotateTokenOut_1.RotateTokenOutSerializer._fromJsonObject);
    }
  }]);

  return IngestSource;
}();

exports.IngestSource = IngestSource;
