// svix/dist/api/dist/integration.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Integration = void 0;

var integrationIn_1 = require("../models/integrationIn");

var integrationKeyOut_1 = require("../models/integrationKeyOut");

var integrationOut_1 = require("../models/integrationOut");

var integrationUpdate_1 = require("../models/integrationUpdate");

var listResponseIntegrationOut_1 = require("../models/listResponseIntegrationOut");

var request_1 = require("../request");

var Integration =
/*#__PURE__*/
function () {
  function Integration(requestCtx) {
    _classCallCheck(this, Integration);

    this.requestCtx = requestCtx;
  }

  _createClass(Integration, [{
    key: "list",
    value: function list(appId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/integration");
      request.setPathParam("app_id", appId);
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        order: options === null || options === void 0 ? void 0 : options.order
      });
      return request.send(this.requestCtx, listResponseIntegrationOut_1.ListResponseIntegrationOutSerializer._fromJsonObject);
    }
  }, {
    key: "create",
    value: function create(appId, integrationIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/integration");
      request.setPathParam("app_id", appId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(integrationIn_1.IntegrationInSerializer._toJsonObject(integrationIn));
      return request.send(this.requestCtx, integrationOut_1.IntegrationOutSerializer._fromJsonObject);
    }
  }, {
    key: "get",
    value: function get(appId, integId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/integration/{integ_id}");
      request.setPathParam("app_id", appId);
      request.setPathParam("integ_id", integId);
      return request.send(this.requestCtx, integrationOut_1.IntegrationOutSerializer._fromJsonObject);
    }
  }, {
    key: "update",
    value: function update(appId, integId, integrationUpdate) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/app/{app_id}/integration/{integ_id}");
      request.setPathParam("app_id", appId);
      request.setPathParam("integ_id", integId);
      request.setBody(integrationUpdate_1.IntegrationUpdateSerializer._toJsonObject(integrationUpdate));
      return request.send(this.requestCtx, integrationOut_1.IntegrationOutSerializer._fromJsonObject);
    }
  }, {
    key: "delete",
    value: function _delete(appId, integId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/app/{app_id}/integration/{integ_id}");
      request.setPathParam("app_id", appId);
      request.setPathParam("integ_id", integId);
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "getKey",
    value: function getKey(appId, integId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/integration/{integ_id}/key");
      request.setPathParam("app_id", appId);
      request.setPathParam("integ_id", integId);
      return request.send(this.requestCtx, integrationKeyOut_1.IntegrationKeyOutSerializer._fromJsonObject);
    }
  }, {
    key: "rotateKey",
    value: function rotateKey(appId, integId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/integration/{integ_id}/key/rotate");
      request.setPathParam("app_id", appId);
      request.setPathParam("integ_id", integId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      return request.send(this.requestCtx, integrationKeyOut_1.IntegrationKeyOutSerializer._fromJsonObject);
    }
  }]);

  return Integration;
}();

exports.Integration = Integration;
