// svix/dist/api/dist/message.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messageInRaw = exports.Message = void 0;

var expungeAllContentsOut_1 = require("../models/expungeAllContentsOut");

var listResponseMessageOut_1 = require("../models/listResponseMessageOut");

var messageOut_1 = require("../models/messageOut");

var messagePrecheckIn_1 = require("../models/messagePrecheckIn");

var messagePrecheckOut_1 = require("../models/messagePrecheckOut");

var messagePoller_1 = require("./messagePoller");

var request_1 = require("../request");

var messageIn_1 = require("../models/messageIn");

var Message =
/*#__PURE__*/
function () {
  function Message(requestCtx) {
    _classCallCheck(this, Message);

    this.requestCtx = requestCtx;
  }

  _createClass(Message, [{
    key: "list",
    value: function list(appId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/msg");
      request.setPathParam("app_id", appId);
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        channel: options === null || options === void 0 ? void 0 : options.channel,
        before: options === null || options === void 0 ? void 0 : options.before,
        after: options === null || options === void 0 ? void 0 : options.after,
        with_content: options === null || options === void 0 ? void 0 : options.withContent,
        tag: options === null || options === void 0 ? void 0 : options.tag,
        event_types: options === null || options === void 0 ? void 0 : options.eventTypes
      });
      return request.send(this.requestCtx, listResponseMessageOut_1.ListResponseMessageOutSerializer._fromJsonObject);
    }
  }, {
    key: "create",
    value: function create(appId, messageIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/msg");
      request.setPathParam("app_id", appId);
      request.setQueryParams({
        with_content: options === null || options === void 0 ? void 0 : options.withContent
      });
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(messageIn_1.MessageInSerializer._toJsonObject(messageIn));
      return request.send(this.requestCtx, messageOut_1.MessageOutSerializer._fromJsonObject);
    }
  }, {
    key: "expungeAllContents",
    value: function expungeAllContents(appId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/msg/expunge-all-contents");
      request.setPathParam("app_id", appId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      return request.send(this.requestCtx, expungeAllContentsOut_1.ExpungeAllContentsOutSerializer._fromJsonObject);
    }
  }, {
    key: "precheck",
    value: function precheck(appId, messagePrecheckIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/msg/precheck/active");
      request.setPathParam("app_id", appId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(messagePrecheckIn_1.MessagePrecheckInSerializer._toJsonObject(messagePrecheckIn));
      return request.send(this.requestCtx, messagePrecheckOut_1.MessagePrecheckOutSerializer._fromJsonObject);
    }
  }, {
    key: "get",
    value: function get(appId, msgId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/msg/{msg_id}");
      request.setPathParam("app_id", appId);
      request.setPathParam("msg_id", msgId);
      request.setQueryParams({
        with_content: options === null || options === void 0 ? void 0 : options.withContent
      });
      return request.send(this.requestCtx, messageOut_1.MessageOutSerializer._fromJsonObject);
    }
  }, {
    key: "expungeContent",
    value: function expungeContent(appId, msgId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/app/{app_id}/msg/{msg_id}/content");
      request.setPathParam("app_id", appId);
      request.setPathParam("msg_id", msgId);
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "poller",
    get: function get() {
      return new messagePoller_1.MessagePoller(this.requestCtx);
    }
  }]);

  return Message;
}();

exports.Message = Message;

function messageInRaw(eventType, payload, contentType) {
  var headers = contentType ? {
    "content-type": contentType
  } : undefined;
  return {
    eventType: eventType,
    payload: {},
    transformationsParams: {
      rawPayload: payload,
      headers: headers
    }
  };
}

exports.messageInRaw = messageInRaw;
