// svix/dist/api/dist/messageAttempt.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MessageAttempt = void 0;

var emptyResponse_1 = require("../models/emptyResponse");

var listResponseEndpointMessageOut_1 = require("../models/listResponseEndpointMessageOut");

var listResponseMessageAttemptOut_1 = require("../models/listResponseMessageAttemptOut");

var listResponseMessageEndpointOut_1 = require("../models/listResponseMessageEndpointOut");

var messageAttemptOut_1 = require("../models/messageAttemptOut");

var request_1 = require("../request");

var MessageAttempt =
/*#__PURE__*/
function () {
  function MessageAttempt(requestCtx) {
    _classCallCheck(this, MessageAttempt);

    this.requestCtx = requestCtx;
  }

  _createClass(MessageAttempt, [{
    key: "listByEndpoint",
    value: function listByEndpoint(appId, endpointId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/attempt/endpoint/{endpoint_id}");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        status: options === null || options === void 0 ? void 0 : options.status,
        status_code_class: options === null || options === void 0 ? void 0 : options.statusCodeClass,
        channel: options === null || options === void 0 ? void 0 : options.channel,
        tag: options === null || options === void 0 ? void 0 : options.tag,
        before: options === null || options === void 0 ? void 0 : options.before,
        after: options === null || options === void 0 ? void 0 : options.after,
        with_content: options === null || options === void 0 ? void 0 : options.withContent,
        with_msg: options === null || options === void 0 ? void 0 : options.withMsg,
        event_types: options === null || options === void 0 ? void 0 : options.eventTypes
      });
      return request.send(this.requestCtx, listResponseMessageAttemptOut_1.ListResponseMessageAttemptOutSerializer._fromJsonObject);
    }
  }, {
    key: "listByMsg",
    value: function listByMsg(appId, msgId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/attempt/msg/{msg_id}");
      request.setPathParam("app_id", appId);
      request.setPathParam("msg_id", msgId);
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        status: options === null || options === void 0 ? void 0 : options.status,
        status_code_class: options === null || options === void 0 ? void 0 : options.statusCodeClass,
        channel: options === null || options === void 0 ? void 0 : options.channel,
        tag: options === null || options === void 0 ? void 0 : options.tag,
        endpoint_id: options === null || options === void 0 ? void 0 : options.endpointId,
        before: options === null || options === void 0 ? void 0 : options.before,
        after: options === null || options === void 0 ? void 0 : options.after,
        with_content: options === null || options === void 0 ? void 0 : options.withContent,
        event_types: options === null || options === void 0 ? void 0 : options.eventTypes
      });
      return request.send(this.requestCtx, listResponseMessageAttemptOut_1.ListResponseMessageAttemptOutSerializer._fromJsonObject);
    }
  }, {
    key: "listAttemptedMessages",
    value: function listAttemptedMessages(appId, endpointId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/msg");
      request.setPathParam("app_id", appId);
      request.setPathParam("endpoint_id", endpointId);
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        channel: options === null || options === void 0 ? void 0 : options.channel,
        tag: options === null || options === void 0 ? void 0 : options.tag,
        status: options === null || options === void 0 ? void 0 : options.status,
        before: options === null || options === void 0 ? void 0 : options.before,
        after: options === null || options === void 0 ? void 0 : options.after,
        with_content: options === null || options === void 0 ? void 0 : options.withContent,
        event_types: options === null || options === void 0 ? void 0 : options.eventTypes
      });
      return request.send(this.requestCtx, listResponseEndpointMessageOut_1.ListResponseEndpointMessageOutSerializer._fromJsonObject);
    }
  }, {
    key: "get",
    value: function get(appId, msgId, attemptId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/msg/{msg_id}/attempt/{attempt_id}");
      request.setPathParam("app_id", appId);
      request.setPathParam("msg_id", msgId);
      request.setPathParam("attempt_id", attemptId);
      return request.send(this.requestCtx, messageAttemptOut_1.MessageAttemptOutSerializer._fromJsonObject);
    }
  }, {
    key: "expungeContent",
    value: function expungeContent(appId, msgId, attemptId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/app/{app_id}/msg/{msg_id}/attempt/{attempt_id}/content");
      request.setPathParam("app_id", appId);
      request.setPathParam("msg_id", msgId);
      request.setPathParam("attempt_id", attemptId);
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "listAttemptedDestinations",
    value: function listAttemptedDestinations(appId, msgId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/msg/{msg_id}/endpoint");
      request.setPathParam("app_id", appId);
      request.setPathParam("msg_id", msgId);
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator
      });
      return request.send(this.requestCtx, listResponseMessageEndpointOut_1.ListResponseMessageEndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "resend",
    value: function resend(appId, msgId, endpointId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/msg/{msg_id}/endpoint/{endpoint_id}/resend");
      request.setPathParam("app_id", appId);
      request.setPathParam("msg_id", msgId);
      request.setPathParam("endpoint_id", endpointId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      return request.send(this.requestCtx, emptyResponse_1.EmptyResponseSerializer._fromJsonObject);
    }
  }]);

  return MessageAttempt;
}();

exports.MessageAttempt = MessageAttempt;
