// svix/dist/api/dist/messagePoller.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MessagePoller = void 0;

var pollingEndpointConsumerSeekIn_1 = require("../models/pollingEndpointConsumerSeekIn");

var pollingEndpointConsumerSeekOut_1 = require("../models/pollingEndpointConsumerSeekOut");

var pollingEndpointOut_1 = require("../models/pollingEndpointOut");

var request_1 = require("../request");

var MessagePoller =
/*#__PURE__*/
function () {
  function MessagePoller(requestCtx) {
    _classCallCheck(this, MessagePoller);

    this.requestCtx = requestCtx;
  }

  _createClass(MessagePoller, [{
    key: "poll",
    value: function poll(appId, sinkId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/poller/{sink_id}");
      request.setPathParam("app_id", appId);
      request.setPathParam("sink_id", sinkId);
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        event_type: options === null || options === void 0 ? void 0 : options.eventType,
        channel: options === null || options === void 0 ? void 0 : options.channel,
        after: options === null || options === void 0 ? void 0 : options.after
      });
      return request.send(this.requestCtx, pollingEndpointOut_1.PollingEndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "consumerPoll",
    value: function consumerPoll(appId, sinkId, consumerId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/poller/{sink_id}/consumer/{consumer_id}");
      request.setPathParam("app_id", appId);
      request.setPathParam("sink_id", sinkId);
      request.setPathParam("consumer_id", consumerId);
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator
      });
      return request.send(this.requestCtx, pollingEndpointOut_1.PollingEndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "consumerSeek",
    value: function consumerSeek(appId, sinkId, consumerId, pollingEndpointConsumerSeekIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/poller/{sink_id}/consumer/{consumer_id}/seek");
      request.setPathParam("app_id", appId);
      request.setPathParam("sink_id", sinkId);
      request.setPathParam("consumer_id", consumerId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(pollingEndpointConsumerSeekIn_1.PollingEndpointConsumerSeekInSerializer._toJsonObject(pollingEndpointConsumerSeekIn));
      return request.send(this.requestCtx, pollingEndpointConsumerSeekOut_1.PollingEndpointConsumerSeekOutSerializer._fromJsonObject);
    }
  }]);

  return MessagePoller;
}();

exports.MessagePoller = MessagePoller;
