// svix/dist/api/dist/operationalWebhookEndpoint.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OperationalWebhookEndpoint = void 0;

var listResponseOperationalWebhookEndpointOut_1 = require("../models/listResponseOperationalWebhookEndpointOut");

var operationalWebhookEndpointHeadersIn_1 = require("../models/operationalWebhookEndpointHeadersIn");

var operationalWebhookEndpointHeadersOut_1 = require("../models/operationalWebhookEndpointHeadersOut");

var operationalWebhookEndpointIn_1 = require("../models/operationalWebhookEndpointIn");

var operationalWebhookEndpointOut_1 = require("../models/operationalWebhookEndpointOut");

var operationalWebhookEndpointSecretIn_1 = require("../models/operationalWebhookEndpointSecretIn");

var operationalWebhookEndpointSecretOut_1 = require("../models/operationalWebhookEndpointSecretOut");

var operationalWebhookEndpointUpdate_1 = require("../models/operationalWebhookEndpointUpdate");

var request_1 = require("../request");

var OperationalWebhookEndpoint =
/*#__PURE__*/
function () {
  function OperationalWebhookEndpoint(requestCtx) {
    _classCallCheck(this, OperationalWebhookEndpoint);

    this.requestCtx = requestCtx;
  }

  _createClass(OperationalWebhookEndpoint, [{
    key: "list",
    value: function list(options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/operational-webhook/endpoint");
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        order: options === null || options === void 0 ? void 0 : options.order
      });
      return request.send(this.requestCtx, listResponseOperationalWebhookEndpointOut_1.ListResponseOperationalWebhookEndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "create",
    value: function create(operationalWebhookEndpointIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/operational-webhook/endpoint");
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(operationalWebhookEndpointIn_1.OperationalWebhookEndpointInSerializer._toJsonObject(operationalWebhookEndpointIn));
      return request.send(this.requestCtx, operationalWebhookEndpointOut_1.OperationalWebhookEndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "get",
    value: function get(endpointId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/operational-webhook/endpoint/{endpoint_id}");
      request.setPathParam("endpoint_id", endpointId);
      return request.send(this.requestCtx, operationalWebhookEndpointOut_1.OperationalWebhookEndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "update",
    value: function update(endpointId, operationalWebhookEndpointUpdate) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/operational-webhook/endpoint/{endpoint_id}");
      request.setPathParam("endpoint_id", endpointId);
      request.setBody(operationalWebhookEndpointUpdate_1.OperationalWebhookEndpointUpdateSerializer._toJsonObject(operationalWebhookEndpointUpdate));
      return request.send(this.requestCtx, operationalWebhookEndpointOut_1.OperationalWebhookEndpointOutSerializer._fromJsonObject);
    }
  }, {
    key: "delete",
    value: function _delete(endpointId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/operational-webhook/endpoint/{endpoint_id}");
      request.setPathParam("endpoint_id", endpointId);
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "getHeaders",
    value: function getHeaders(endpointId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/operational-webhook/endpoint/{endpoint_id}/headers");
      request.setPathParam("endpoint_id", endpointId);
      return request.send(this.requestCtx, operationalWebhookEndpointHeadersOut_1.OperationalWebhookEndpointHeadersOutSerializer._fromJsonObject);
    }
  }, {
    key: "updateHeaders",
    value: function updateHeaders(endpointId, operationalWebhookEndpointHeadersIn) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/operational-webhook/endpoint/{endpoint_id}/headers");
      request.setPathParam("endpoint_id", endpointId);
      request.setBody(operationalWebhookEndpointHeadersIn_1.OperationalWebhookEndpointHeadersInSerializer._toJsonObject(operationalWebhookEndpointHeadersIn));
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "getSecret",
    value: function getSecret(endpointId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/operational-webhook/endpoint/{endpoint_id}/secret");
      request.setPathParam("endpoint_id", endpointId);
      return request.send(this.requestCtx, operationalWebhookEndpointSecretOut_1.OperationalWebhookEndpointSecretOutSerializer._fromJsonObject);
    }
  }, {
    key: "rotateSecret",
    value: function rotateSecret(endpointId, operationalWebhookEndpointSecretIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/operational-webhook/endpoint/{endpoint_id}/secret/rotate");
      request.setPathParam("endpoint_id", endpointId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(operationalWebhookEndpointSecretIn_1.OperationalWebhookEndpointSecretInSerializer._toJsonObject(operationalWebhookEndpointSecretIn));
      return request.sendNoResponseBody(this.requestCtx);
    }
  }]);

  return OperationalWebhookEndpoint;
}();

exports.OperationalWebhookEndpoint = OperationalWebhookEndpoint;
