// svix/dist/api/dist/streamingEvents.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StreamingEvents = void 0;

var createStreamEventsIn_1 = require("../models/createStreamEventsIn");

var createStreamEventsOut_1 = require("../models/createStreamEventsOut");

var eventStreamOut_1 = require("../models/eventStreamOut");

var request_1 = require("../request");

var StreamingEvents =
/*#__PURE__*/
function () {
  function StreamingEvents(requestCtx) {
    _classCallCheck(this, StreamingEvents);

    this.requestCtx = requestCtx;
  }

  _createClass(StreamingEvents, [{
    key: "create",
    value: function create(streamId, createStreamEventsIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/stream/{stream_id}/events");
      request.setPathParam("stream_id", streamId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(createStreamEventsIn_1.CreateStreamEventsInSerializer._toJsonObject(createStreamEventsIn));
      return request.send(this.requestCtx, createStreamEventsOut_1.CreateStreamEventsOutSerializer._fromJsonObject);
    }
  }, {
    key: "get",
    value: function get(streamId, sinkId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}/sink/{sink_id}/events");
      request.setPathParam("stream_id", streamId);
      request.setPathParam("sink_id", sinkId);
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        after: options === null || options === void 0 ? void 0 : options.after
      });
      return request.send(this.requestCtx, eventStreamOut_1.EventStreamOutSerializer._fromJsonObject);
    }
  }]);

  return StreamingEvents;
}();

exports.StreamingEvents = StreamingEvents;
