// svix/dist/api/dist/streamingEventType.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StreamingEventType = void 0;

var listResponseStreamEventTypeOut_1 = require("../models/listResponseStreamEventTypeOut");

var streamEventTypeIn_1 = require("../models/streamEventTypeIn");

var streamEventTypeOut_1 = require("../models/streamEventTypeOut");

var streamEventTypePatch_1 = require("../models/streamEventTypePatch");

var request_1 = require("../request");

var StreamingEventType =
/*#__PURE__*/
function () {
  function StreamingEventType(requestCtx) {
    _classCallCheck(this, StreamingEventType);

    this.requestCtx = requestCtx;
  }

  _createClass(StreamingEventType, [{
    key: "list",
    value: function list(options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/event-type");
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        order: options === null || options === void 0 ? void 0 : options.order,
        include_archived: options === null || options === void 0 ? void 0 : options.includeArchived
      });
      return request.send(this.requestCtx, listResponseStreamEventTypeOut_1.ListResponseStreamEventTypeOutSerializer._fromJsonObject);
    }
  }, {
    key: "create",
    value: function create(streamEventTypeIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/stream/event-type");
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(streamEventTypeIn_1.StreamEventTypeInSerializer._toJsonObject(streamEventTypeIn));
      return request.send(this.requestCtx, streamEventTypeOut_1.StreamEventTypeOutSerializer._fromJsonObject);
    }
  }, {
    key: "get",
    value: function get(name) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/event-type/{name}");
      request.setPathParam("name", name);
      return request.send(this.requestCtx, streamEventTypeOut_1.StreamEventTypeOutSerializer._fromJsonObject);
    }
  }, {
    key: "update",
    value: function update(name, streamEventTypeIn) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/stream/event-type/{name}");
      request.setPathParam("name", name);
      request.setBody(streamEventTypeIn_1.StreamEventTypeInSerializer._toJsonObject(streamEventTypeIn));
      return request.send(this.requestCtx, streamEventTypeOut_1.StreamEventTypeOutSerializer._fromJsonObject);
    }
  }, {
    key: "delete",
    value: function _delete(name, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/stream/event-type/{name}");
      request.setPathParam("name", name);
      request.setQueryParams({
        expunge: options === null || options === void 0 ? void 0 : options.expunge
      });
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "patch",
    value: function patch(name, streamEventTypePatch) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/stream/event-type/{name}");
      request.setPathParam("name", name);
      request.setBody(streamEventTypePatch_1.StreamEventTypePatchSerializer._toJsonObject(streamEventTypePatch));
      return request.send(this.requestCtx, streamEventTypeOut_1.StreamEventTypeOutSerializer._fromJsonObject);
    }
  }]);

  return StreamingEventType;
}();

exports.StreamingEventType = StreamingEventType;
