// svix/dist/api/dist/streamingSink.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StreamingSink = void 0;

var emptyResponse_1 = require("../models/emptyResponse");

var endpointSecretRotateIn_1 = require("../models/endpointSecretRotateIn");

var listResponseStreamSinkOut_1 = require("../models/listResponseStreamSinkOut");

var sinkSecretOut_1 = require("../models/sinkSecretOut");

var sinkTransformIn_1 = require("../models/sinkTransformIn");

var streamSinkIn_1 = require("../models/streamSinkIn");

var streamSinkOut_1 = require("../models/streamSinkOut");

var streamSinkPatch_1 = require("../models/streamSinkPatch");

var request_1 = require("../request");

var StreamingSink =
/*#__PURE__*/
function () {
  function StreamingSink(requestCtx) {
    _classCallCheck(this, StreamingSink);

    this.requestCtx = requestCtx;
  }

  _createClass(StreamingSink, [{
    key: "list",
    value: function list(streamId, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}/sink");
      request.setPathParam("stream_id", streamId);
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        order: options === null || options === void 0 ? void 0 : options.order
      });
      return request.send(this.requestCtx, listResponseStreamSinkOut_1.ListResponseStreamSinkOutSerializer._fromJsonObject);
    }
  }, {
    key: "create",
    value: function create(streamId, streamSinkIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/stream/{stream_id}/sink");
      request.setPathParam("stream_id", streamId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(streamSinkIn_1.StreamSinkInSerializer._toJsonObject(streamSinkIn));
      return request.send(this.requestCtx, streamSinkOut_1.StreamSinkOutSerializer._fromJsonObject);
    }
  }, {
    key: "get",
    value: function get(streamId, sinkId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}/sink/{sink_id}");
      request.setPathParam("stream_id", streamId);
      request.setPathParam("sink_id", sinkId);
      return request.send(this.requestCtx, streamSinkOut_1.StreamSinkOutSerializer._fromJsonObject);
    }
  }, {
    key: "update",
    value: function update(streamId, sinkId, streamSinkIn) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/stream/{stream_id}/sink/{sink_id}");
      request.setPathParam("stream_id", streamId);
      request.setPathParam("sink_id", sinkId);
      request.setBody(streamSinkIn_1.StreamSinkInSerializer._toJsonObject(streamSinkIn));
      return request.send(this.requestCtx, streamSinkOut_1.StreamSinkOutSerializer._fromJsonObject);
    }
  }, {
    key: "delete",
    value: function _delete(streamId, sinkId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/stream/{stream_id}/sink/{sink_id}");
      request.setPathParam("stream_id", streamId);
      request.setPathParam("sink_id", sinkId);
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "patch",
    value: function patch(streamId, sinkId, streamSinkPatch) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/stream/{stream_id}/sink/{sink_id}");
      request.setPathParam("stream_id", streamId);
      request.setPathParam("sink_id", sinkId);
      request.setBody(streamSinkPatch_1.StreamSinkPatchSerializer._toJsonObject(streamSinkPatch));
      return request.send(this.requestCtx, streamSinkOut_1.StreamSinkOutSerializer._fromJsonObject);
    }
  }, {
    key: "getSecret",
    value: function getSecret(streamId, sinkId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}/sink/{sink_id}/secret");
      request.setPathParam("stream_id", streamId);
      request.setPathParam("sink_id", sinkId);
      return request.send(this.requestCtx, sinkSecretOut_1.SinkSecretOutSerializer._fromJsonObject);
    }
  }, {
    key: "rotateSecret",
    value: function rotateSecret(streamId, sinkId, endpointSecretRotateIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/stream/{stream_id}/sink/{sink_id}/secret/rotate");
      request.setPathParam("stream_id", streamId);
      request.setPathParam("sink_id", sinkId);
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(endpointSecretRotateIn_1.EndpointSecretRotateInSerializer._toJsonObject(endpointSecretRotateIn));
      return request.send(this.requestCtx, emptyResponse_1.EmptyResponseSerializer._fromJsonObject);
    }
  }, {
    key: "transformationPartialUpdate",
    value: function transformationPartialUpdate(streamId, sinkId, sinkTransformIn) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/stream/{stream_id}/sink/{sink_id}/transformation");
      request.setPathParam("stream_id", streamId);
      request.setPathParam("sink_id", sinkId);
      request.setBody(sinkTransformIn_1.SinkTransformInSerializer._toJsonObject(sinkTransformIn));
      return request.send(this.requestCtx, emptyResponse_1.EmptyResponseSerializer._fromJsonObject);
    }
  }]);

  return StreamingSink;
}();

exports.StreamingSink = StreamingSink;
