// svix/dist/api/dist/streamingStream.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StreamingStream = void 0;

var listResponseStreamOut_1 = require("../models/listResponseStreamOut");

var streamIn_1 = require("../models/streamIn");

var streamOut_1 = require("../models/streamOut");

var streamPatch_1 = require("../models/streamPatch");

var request_1 = require("../request");

var StreamingStream =
/*#__PURE__*/
function () {
  function StreamingStream(requestCtx) {
    _classCallCheck(this, StreamingStream);

    this.requestCtx = requestCtx;
  }

  _createClass(StreamingStream, [{
    key: "list",
    value: function list(options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream");
      request.setQueryParams({
        limit: options === null || options === void 0 ? void 0 : options.limit,
        iterator: options === null || options === void 0 ? void 0 : options.iterator,
        order: options === null || options === void 0 ? void 0 : options.order
      });
      return request.send(this.requestCtx, listResponseStreamOut_1.ListResponseStreamOutSerializer._fromJsonObject);
    }
  }, {
    key: "create",
    value: function create(streamIn, options) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/stream");
      request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
      request.setBody(streamIn_1.StreamInSerializer._toJsonObject(streamIn));
      return request.send(this.requestCtx, streamOut_1.StreamOutSerializer._fromJsonObject);
    }
  }, {
    key: "get",
    value: function get(streamId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}");
      request.setPathParam("stream_id", streamId);
      return request.send(this.requestCtx, streamOut_1.StreamOutSerializer._fromJsonObject);
    }
  }, {
    key: "update",
    value: function update(streamId, streamIn) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/stream/{stream_id}");
      request.setPathParam("stream_id", streamId);
      request.setBody(streamIn_1.StreamInSerializer._toJsonObject(streamIn));
      return request.send(this.requestCtx, streamOut_1.StreamOutSerializer._fromJsonObject);
    }
  }, {
    key: "delete",
    value: function _delete(streamId) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/stream/{stream_id}");
      request.setPathParam("stream_id", streamId);
      return request.sendNoResponseBody(this.requestCtx);
    }
  }, {
    key: "patch",
    value: function patch(streamId, streamPatch) {
      var request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/stream/{stream_id}");
      request.setPathParam("stream_id", streamId);
      request.setBody(streamPatch_1.StreamPatchSerializer._toJsonObject(streamPatch));
      return request.send(this.requestCtx, streamOut_1.StreamOutSerializer._fromJsonObject);
    }
  }]);

  return StreamingStream;
}();

exports.StreamingStream = StreamingStream;
