// svix/src/models/dist/appPortalAccessIn.js
"use strict";
exports.__esModule = true;
exports.AppPortalAccessInSerializer = void 0;
// this file is @generated
var appPortalCapability_1 = require("./appPortalCapability");
var applicationIn_1 = require("./applicationIn");
exports.AppPortalAccessInSerializer = {
    _fromJsonObject: function (object) {
        var _a;
        return {
            application: object["application"] != null
                ? applicationIn_1.ApplicationInSerializer._fromJsonObject(object["application"])
                : undefined,
            capabilities: (_a = object["capabilities"]) === null || _a === void 0 ? void 0 : _a.map(function (item) {
                return appPortalCapability_1.AppPortalCapabilitySerializer._fromJsonObject(item);
            }),
            expiry: object["expiry"],
            featureFlags: object["featureFlags"],
            readOnly: object["readOnly"],
            sessionId: object["sessionId"]
        };
    },
    _toJsonObject: function (self) {
        var _a;
        return {
            application: self.application != null
                ? applicationIn_1.ApplicationInSerializer._toJsonObject(self.application)
                : undefined,
            capabilities: (_a = self.capabilities) === null || _a === void 0 ? void 0 : _a.map(function (item) {
                return appPortalCapability_1.AppPortalCapabilitySerializer._toJsonObject(item);
            }),
            expiry: self.expiry,
            featureFlags: self.featureFlags,
            readOnly: self.readOnly,
            sessionId: self.sessionId
        };
    }
};
