// svix/src/models/dist/appUsageStatsOut.js
"use strict";
exports.__esModule = true;
exports.AppUsageStatsOutSerializer = void 0;
// this file is @generated
var backgroundTaskStatus_1 = require("./backgroundTaskStatus");
var backgroundTaskType_1 = require("./backgroundTaskType");
exports.AppUsageStatsOutSerializer = {
    _fromJsonObject: function (object) {
        return {
            id: object["id"],
            status: backgroundTaskStatus_1.BackgroundTaskStatusSerializer._fromJsonObject(object["status"]),
            task: backgroundTaskType_1.BackgroundTaskTypeSerializer._fromJsonObject(object["task"]),
            unresolvedAppIds: object["unresolvedAppIds"],
            updatedAt: new Date(object["updatedAt"])
        };
    },
    _toJsonObject: function (self) {
        return {
            id: self.id,
            status: backgroundTaskStatus_1.BackgroundTaskStatusSerializer._toJsonObject(self.status),
            task: backgroundTaskType_1.BackgroundTaskTypeSerializer._toJsonObject(self.task),
            unresolvedAppIds: self.unresolvedAppIds,
            updatedAt: self.updatedAt
        };
    }
};
