// svix/src/models/dist/bulkReplayIn.js
"use strict";
exports.__esModule = true;
exports.BulkReplayInSerializer = void 0;
// this file is @generated
var messageStatus_1 = require("./messageStatus");
var statusCodeClass_1 = require("./statusCodeClass");
exports.BulkReplayInSerializer = {
    _fromJsonObject: function (object) {
        return {
            channel: object["channel"],
            eventTypes: object["eventTypes"],
            since: new Date(object["since"]),
            status: object["status"] != null
                ? messageStatus_1.MessageStatusSerializer._fromJsonObject(object["status"])
                : undefined,
            statusCodeClass: object["statusCodeClass"] != null
                ? statusCodeClass_1.StatusCodeClassSerializer._fromJsonObject(object["statusCodeClass"])
                : undefined,
            tag: object["tag"],
            until: object["until"] ? new Date(object["until"]) : null
        };
    },
    _toJsonObject: function (self) {
        return {
            channel: self.channel,
            eventTypes: self.eventTypes,
            since: self.since,
            status: self.status != null
                ? messageStatus_1.MessageStatusSerializer._toJsonObject(self.status)
                : undefined,
            statusCodeClass: self.statusCodeClass != null
                ? statusCodeClass_1.StatusCodeClassSerializer._toJsonObject(self.statusCodeClass)
                : undefined,
            tag: self.tag,
            until: self.until
        };
    }
};
