// svix/src/models/dist/endpointDisabledEventData.js
"use strict";
exports.__esModule = true;
exports.EndpointDisabledEventDataSerializer = void 0;
// this file is @generated
var endpointDisabledTrigger_1 = require("./endpointDisabledTrigger");
exports.EndpointDisabledEventDataSerializer = {
    _fromJsonObject: function (object) {
        return {
            appId: object["appId"],
            appUid: object["appUid"],
            endpointId: object["endpointId"],
            endpointUid: object["endpointUid"],
            failSince: object["failSince"] ? new Date(object["failSince"]) : null,
            trigger: object["trigger"] != null
                ? endpointDisabledTrigger_1.EndpointDisabledTriggerSerializer._fromJsonObject(object["trigger"])
                : undefined
        };
    },
    _toJsonObject: function (self) {
        return {
            appId: self.appId,
            appUid: self.appUid,
            endpointId: self.endpointId,
            endpointUid: self.endpointUid,
            failSince: self.failSince,
            trigger: self.trigger != null
                ? endpointDisabledTrigger_1.EndpointDisabledTriggerSerializer._toJsonObject(self.trigger)
                : undefined
        };
    }
};
