// svix/src/models/dist/environmentOut.js
"use strict";
exports.__esModule = true;
exports.EnvironmentOutSerializer = void 0;
// this file is @generated
var connectorOut_1 = require("./connectorOut");
var eventTypeOut_1 = require("./eventTypeOut");
exports.EnvironmentOutSerializer = {
    _fromJsonObject: function (object) {
        return {
            connectors: object["connectors"].map(function (item) {
                return connectorOut_1.ConnectorOutSerializer._fromJsonObject(item);
            }),
            createdAt: new Date(object["createdAt"]),
            eventTypes: object["eventTypes"].map(function (item) {
                return eventTypeOut_1.EventTypeOutSerializer._fromJsonObject(item);
            }),
            settings: object["settings"],
            version: object["version"]
        };
    },
    _toJsonObject: function (self) {
        return {
            connectors: self.connectors.map(function (item) {
                return connectorOut_1.ConnectorOutSerializer._toJsonObject(item);
            }),
            createdAt: self.createdAt,
            eventTypes: self.eventTypes.map(function (item) {
                return eventTypeOut_1.EventTypeOutSerializer._toJsonObject(item);
            }),
            settings: self.settings,
            version: self.version
        };
    }
};
