// svix/src/models/dist/ingestEndpointDisabledEventData.js
"use strict";
exports.__esModule = true;
exports.IngestEndpointDisabledEventDataSerializer = void 0;
// this file is @generated
var endpointDisabledTrigger_1 = require("./endpointDisabledTrigger");
exports.IngestEndpointDisabledEventDataSerializer = {
    _fromJsonObject: function (object) {
        return {
            endpointId: object["endpointId"],
            endpointUid: object["endpointUid"],
            failSince: object["failSince"] ? new Date(object["failSince"]) : null,
            sourceId: object["sourceId"],
            trigger: object["trigger"] != null
                ? endpointDisabledTrigger_1.EndpointDisabledTriggerSerializer._fromJsonObject(object["trigger"])
                : undefined
        };
    },
    _toJsonObject: function (self) {
        return {
            endpointId: self.endpointId,
            endpointUid: self.endpointUid,
            failSince: self.failSince,
            sourceId: self.sourceId,
            trigger: self.trigger != null
                ? endpointDisabledTrigger_1.EndpointDisabledTriggerSerializer._toJsonObject(self.trigger)
                : undefined
        };
    }
};
