// svix/src/models/dist/messageAttemptExhaustedEventData.js
"use strict";
exports.__esModule = true;
exports.MessageAttemptExhaustedEventDataSerializer = void 0;
// this file is @generated
var messageAttemptFailedData_1 = require("./messageAttemptFailedData");
exports.MessageAttemptExhaustedEventDataSerializer = {
    _fromJsonObject: function (object) {
        return {
            appId: object["appId"],
            appUid: object["appUid"],
            endpointId: object["endpointId"],
            lastAttempt: messageAttemptFailedData_1.MessageAttemptFailedDataSerializer._fromJsonObject(object["lastAttempt"]),
            msgEventId: object["msgEventId"],
            msgId: object["msgId"]
        };
    },
    _toJsonObject: function (self) {
        return {
            appId: self.appId,
            appUid: self.appUid,
            endpointId: self.endpointId,
            lastAttempt: messageAttemptFailedData_1.MessageAttemptFailedDataSerializer._toJsonObject(self.lastAttempt),
            msgEventId: self.msgEventId,
            msgId: self.msgId
        };
    }
};
