// svix/src/models/dist/messageAttemptOut.js
"use strict";
exports.__esModule = true;
exports.MessageAttemptOutSerializer = void 0;
// this file is @generated
var messageAttemptTriggerType_1 = require("./messageAttemptTriggerType");
var messageOut_1 = require("./messageOut");
var messageStatus_1 = require("./messageStatus");
var messageStatusText_1 = require("./messageStatusText");
exports.MessageAttemptOutSerializer = {
    _fromJsonObject: function (object) {
        return {
            endpointId: object["endpointId"],
            id: object["id"],
            msg: object["msg"] != null
                ? messageOut_1.MessageOutSerializer._fromJsonObject(object["msg"])
                : undefined,
            msgId: object["msgId"],
            response: object["response"],
            responseDurationMs: object["responseDurationMs"],
            responseStatusCode: object["responseStatusCode"],
            status: messageStatus_1.MessageStatusSerializer._fromJsonObject(object["status"]),
            statusText: messageStatusText_1.MessageStatusTextSerializer._fromJsonObject(object["statusText"]),
            timestamp: new Date(object["timestamp"]),
            triggerType: messageAttemptTriggerType_1.MessageAttemptTriggerTypeSerializer._fromJsonObject(object["triggerType"]),
            url: object["url"]
        };
    },
    _toJsonObject: function (self) {
        return {
            endpointId: self.endpointId,
            id: self.id,
            msg: self.msg != null ? messageOut_1.MessageOutSerializer._toJsonObject(self.msg) : undefined,
            msgId: self.msgId,
            response: self.response,
            responseDurationMs: self.responseDurationMs,
            responseStatusCode: self.responseStatusCode,
            status: messageStatus_1.MessageStatusSerializer._toJsonObject(self.status),
            statusText: messageStatusText_1.MessageStatusTextSerializer._toJsonObject(self.statusText),
            timestamp: self.timestamp,
            triggerType: messageAttemptTriggerType_1.MessageAttemptTriggerTypeSerializer._toJsonObject(self.triggerType),
            url: self.url
        };
    }
};
