// svix/src/models/dist/statusCodeClass.js
"use strict";
exports.__esModule = true;
exports.StatusCodeClassSerializer = exports.StatusCodeClass = void 0;
// this file is @generated
/**
 * The different classes of HTTP status codes:
 *
 * - CodeNone = 0
 * - Code1xx = 100
 * - Code2xx = 200
 * - Code3xx = 300
 * - Code4xx = 400
 * - Code5xx = 500
 */
var StatusCodeClass;
(function (StatusCodeClass) {
    StatusCodeClass[StatusCodeClass["CodeNone"] = 0] = "CodeNone";
    StatusCodeClass[StatusCodeClass["Code1xx"] = 100] = "Code1xx";
    StatusCodeClass[StatusCodeClass["Code2xx"] = 200] = "Code2xx";
    StatusCodeClass[StatusCodeClass["Code3xx"] = 300] = "Code3xx";
    StatusCodeClass[StatusCodeClass["Code4xx"] = 400] = "Code4xx";
    StatusCodeClass[StatusCodeClass["Code5xx"] = 500] = "Code5xx";
})(StatusCodeClass = exports.StatusCodeClass || (exports.StatusCodeClass = {}));
exports.StatusCodeClassSerializer = {
    _fromJsonObject: function (object) {
        return object;
    },
    _toJsonObject: function (self) {
        return self;
    }
};
