// postal-mime/src/dist/base64-decoder.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _decodeStrings = require("./decode-strings.js");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var Base64Decoder =
/*#__PURE__*/
function () {
  function Base64Decoder(opts) {
    _classCallCheck(this, Base64Decoder);

    opts = opts || {};
    this.decoder = opts.decoder || new TextDecoder();
    this.maxChunkSize = 100 * 1024;
    this.chunks = [];
    this.remainder = '';
  }

  _createClass(Base64Decoder, [{
    key: "update",
    value: function update(buffer) {
      var str = this.decoder.decode(buffer);
      str = str.replace(/[^a-zA-Z0-9+\/]+/g, '');
      this.remainder += str;

      if (this.remainder.length >= this.maxChunkSize) {
        var allowedBytes = Math.floor(this.remainder.length / 4) * 4;
        var base64Str;

        if (allowedBytes === this.remainder.length) {
          base64Str = this.remainder;
          this.remainder = '';
        } else {
          base64Str = this.remainder.substr(0, allowedBytes);
          this.remainder = this.remainder.substr(allowedBytes);
        }

        if (base64Str.length) {
          this.chunks.push((0, _decodeStrings.decodeBase64)(base64Str));
        }
      }
    }
  }, {
    key: "finalize",
    value: function finalize() {
      if (this.remainder && !/^=+$/.test(this.remainder)) {
        this.chunks.push((0, _decodeStrings.decodeBase64)(this.remainder));
      }

      return (0, _decodeStrings.blobToArrayBuffer)(new Blob(this.chunks, {
        type: 'application/octet-stream'
      }));
    }
  }]);

  return Base64Decoder;
}();

exports["default"] = Base64Decoder;
