// postal-mime/src/dist/decode-strings.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.decodeBase64 = decodeBase64;
exports.getDecoder = getDecoder;
exports.blobToArrayBuffer = blobToArrayBuffer;
exports.getHex = getHex;
exports.decodeWord = decodeWord;
exports.decodeWords = decodeWords;
exports.decodeURIComponentWithCharset = decodeURIComponentWithCharset;
exports.decodeParameterValueContinuations = decodeParameterValueContinuations;
exports.textEncoder = void 0;
var textEncoder = new TextEncoder();
exports.textEncoder = textEncoder;
var base64Chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/'; // Use a lookup table to find the index.

var base64Lookup = new Uint8Array(256);

for (var i = 0; i < base64Chars.length; i++) {
  base64Lookup[base64Chars.charCodeAt(i)] = i;
}

function decodeBase64(base64) {
  var bufferLength = Math.ceil(base64.length / 4) * 3;
  var len = base64.length;
  var p = 0;

  if (base64.length % 4 === 3) {
    bufferLength--;
  } else if (base64.length % 4 === 2) {
    bufferLength -= 2;
  } else if (base64[base64.length - 1] === '=') {
    bufferLength--;

    if (base64[base64.length - 2] === '=') {
      bufferLength--;
    }
  }

  var arrayBuffer = new ArrayBuffer(bufferLength);
  var bytes = new Uint8Array(arrayBuffer);

  for (var _i = 0; _i < len; _i += 4) {
    var encoded1 = base64Lookup[base64.charCodeAt(_i)];
    var encoded2 = base64Lookup[base64.charCodeAt(_i + 1)];
    var encoded3 = base64Lookup[base64.charCodeAt(_i + 2)];
    var encoded4 = base64Lookup[base64.charCodeAt(_i + 3)];
    bytes[p++] = encoded1 << 2 | encoded2 >> 4;
    bytes[p++] = (encoded2 & 15) << 4 | encoded3 >> 2;
    bytes[p++] = (encoded3 & 3) << 6 | encoded4 & 63;
  }

  return arrayBuffer;
}

function getDecoder(charset) {
  charset = charset || 'utf8';
  var decoder;

  try {
    decoder = new TextDecoder(charset);
  } catch (err) {
    decoder = new TextDecoder('windows-1252');
  }

  return decoder;
}
/**
 * Converts a Blob into an ArrayBuffer
 * @param {Blob} blob Blob to convert
 * @returns {ArrayBuffer} Converted value
 */


function blobToArrayBuffer(blob) {
  var fr;
  return regeneratorRuntime.async(function blobToArrayBuffer$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          if (!('arrayBuffer' in blob)) {
            _context.next = 4;
            break;
          }

          _context.next = 3;
          return regeneratorRuntime.awrap(blob.arrayBuffer());

        case 3:
          return _context.abrupt("return", _context.sent);

        case 4:
          fr = new FileReader();
          return _context.abrupt("return", new Promise(function (resolve, reject) {
            fr.onload = function (e) {
              resolve(e.target.result);
            };

            fr.onerror = function (e) {
              reject(fr.error);
            };

            fr.readAsArrayBuffer(blob);
          }));

        case 6:
        case "end":
          return _context.stop();
      }
    }
  });
}

function getHex(c) {
  if (c >= 0x30
  /* 0 */
  && c <= 0x39 ||
  /* 9 */
  c >= 0x61
  /* a */
  && c <= 0x66
  /* f */
  || c >= 0x41
  /* A */
  && c <= 0x46
  /* F */
  ) {
      return String.fromCharCode(c);
    }

  return false;
}
/**
 * Decode a complete mime word encoded string
 *
 * @param {String} str Mime word encoded string
 * @return {String} Decoded unicode string
 */


function decodeWord(charset, encoding, str) {
  // RFC2231 added language tag to the encoding
  // see: https://tools.ietf.org/html/rfc2231#section-5
  // this implementation silently ignores this tag
  var splitPos = charset.indexOf('*');

  if (splitPos >= 0) {
    charset = charset.substr(0, splitPos);
  }

  encoding = encoding.toUpperCase();
  var byteStr;

  if (encoding === 'Q') {
    str = str // remove spaces between = and hex char, this might indicate invalidly applied line splitting
    .replace(/=\s+([0-9a-fA-F])/g, '=$1') // convert all underscores to spaces
    .replace(/[_\s]/g, ' ');
    var buf = textEncoder.encode(str);
    var encodedBytes = [];

    for (var _i2 = 0, len = buf.length; _i2 < len; _i2++) {
      var c = buf[_i2];

      if (_i2 <= len - 2 && c === 0x3d
      /* = */
      ) {
          var c1 = getHex(buf[_i2 + 1]);
          var c2 = getHex(buf[_i2 + 2]);

          if (c1 && c2) {
            var _c = parseInt(c1 + c2, 16);

            encodedBytes.push(_c);
            _i2 += 2;
            continue;
          }
        }

      encodedBytes.push(c);
    }

    byteStr = new ArrayBuffer(encodedBytes.length);
    var dataView = new DataView(byteStr);

    for (var _i3 = 0, _len = encodedBytes.length; _i3 < _len; _i3++) {
      dataView.setUint8(_i3, encodedBytes[_i3]);
    }
  } else if (encoding === 'B') {
    byteStr = decodeBase64(str.replace(/[^a-zA-Z0-9\+\/=]+/g, ''));
  } else {
    // keep as is, convert ArrayBuffer to unicode string, assume utf8
    byteStr = textEncoder.encode(str);
  }

  return getDecoder(charset).decode(byteStr);
}

function decodeWords(str) {
  var joinString = true;
  var done = false;

  while (!done) {
    var result = (str || '').toString() // find base64 words that can be joined
    .replace(/(=\?([^?]+)\?[Bb]\?([^?]*)\?=)\s*(?==\?([^?]+)\?[Bb]\?[^?]*\?=)/g, function (match, left, chLeft, encodedLeftStr, chRight) {
      if (!joinString) {
        return match;
      } // only mark b64 chunks to be joined if charsets match and left side does not end with =


      if (chLeft === chRight && encodedLeftStr.length % 4 === 0 && !/=$/.test(encodedLeftStr)) {
        // set a joiner marker
        return left + '__\x00JOIN\x00__';
      }

      return match;
    }) // find QP words that can be joined
    .replace(/(=\?([^?]+)\?[Qq]\?[^?]*\?=)\s*(?==\?([^?]+)\?[Qq]\?[^?]*\?=)/g, function (match, left, chLeft, chRight) {
      if (!joinString) {
        return match;
      } // only mark QP chunks to be joined if charsets match


      if (chLeft === chRight) {
        // set a joiner marker
        return left + '__\x00JOIN\x00__';
      }

      return match;
    }) // join base64 encoded words
    .replace(/(\?=)?__\x00JOIN\x00__(=\?([^?]+)\?[QqBb]\?)?/g, '') // remove spaces between mime encoded words
    .replace(/(=\?[^?]+\?[QqBb]\?[^?]*\?=)\s+(?==\?[^?]+\?[QqBb]\?[^?]*\?=)/g, '$1') // decode words
    .replace(/=\?([\w_\-*]+)\?([QqBb])\?([^?]*)\?=/g, function (m, charset, encoding, text) {
      return decodeWord(charset, encoding, text);
    });

    if (joinString && result.indexOf("\uFFFD") >= 0) {
      // text contains \ufffd (EF BF BD), so unicode conversion failed, retry without joining strings
      joinString = false;
    } else {
      return result;
    }
  }
}

function decodeURIComponentWithCharset(encodedStr, charset) {
  charset = charset || 'utf-8';
  var encodedBytes = [];

  for (var _i4 = 0; _i4 < encodedStr.length; _i4++) {
    var c = encodedStr.charAt(_i4);

    if (c === '%' && /^[a-f0-9]{2}/i.test(encodedStr.substr(_i4 + 1, 2))) {
      // encoded sequence
      var _byte = encodedStr.substr(_i4 + 1, 2);

      _i4 += 2;
      encodedBytes.push(parseInt(_byte, 16));
    } else if (c.charCodeAt(0) > 126) {
      c = textEncoder.encode(c);

      for (var j = 0; j < c.length; j++) {
        encodedBytes.push(c[j]);
      }
    } else {
      // "normal" char
      encodedBytes.push(c.charCodeAt(0));
    }
  }

  var byteStr = new ArrayBuffer(encodedBytes.length);
  var dataView = new DataView(byteStr);

  for (var _i5 = 0, len = encodedBytes.length; _i5 < len; _i5++) {
    dataView.setUint8(_i5, encodedBytes[_i5]);
  }

  return getDecoder(charset).decode(byteStr);
}

function decodeParameterValueContinuations(header) {
  // handle parameter value continuations
  // https://tools.ietf.org/html/rfc2231#section-3
  // preprocess values
  var paramKeys = new Map();
  Object.keys(header.params).forEach(function (key) {
    var match = key.match(/\*((\d+)\*?)?$/);

    if (!match) {
      // nothing to do here, does not seem like a continuation param
      return;
    }

    var actualKey = key.substr(0, match.index).toLowerCase();
    var nr = Number(match[2]) || 0;
    var paramVal;

    if (!paramKeys.has(actualKey)) {
      paramVal = {
        charset: false,
        values: []
      };
      paramKeys.set(actualKey, paramVal);
    } else {
      paramVal = paramKeys.get(actualKey);
    }

    var value = header.params[key];

    if (nr === 0 && match[0].charAt(match[0].length - 1) === '*' && (match = value.match(/^([^']*)'[^']*'(.*)$/))) {
      paramVal.charset = match[1] || 'utf-8';
      value = match[2];
    }

    paramVal.values.push({
      nr: nr,
      value: value
    }); // remove the old reference

    delete header.params[key];
  });
  paramKeys.forEach(function (paramVal, key) {
    header.params[key] = decodeURIComponentWithCharset(paramVal.values.sort(function (a, b) {
      return a.nr - b.nr;
    }).map(function (a) {
      return a.value;
    }).join(''), paramVal.charset);
  });
}
