// postal-mime/src/dist/mime-node.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _decodeStrings = require("./decode-strings.js");

var _passThroughDecoder = _interopRequireDefault(require("./pass-through-decoder.js"));

var _base64Decoder = _interopRequireDefault(require("./base64-decoder.js"));

var _qpDecoder = _interopRequireDefault(require("./qp-decoder.js"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var defaultDecoder = (0, _decodeStrings.getDecoder)();

var MimeNode =
/*#__PURE__*/
function () {
  function MimeNode(options) {
    _classCallCheck(this, MimeNode);

    this.options = options || {};
    this.postalMime = this.options.postalMime;
    this.root = !!this.options.parentNode;
    this.childNodes = [];

    if (this.options.parentNode) {
      this.parentNode = this.options.parentNode;
      this.depth = this.parentNode.depth + 1;

      if (this.depth > this.options.maxNestingDepth) {
        throw new Error("Maximum MIME nesting depth of ".concat(this.options.maxNestingDepth, " levels exceeded"));
      }

      this.options.parentNode.childNodes.push(this);
    } else {
      this.depth = 0;
    }

    this.state = 'header';
    this.headerLines = [];
    this.headerSize = 0; // RFC 2046 Section 5.1.5: multipart/digest defaults to message/rfc822

    var parentMultipartType = this.options.parentMultipartType || null;
    var defaultContentType = parentMultipartType === 'digest' ? 'message/rfc822' : 'text/plain';
    this.contentType = {
      value: defaultContentType,
      "default": true
    };
    this.contentTransferEncoding = {
      value: '8bit'
    };
    this.contentDisposition = {
      value: ''
    };
    this.headers = [];
    this.contentDecoder = false;
  }

  _createClass(MimeNode, [{
    key: "setupContentDecoder",
    value: function setupContentDecoder(transferEncoding) {
      if (/base64/i.test(transferEncoding)) {
        this.contentDecoder = new _base64Decoder["default"]();
      } else if (/quoted-printable/i.test(transferEncoding)) {
        this.contentDecoder = new _qpDecoder["default"]({
          decoder: (0, _decodeStrings.getDecoder)(this.contentType.parsed.params.charset)
        });
      } else {
        this.contentDecoder = new _passThroughDecoder["default"]();
      }
    }
  }, {
    key: "finalize",
    value: function finalize() {
      var boundaries, i, boundary;
      return regeneratorRuntime.async(function finalize$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              if (!(this.state === 'finished')) {
                _context.next = 2;
                break;
              }

              return _context.abrupt("return");

            case 2:
              if (this.state === 'header') {
                this.processHeaders();
              } // remove self from boundary listing


              boundaries = this.postalMime.boundaries;
              i = boundaries.length - 1;

            case 5:
              if (!(i >= 0)) {
                _context.next = 13;
                break;
              }

              boundary = boundaries[i];

              if (!(boundary.node === this)) {
                _context.next = 10;
                break;
              }

              boundaries.splice(i, 1);
              return _context.abrupt("break", 13);

            case 10:
              i--;
              _context.next = 5;
              break;

            case 13:
              _context.next = 15;
              return regeneratorRuntime.awrap(this.finalizeChildNodes());

            case 15:
              if (!this.contentDecoder) {
                _context.next = 21;
                break;
              }

              _context.next = 18;
              return regeneratorRuntime.awrap(this.contentDecoder.finalize());

            case 18:
              _context.t0 = _context.sent;
              _context.next = 22;
              break;

            case 21:
              _context.t0 = null;

            case 22:
              this.content = _context.t0;
              this.state = 'finished';

            case 24:
            case "end":
              return _context.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: "finalizeChildNodes",
    value: function finalizeChildNodes() {
      var _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, childNode;

      return regeneratorRuntime.async(function finalizeChildNodes$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _iteratorNormalCompletion = true;
              _didIteratorError = false;
              _iteratorError = undefined;
              _context2.prev = 3;
              _iterator = this.childNodes[Symbol.iterator]();

            case 5:
              if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
                _context2.next = 12;
                break;
              }

              childNode = _step.value;
              _context2.next = 9;
              return regeneratorRuntime.awrap(childNode.finalize());

            case 9:
              _iteratorNormalCompletion = true;
              _context2.next = 5;
              break;

            case 12:
              _context2.next = 18;
              break;

            case 14:
              _context2.prev = 14;
              _context2.t0 = _context2["catch"](3);
              _didIteratorError = true;
              _iteratorError = _context2.t0;

            case 18:
              _context2.prev = 18;
              _context2.prev = 19;

              if (!_iteratorNormalCompletion && _iterator["return"] != null) {
                _iterator["return"]();
              }

            case 21:
              _context2.prev = 21;

              if (!_didIteratorError) {
                _context2.next = 24;
                break;
              }

              throw _iteratorError;

            case 24:
              return _context2.finish(21);

            case 25:
              return _context2.finish(18);

            case 26:
            case "end":
              return _context2.stop();
          }
        }
      }, null, this, [[3, 14, 18, 26], [19,, 21, 25]]);
    } // Strip RFC 822 comments (parenthesized text) from structured header values

  }, {
    key: "stripComments",
    value: function stripComments(str) {
      var result = '';
      var depth = 0;
      var escaped = false;
      var inQuote = false;

      for (var i = 0; i < str.length; i++) {
        var chr = str.charAt(i);

        if (escaped) {
          if (depth === 0) {
            result += chr;
          }

          escaped = false;
          continue;
        }

        if (chr === '\\') {
          escaped = true;

          if (depth === 0) {
            result += chr;
          }

          continue;
        }

        if (chr === '"' && depth === 0) {
          inQuote = !inQuote;
          result += chr;
          continue;
        }

        if (!inQuote) {
          if (chr === '(') {
            depth++;
            continue;
          }

          if (chr === ')' && depth > 0) {
            depth--;
            continue;
          }
        }

        if (depth === 0) {
          result += chr;
        }
      }

      return result;
    }
  }, {
    key: "parseStructuredHeader",
    value: function parseStructuredHeader(str) {
      // Strip RFC 822 comments before parsing
      str = this.stripComments(str);
      var response = {
        value: false,
        params: {}
      };
      var key = false;
      var value = '';
      var stage = 'value';
      var quote = false;
      var escaped = false;
      var chr;

      for (var i = 0, len = str.length; i < len; i++) {
        chr = str.charAt(i);

        switch (stage) {
          case 'key':
            if (chr === '=') {
              key = value.trim().toLowerCase();
              stage = 'value';
              value = '';
              break;
            }

            value += chr;
            break;

          case 'value':
            if (escaped) {
              value += chr;
            } else if (chr === '\\') {
              escaped = true;
              continue;
            } else if (quote && chr === quote) {
              quote = false;
            } else if (!quote && chr === '"') {
              quote = chr;
            } else if (!quote && chr === ';') {
              if (key === false) {
                response.value = value.trim();
              } else {
                response.params[key] = value.trim();
              }

              stage = 'key';
              value = '';
            } else {
              value += chr;
            }

            escaped = false;
            break;
        }
      } // finalize remainder


      value = value.trim();

      if (stage === 'value') {
        if (key === false) {
          // default value
          response.value = value;
        } else {
          // subkey value
          response.params[key] = value;
        }
      } else if (value) {
        // treat as key without value, see emptykey:
        // Header-Key: somevalue; key=value; emptykey
        response.params[value.toLowerCase()] = '';
      }

      if (response.value) {
        response.value = response.value.toLowerCase();
      } // convert Parameter Value Continuations into single strings


      (0, _decodeStrings.decodeParameterValueContinuations)(response);
      return response;
    }
  }, {
    key: "decodeFlowedText",
    value: function decodeFlowedText(str, delSp) {
      return str.split(/\r?\n/) // remove soft linebreaks
      // soft linebreaks are added after space symbols
      .reduce(function (previousValue, currentValue) {
        if (previousValue.endsWith(' ') && previousValue !== '-- ' && !previousValue.endsWith('\n-- ')) {
          if (delSp) {
            // delsp adds space to text to be able to fold it
            // these spaces can be removed once the text is unfolded
            return previousValue.slice(0, -1) + currentValue;
          } else {
            return previousValue + currentValue;
          }
        } else {
          return previousValue + '\n' + currentValue;
        }
      }) // remove whitespace stuffing
      // http://tools.ietf.org/html/rfc3676#section-4.4
      .replace(/^ /gm, '');
    }
  }, {
    key: "getTextContent",
    value: function getTextContent() {
      if (!this.content) {
        return '';
      }

      var str = (0, _decodeStrings.getDecoder)(this.contentType.parsed.params.charset).decode(this.content);

      if (/^flowed$/i.test(this.contentType.parsed.params.format)) {
        str = this.decodeFlowedText(str, /^yes$/i.test(this.contentType.parsed.params.delsp));
      }

      return str;
    }
  }, {
    key: "processHeaders",
    value: function processHeaders() {
      // First pass: merge folded headers (backward iteration)
      for (var i = this.headerLines.length - 1; i >= 0; i--) {
        var line = this.headerLines[i];

        if (i && /^\s/.test(line)) {
          this.headerLines[i - 1] += '\n' + line;
          this.headerLines.splice(i, 1);
        }
      } // Initialize rawHeaderLines to store unmodified lines


      this.rawHeaderLines = []; // Second pass: process headers (MUST be backward to maintain this.headers order)
      // The existing code iterates backward and postal-mime.js calls .reverse()
      // We must preserve this behavior to avoid breaking changes

      for (var _i = this.headerLines.length - 1; _i >= 0; _i--) {
        var rawLine = this.headerLines[_i]; // Extract key from raw line for rawHeaderLines

        var sep = rawLine.indexOf(':');
        var rawKey = sep < 0 ? rawLine.trim() : rawLine.substr(0, sep).trim(); // Store raw line with lowercase key

        this.rawHeaderLines.push({
          key: rawKey.toLowerCase(),
          line: rawLine
        }); // Normalize for this.headers (existing behavior - order preserved)

        var normalizedLine = rawLine.replace(/\s+/g, ' ');
        sep = normalizedLine.indexOf(':');
        var key = sep < 0 ? normalizedLine.trim() : normalizedLine.substr(0, sep).trim();
        var value = sep < 0 ? '' : normalizedLine.substr(sep + 1).trim();
        this.headers.push({
          key: key.toLowerCase(),
          originalKey: key,
          value: value
        });

        switch (key.toLowerCase()) {
          case 'content-type':
            if (this.contentType["default"]) {
              this.contentType = {
                value: value,
                parsed: {}
              };
            }

            break;

          case 'content-transfer-encoding':
            this.contentTransferEncoding = {
              value: value,
              parsed: {}
            };
            break;

          case 'content-disposition':
            this.contentDisposition = {
              value: value,
              parsed: {}
            };
            break;

          case 'content-id':
            this.contentId = value;
            break;

          case 'content-description':
            this.contentDescription = value;
            break;
        }
      }

      this.contentType.parsed = this.parseStructuredHeader(this.contentType.value);
      this.contentType.multipart = /^multipart\//i.test(this.contentType.parsed.value) ? this.contentType.parsed.value.substr(this.contentType.parsed.value.indexOf('/') + 1) : false;

      if (this.contentType.multipart && this.contentType.parsed.params.boundary) {
        // add self to boundary terminator listing
        this.postalMime.boundaries.push({
          value: _decodeStrings.textEncoder.encode(this.contentType.parsed.params.boundary),
          node: this
        });
      }

      this.contentDisposition.parsed = this.parseStructuredHeader(this.contentDisposition.value);
      this.contentTransferEncoding.encoding = this.contentTransferEncoding.value.toLowerCase().split(/[^\w-]/).shift();
      this.setupContentDecoder(this.contentTransferEncoding.encoding);
    }
  }, {
    key: "feed",
    value: function feed(line) {
      switch (this.state) {
        case 'header':
          if (!line.length) {
            this.state = 'body';
            return this.processHeaders();
          }

          this.headerSize += line.length;

          if (this.headerSize > this.options.maxHeadersSize) {
            var error = new Error("Maximum header size of ".concat(this.options.maxHeadersSize, " bytes exceeded"));
            throw error;
          }

          this.headerLines.push(defaultDecoder.decode(line));
          break;

        case 'body':
          {
            // add line to body
            this.contentDecoder.update(line);
          }
      }
    }
  }]);

  return MimeNode;
}();

exports["default"] = MimeNode;
