// postal-mime/src/dist/qp-decoder.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _decodeStrings = require("./decode-strings.js");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

// Regex patterns compiled once for performance
var VALID_QP_REGEX = /^=[a-f0-9]{2}$/i;
var QP_SPLIT_REGEX = /(?==[a-f0-9]{2})/i;
var SOFT_LINE_BREAK_REGEX = /=\r?\n/g;
var PARTIAL_QP_ENDING_REGEX = /=[a-fA-F0-9]?$/;

var QPDecoder =
/*#__PURE__*/
function () {
  function QPDecoder(opts) {
    _classCallCheck(this, QPDecoder);

    opts = opts || {};
    this.decoder = opts.decoder || new TextDecoder();
    this.maxChunkSize = 100 * 1024;
    this.remainder = '';
    this.chunks = [];
  }

  _createClass(QPDecoder, [{
    key: "decodeQPBytes",
    value: function decodeQPBytes(encodedBytes) {
      var buf = new ArrayBuffer(encodedBytes.length);
      var dataView = new DataView(buf);

      for (var i = 0, len = encodedBytes.length; i < len; i++) {
        dataView.setUint8(i, parseInt(encodedBytes[i], 16));
      }

      return buf;
    }
  }, {
    key: "decodeChunks",
    value: function decodeChunks(str) {
      // unwrap newlines
      str = str.replace(SOFT_LINE_BREAK_REGEX, '');
      var list = str.split(QP_SPLIT_REGEX);
      var encodedBytes = [];
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = list[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var part = _step.value;

          if (part.charAt(0) !== '=') {
            if (encodedBytes.length) {
              this.chunks.push(this.decodeQPBytes(encodedBytes));
              encodedBytes = [];
            }

            this.chunks.push(part);
            continue;
          }

          if (part.length === 3) {
            // Validate that this is actually a valid QP sequence
            if (VALID_QP_REGEX.test(part)) {
              encodedBytes.push(part.substr(1));
            } else {
              // Not a valid QP sequence, treat as literal text
              if (encodedBytes.length) {
                this.chunks.push(this.decodeQPBytes(encodedBytes));
                encodedBytes = [];
              }

              this.chunks.push(part);
            }

            continue;
          }

          if (part.length > 3) {
            // First 3 chars should be a valid QP sequence
            var firstThree = part.substr(0, 3);

            if (VALID_QP_REGEX.test(firstThree)) {
              encodedBytes.push(part.substr(1, 2));
              this.chunks.push(this.decodeQPBytes(encodedBytes));
              encodedBytes = [];
              part = part.substr(3);
              this.chunks.push(part);
            } else {
              // Not a valid QP sequence, treat entire part as literal
              if (encodedBytes.length) {
                this.chunks.push(this.decodeQPBytes(encodedBytes));
                encodedBytes = [];
              }

              this.chunks.push(part);
            }
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      if (encodedBytes.length) {
        this.chunks.push(this.decodeQPBytes(encodedBytes));
      }
    }
  }, {
    key: "update",
    value: function update(buffer) {
      // expect full lines, so add line terminator as well
      var str = this.decoder.decode(buffer) + '\n';
      str = this.remainder + str;

      if (str.length < this.maxChunkSize) {
        this.remainder = str;
        return;
      }

      this.remainder = '';
      var partialEnding = str.match(PARTIAL_QP_ENDING_REGEX);

      if (partialEnding) {
        if (partialEnding.index === 0) {
          this.remainder = str;
          return;
        }

        this.remainder = str.substr(partialEnding.index);
        str = str.substr(0, partialEnding.index);
      }

      this.decodeChunks(str);
    }
  }, {
    key: "finalize",
    value: function finalize() {
      if (this.remainder.length) {
        this.decodeChunks(this.remainder);
        this.remainder = '';
      } // convert an array of arraybuffers into a blob and then back into a single arraybuffer


      return (0, _decodeStrings.blobToArrayBuffer)(new Blob(this.chunks, {
        type: 'application/octet-stream'
      }));
    }
  }]);

  return QPDecoder;
}();

exports["default"] = QPDecoder;
