// standardwebhooks/dist/dist/index.dev.js
"use strict";

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _wrapNativeSuper(Class) { var _cache = typeof Map === "function" ? new Map() : undefined; _wrapNativeSuper = function _wrapNativeSuper(Class) { if (Class === null || !_isNativeFunction(Class)) return Class; if (typeof Class !== "function") { throw new TypeError("Super expression must either be null or a function"); } if (typeof _cache !== "undefined") { if (_cache.has(Class)) return _cache.get(Class); _cache.set(Class, Wrapper); } function Wrapper() { return _construct(Class, arguments, _getPrototypeOf(this).constructor); } Wrapper.prototype = Object.create(Class.prototype, { constructor: { value: Wrapper, enumerable: false, writable: true, configurable: true } }); return _setPrototypeOf(Wrapper, Class); }; return _wrapNativeSuper(Class); }

function isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _construct(Parent, args, Class) { if (isNativeReflectConstruct()) { _construct = Reflect.construct; } else { _construct = function _construct(Parent, args, Class) { var a = [null]; a.push.apply(a, args); var Constructor = Function.bind.apply(Parent, a); var instance = new Constructor(); if (Class) _setPrototypeOf(instance, Class.prototype); return instance; }; } return _construct.apply(null, arguments); }

function _isNativeFunction(fn) { return Function.toString.call(fn).indexOf("[native code]") !== -1; }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Webhook = exports.WebhookVerificationError = void 0;

var timing_safe_equal_1 = require("./timing_safe_equal");

var base64 = require("@stablelib/base64");

var sha256 = require("fast-sha256");

var WEBHOOK_TOLERANCE_IN_SECONDS = 5 * 60;

var ExtendableError =
/*#__PURE__*/
function (_Error) {
  _inherits(ExtendableError, _Error);

  function ExtendableError(message) {
    var _this;

    _classCallCheck(this, ExtendableError);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(ExtendableError).call(this, message));
    Object.setPrototypeOf(_assertThisInitialized(_this), ExtendableError.prototype);
    _this.name = "ExtendableError";
    _this.stack = new Error(message).stack;
    return _this;
  }

  return ExtendableError;
}(_wrapNativeSuper(Error));

var WebhookVerificationError =
/*#__PURE__*/
function (_ExtendableError) {
  _inherits(WebhookVerificationError, _ExtendableError);

  function WebhookVerificationError(message) {
    var _this2;

    _classCallCheck(this, WebhookVerificationError);

    _this2 = _possibleConstructorReturn(this, _getPrototypeOf(WebhookVerificationError).call(this, message));
    Object.setPrototypeOf(_assertThisInitialized(_this2), WebhookVerificationError.prototype);
    _this2.name = "WebhookVerificationError";
    return _this2;
  }

  return WebhookVerificationError;
}(ExtendableError);

exports.WebhookVerificationError = WebhookVerificationError;

var Webhook =
/*#__PURE__*/
function () {
  function Webhook(secret, options) {
    _classCallCheck(this, Webhook);

    if (!secret) {
      throw new Error("Secret can't be empty.");
    }

    if ((options === null || options === void 0 ? void 0 : options.format) === "raw") {
      if (secret instanceof Uint8Array) {
        this.key = secret;
      } else {
        this.key = Uint8Array.from(secret, function (c) {
          return c.charCodeAt(0);
        });
      }
    } else {
      if (typeof secret !== "string") {
        throw new Error("Expected secret to be of type string");
      }

      if (secret.startsWith(Webhook.prefix)) {
        secret = secret.substring(Webhook.prefix.length);
      }

      this.key = base64.decode(secret);
    }
  }

  _createClass(Webhook, [{
    key: "verify",
    value: function verify(payload, headers_) {
      var headers = {};

      for (var _i = 0, _Object$keys = Object.keys(headers_); _i < _Object$keys.length; _i++) {
        var key = _Object$keys[_i];
        headers[key.toLowerCase()] = headers_[key];
      }

      var msgId = headers["webhook-id"];
      var msgSignature = headers["webhook-signature"];
      var msgTimestamp = headers["webhook-timestamp"];

      if (!msgSignature || !msgId || !msgTimestamp) {
        throw new WebhookVerificationError("Missing required headers");
      }

      var timestamp = this.verifyTimestamp(msgTimestamp);
      var computedSignature = this.sign(msgId, timestamp, payload);
      var expectedSignature = computedSignature.split(",")[1];
      var passedSignatures = msgSignature.split(" ");
      var encoder = new globalThis.TextEncoder();
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = passedSignatures[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var versionedSignature = _step.value;

          var _versionedSignature$s = versionedSignature.split(","),
              _versionedSignature$s2 = _slicedToArray(_versionedSignature$s, 2),
              version = _versionedSignature$s2[0],
              signature = _versionedSignature$s2[1];

          if (version !== "v1") {
            continue;
          }

          if ((0, timing_safe_equal_1.timingSafeEqual)(encoder.encode(signature), encoder.encode(expectedSignature))) {
            return JSON.parse(payload.toString());
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      throw new WebhookVerificationError("No matching signature found");
    }
  }, {
    key: "sign",
    value: function sign(msgId, timestamp, payload) {
      if (typeof payload === "string") {} else if (payload.constructor.name === "Buffer") {
        payload = payload.toString();
      } else {
        throw new Error("Expected payload to be of type string or Buffer.");
      }

      var encoder = new TextEncoder();
      var timestampNumber = Math.floor(timestamp.getTime() / 1000);
      var toSign = encoder.encode("".concat(msgId, ".").concat(timestampNumber, ".").concat(payload));
      var expectedSignature = base64.encode(sha256.hmac(this.key, toSign));
      return "v1,".concat(expectedSignature);
    }
  }, {
    key: "verifyTimestamp",
    value: function verifyTimestamp(timestampHeader) {
      var now = Math.floor(Date.now() / 1000);
      var timestamp = parseInt(timestampHeader, 10);

      if (isNaN(timestamp)) {
        throw new WebhookVerificationError("Invalid Signature Headers");
      }

      if (now - timestamp > WEBHOOK_TOLERANCE_IN_SECONDS) {
        throw new WebhookVerificationError("Message timestamp too old");
      }

      if (timestamp > now + WEBHOOK_TOLERANCE_IN_SECONDS) {
        throw new WebhookVerificationError("Message timestamp too new");
      }

      return new Date(timestamp * 1000);
    }
  }]);

  return Webhook;
}();

exports.Webhook = Webhook;
Webhook.prefix = "whsec_";
