// svix/dist/dist/HttpErrors.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HTTPValidationError = exports.ValidationError = exports.HttpErrorOut = void 0;

var HttpErrorOut =
/*#__PURE__*/
function () {
  function HttpErrorOut() {
    _classCallCheck(this, HttpErrorOut);
  }

  _createClass(HttpErrorOut, null, [{
    key: "getAttributeTypeMap",
    value: function getAttributeTypeMap() {
      return HttpErrorOut.attributeTypeMap;
    }
  }]);

  return HttpErrorOut;
}();

exports.HttpErrorOut = HttpErrorOut;
HttpErrorOut.discriminator = undefined;
HttpErrorOut.mapping = undefined;
HttpErrorOut.attributeTypeMap = [{
  name: "code",
  baseName: "code",
  type: "string",
  format: ""
}, {
  name: "detail",
  baseName: "detail",
  type: "string",
  format: ""
}];

var ValidationError =
/*#__PURE__*/
function () {
  function ValidationError() {
    _classCallCheck(this, ValidationError);
  }

  _createClass(ValidationError, null, [{
    key: "getAttributeTypeMap",
    value: function getAttributeTypeMap() {
      return ValidationError.attributeTypeMap;
    }
  }]);

  return ValidationError;
}();

exports.ValidationError = ValidationError;
ValidationError.discriminator = undefined;
ValidationError.mapping = undefined;
ValidationError.attributeTypeMap = [{
  name: "loc",
  baseName: "loc",
  type: "Array<string>",
  format: ""
}, {
  name: "msg",
  baseName: "msg",
  type: "string",
  format: ""
}, {
  name: "type",
  baseName: "type",
  type: "string",
  format: ""
}];

var HTTPValidationError =
/*#__PURE__*/
function () {
  function HTTPValidationError() {
    _classCallCheck(this, HTTPValidationError);
  }

  _createClass(HTTPValidationError, null, [{
    key: "getAttributeTypeMap",
    value: function getAttributeTypeMap() {
      return HTTPValidationError.attributeTypeMap;
    }
  }]);

  return HTTPValidationError;
}();

exports.HTTPValidationError = HTTPValidationError;
HTTPValidationError.discriminator = undefined;
HTTPValidationError.mapping = undefined;
HTTPValidationError.attributeTypeMap = [{
  name: "detail",
  baseName: "detail",
  type: "Array<ValidationError>",
  format: ""
}];
