// svix/dist/dist/index.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var __createBinding = void 0 && (void 0).__createBinding || (Object.create ? function (o, m, k, k2) {
  if (k2 === undefined) k2 = k;
  var desc = Object.getOwnPropertyDescriptor(m, k);

  if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
    desc = {
      enumerable: true,
      get: function get() {
        return m[k];
      }
    };
  }

  Object.defineProperty(o, k2, desc);
} : function (o, m, k, k2) {
  if (k2 === undefined) k2 = k;
  o[k2] = m[k];
});

var __exportStar = void 0 && (void 0).__exportStar || function (m, exports) {
  for (var p in m) {
    if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
  }
};

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Svix = exports.messageInRaw = exports.ValidationError = exports.HttpErrorOut = exports.HTTPValidationError = exports.ApiException = void 0;

var application_1 = require("./api/application");

var authentication_1 = require("./api/authentication");

var backgroundTask_1 = require("./api/backgroundTask");

var connector_1 = require("./api/connector");

var endpoint_1 = require("./api/endpoint");

var environment_1 = require("./api/environment");

var eventType_1 = require("./api/eventType");

var health_1 = require("./api/health");

var ingest_1 = require("./api/ingest");

var integration_1 = require("./api/integration");

var message_1 = require("./api/message");

var messageAttempt_1 = require("./api/messageAttempt");

var operationalWebhook_1 = require("./api/operationalWebhook");

var statistics_1 = require("./api/statistics");

var streaming_1 = require("./api/streaming");

var operationalWebhookEndpoint_1 = require("./api/operationalWebhookEndpoint");

var util_1 = require("./util");

Object.defineProperty(exports, "ApiException", {
  enumerable: true,
  get: function get() {
    return util_1.ApiException;
  }
});

var HttpErrors_1 = require("./HttpErrors");

Object.defineProperty(exports, "HTTPValidationError", {
  enumerable: true,
  get: function get() {
    return HttpErrors_1.HTTPValidationError;
  }
});
Object.defineProperty(exports, "HttpErrorOut", {
  enumerable: true,
  get: function get() {
    return HttpErrors_1.HttpErrorOut;
  }
});
Object.defineProperty(exports, "ValidationError", {
  enumerable: true,
  get: function get() {
    return HttpErrors_1.ValidationError;
  }
});

__exportStar(require("./webhook"), exports);

__exportStar(require("./models/index"), exports);

var message_2 = require("./api/message");

Object.defineProperty(exports, "messageInRaw", {
  enumerable: true,
  get: function get() {
    return message_2.messageInRaw;
  }
});
var REGIONS = [{
  region: "us",
  url: "https://api.us.svix.com"
}, {
  region: "eu",
  url: "https://api.eu.svix.com"
}, {
  region: "in",
  url: "https://api.in.svix.com"
}, {
  region: "ca",
  url: "https://api.ca.svix.com"
}, {
  region: "au",
  url: "https://api.au.svix.com"
}];

var Svix =
/*#__PURE__*/
function () {
  function Svix(token) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, Svix);

    var _a, _b, _c;

    var regionalUrl = (_a = REGIONS.find(function (x) {
      return x.region === token.split(".")[1];
    })) === null || _a === void 0 ? void 0 : _a.url;
    var baseUrl = (_c = (_b = options.serverUrl) !== null && _b !== void 0 ? _b : regionalUrl) !== null && _c !== void 0 ? _c : "https://api.svix.com";

    if (options.retryScheduleInMs) {
      this.requestCtx = {
        baseUrl: baseUrl,
        token: token,
        timeout: options.requestTimeout,
        retryScheduleInMs: options.retryScheduleInMs,
        fetch: options.fetch
      };
      return;
    }

    if (options.numRetries) {
      this.requestCtx = {
        baseUrl: baseUrl,
        token: token,
        timeout: options.requestTimeout,
        numRetries: options.numRetries,
        fetch: options.fetch
      };
      return;
    }

    this.requestCtx = {
      baseUrl: baseUrl,
      token: token,
      timeout: options.requestTimeout,
      fetch: options.fetch
    };
  }

  _createClass(Svix, [{
    key: "application",
    get: function get() {
      return new application_1.Application(this.requestCtx);
    }
  }, {
    key: "authentication",
    get: function get() {
      return new authentication_1.Authentication(this.requestCtx);
    }
  }, {
    key: "backgroundTask",
    get: function get() {
      return new backgroundTask_1.BackgroundTask(this.requestCtx);
    }
  }, {
    key: "connector",
    get: function get() {
      return new connector_1.Connector(this.requestCtx);
    }
  }, {
    key: "endpoint",
    get: function get() {
      return new endpoint_1.Endpoint(this.requestCtx);
    }
  }, {
    key: "environment",
    get: function get() {
      return new environment_1.Environment(this.requestCtx);
    }
  }, {
    key: "eventType",
    get: function get() {
      return new eventType_1.EventType(this.requestCtx);
    }
  }, {
    key: "health",
    get: function get() {
      return new health_1.Health(this.requestCtx);
    }
  }, {
    key: "ingest",
    get: function get() {
      return new ingest_1.Ingest(this.requestCtx);
    }
  }, {
    key: "integration",
    get: function get() {
      return new integration_1.Integration(this.requestCtx);
    }
  }, {
    key: "message",
    get: function get() {
      return new message_1.Message(this.requestCtx);
    }
  }, {
    key: "messageAttempt",
    get: function get() {
      return new messageAttempt_1.MessageAttempt(this.requestCtx);
    }
  }, {
    key: "operationalWebhook",
    get: function get() {
      return new operationalWebhook_1.OperationalWebhook(this.requestCtx);
    }
  }, {
    key: "statistics",
    get: function get() {
      return new statistics_1.Statistics(this.requestCtx);
    }
  }, {
    key: "streaming",
    get: function get() {
      return new streaming_1.Streaming(this.requestCtx);
    }
  }, {
    key: "operationalWebhookEndpoint",
    get: function get() {
      return new operationalWebhookEndpoint_1.OperationalWebhookEndpoint(this.requestCtx);
    }
  }]);

  return Svix;
}();

exports.Svix = Svix;
