// svix/dist/dist/request.dev.js
"use strict";

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var __awaiter = void 0 && (void 0).__awaiter || function (thisArg, _arguments, P, generator) {
  function adopt(value) {
    return value instanceof P ? value : new P(function (resolve) {
      resolve(value);
    });
  }

  return new (P || (P = Promise))(function (resolve, reject) {
    function fulfilled(value) {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    }

    function rejected(value) {
      try {
        step(generator["throw"](value));
      } catch (e) {
        reject(e);
      }
    }

    function step(result) {
      result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
    }

    step((generator = generator.apply(thisArg, _arguments || [])).next());
  });
};

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SvixRequest = exports.HttpMethod = exports.LIB_VERSION = void 0;

var util_1 = require("./util");

exports.LIB_VERSION = "1.92.2";
var USER_AGENT = "svix-libs/".concat(exports.LIB_VERSION, "/javascript");
var HttpMethod;

(function (HttpMethod) {
  HttpMethod["GET"] = "GET";
  HttpMethod["HEAD"] = "HEAD";
  HttpMethod["POST"] = "POST";
  HttpMethod["PUT"] = "PUT";
  HttpMethod["DELETE"] = "DELETE";
  HttpMethod["CONNECT"] = "CONNECT";
  HttpMethod["OPTIONS"] = "OPTIONS";
  HttpMethod["TRACE"] = "TRACE";
  HttpMethod["PATCH"] = "PATCH";
})(HttpMethod = exports.HttpMethod || (exports.HttpMethod = {}));

var SvixRequest =
/*#__PURE__*/
function () {
  function SvixRequest(method, path) {
    _classCallCheck(this, SvixRequest);

    this.method = method;
    this.path = path;
    this.queryParams = {};
    this.headerParams = {};
  }

  _createClass(SvixRequest, [{
    key: "setPathParam",
    value: function setPathParam(name, value) {
      var newPath = this.path.replace("{".concat(name, "}"), encodeURIComponent(value));

      if (this.path === newPath) {
        throw new Error("path parameter ".concat(name, " not found"));
      }

      this.path = newPath;
    }
  }, {
    key: "setQueryParams",
    value: function setQueryParams(params) {
      for (var _i = 0, _Object$entries = Object.entries(params); _i < _Object$entries.length; _i++) {
        var _Object$entries$_i = _slicedToArray(_Object$entries[_i], 2),
            name = _Object$entries$_i[0],
            value = _Object$entries$_i[1];

        this.setQueryParam(name, value);
      }
    }
  }, {
    key: "setQueryParam",
    value: function setQueryParam(name, value) {
      if (value === undefined || value === null) {
        return;
      }

      if (typeof value === "string") {
        this.queryParams[name] = value;
      } else if (typeof value === "boolean" || typeof value === "number") {
        this.queryParams[name] = value.toString();
      } else if (value instanceof Date) {
        this.queryParams[name] = value.toISOString();
      } else if (Array.isArray(value)) {
        if (value.length > 0) {
          this.queryParams[name] = value.join(",");
        }
      } else {
        var _assert_unreachable = value;
        throw new Error("query parameter ".concat(name, " has unsupported type"));
      }
    }
  }, {
    key: "setHeaderParam",
    value: function setHeaderParam(name, value) {
      if (value === undefined) {
        return;
      }

      this.headerParams[name] = value;
    }
  }, {
    key: "setBody",
    value: function setBody(value) {
      this.body = JSON.stringify(value);
    }
  }, {
    key: "send",
    value: function send(ctx, parseResponseBody) {
      return __awaiter(this, void 0, void 0,
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee() {
        var response, responseBody;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.sendInner(ctx);

              case 2:
                response = _context.sent;

                if (!(response.status === 204)) {
                  _context.next = 5;
                  break;
                }

                return _context.abrupt("return", null);

              case 5:
                _context.next = 7;
                return response.text();

              case 7:
                responseBody = _context.sent;
                return _context.abrupt("return", parseResponseBody(JSON.parse(responseBody)));

              case 9:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));
    }
  }, {
    key: "sendNoResponseBody",
    value: function sendNoResponseBody(ctx) {
      return __awaiter(this, void 0, void 0,
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee2() {
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.sendInner(ctx);

              case 2:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));
    }
  }, {
    key: "sendInner",
    value: function sendInner(ctx) {
      var _a, _b;

      return __awaiter(this, void 0, void 0,
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee3() {
        var url, _i2, _Object$entries2, _Object$entries2$_i, name, value, randomId, isCredentialsSupported, response;

        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                url = new URL(ctx.baseUrl + this.path);

                for (_i2 = 0, _Object$entries2 = Object.entries(this.queryParams); _i2 < _Object$entries2.length; _i2++) {
                  _Object$entries2$_i = _slicedToArray(_Object$entries2[_i2], 2), name = _Object$entries2$_i[0], value = _Object$entries2$_i[1];
                  url.searchParams.set(name, value);
                }

                if (this.headerParams["idempotency-key"] === undefined && this.method.toUpperCase() === "POST") {
                  this.headerParams["idempotency-key"] = "auto_".concat(crypto.randomUUID());
                }

                randomId = Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);

                if (this.body != null) {
                  this.headerParams["content-type"] = "application/json";
                }

                isCredentialsSupported = "credentials" in Request.prototype;
                _context3.next = 8;
                return sendWithRetry(url, {
                  method: this.method.toString(),
                  body: this.body,
                  headers: Object.assign({
                    accept: "application/json, */*;q=0.8",
                    authorization: "Bearer ".concat(ctx.token),
                    "user-agent": USER_AGENT,
                    "svix-req-id": randomId.toString()
                  }, this.headerParams),
                  credentials: isCredentialsSupported ? "same-origin" : undefined,
                  signal: ctx.timeout !== undefined ? AbortSignal.timeout(ctx.timeout) : undefined
                }, ctx.retryScheduleInMs, (_a = ctx.retryScheduleInMs) === null || _a === void 0 ? void 0 : _a[0], ((_b = ctx.retryScheduleInMs) === null || _b === void 0 ? void 0 : _b.length) || ctx.numRetries, ctx.fetch);

              case 8:
                response = _context3.sent;
                return _context3.abrupt("return", filterResponseForErrors(response));

              case 10:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));
    }
  }]);

  return SvixRequest;
}();

exports.SvixRequest = SvixRequest;

function filterResponseForErrors(response) {
  return __awaiter(this, void 0, void 0,
  /*#__PURE__*/
  regeneratorRuntime.mark(function _callee4() {
    var responseBody;
    return regeneratorRuntime.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            if (!(response.status < 300)) {
              _context4.next = 2;
              break;
            }

            return _context4.abrupt("return", response);

          case 2:
            _context4.next = 4;
            return response.text();

          case 4:
            responseBody = _context4.sent;

            if (!(response.status === 422)) {
              _context4.next = 7;
              break;
            }

            throw new util_1.ApiException(response.status, JSON.parse(responseBody), response.headers);

          case 7:
            if (!(response.status >= 400 && response.status <= 499)) {
              _context4.next = 9;
              break;
            }

            throw new util_1.ApiException(response.status, JSON.parse(responseBody), response.headers);

          case 9:
            throw new util_1.ApiException(response.status, responseBody, response.headers);

          case 10:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));
}

function sendWithRetry(url, init, retryScheduleInMs) {
  var nextInterval = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 50;
  var triesLeft = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : 2;
  var fetchImpl = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : fetch;
  var retryCount = arguments.length > 6 && arguments[6] !== undefined ? arguments[6] : 1;
  return __awaiter(this, void 0, void 0,
  /*#__PURE__*/
  regeneratorRuntime.mark(function _callee5() {
    var sleep, response;
    return regeneratorRuntime.wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            sleep = function sleep(interval) {
              return new Promise(function (resolve) {
                return setTimeout(resolve, interval);
              });
            };

            _context5.prev = 1;
            _context5.next = 4;
            return fetchImpl(url, init);

          case 4:
            response = _context5.sent;

            if (!(triesLeft <= 0 || response.status < 500)) {
              _context5.next = 7;
              break;
            }

            return _context5.abrupt("return", response);

          case 7:
            _context5.next = 13;
            break;

          case 9:
            _context5.prev = 9;
            _context5.t0 = _context5["catch"](1);

            if (!(triesLeft <= 0)) {
              _context5.next = 13;
              break;
            }

            throw _context5.t0;

          case 13:
            _context5.next = 15;
            return sleep(nextInterval);

          case 15:
            init.headers["svix-retry-count"] = retryCount.toString();
            nextInterval = (retryScheduleInMs === null || retryScheduleInMs === void 0 ? void 0 : retryScheduleInMs[retryCount]) || nextInterval * 2;
            _context5.next = 19;
            return sendWithRetry(url, init, retryScheduleInMs, nextInterval, --triesLeft, fetchImpl, ++retryCount);

          case 19:
            return _context5.abrupt("return", _context5.sent);

          case 20:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5, null, [[1, 9]]);
  }));
}
