// svix/src/dist/HttpErrors.js
"use strict";
exports.__esModule = true;
exports.HTTPValidationError = exports.ValidationError = exports.HttpErrorOut = void 0;
var HttpErrorOut = /** @class */ (function () {
    function HttpErrorOut() {
    }
    HttpErrorOut.getAttributeTypeMap = function () {
        return HttpErrorOut.attributeTypeMap;
    };
    HttpErrorOut.discriminator = undefined;
    HttpErrorOut.mapping = undefined;
    HttpErrorOut.attributeTypeMap = [
        {
            name: "code",
            baseName: "code",
            type: "string",
            format: ""
        },
        {
            name: "detail",
            baseName: "detail",
            type: "string",
            format: ""
        },
    ];
    return HttpErrorOut;
}());
exports.HttpErrorOut = HttpErrorOut;
/**
 * Validation errors have their own schema to provide context for invalid requests eg. mismatched types and out of bounds values. There may be any number of these per 422 UNPROCESSABLE ENTITY error.
 */
var ValidationError = /** @class */ (function () {
    function ValidationError() {
    }
    ValidationError.getAttributeTypeMap = function () {
        return ValidationError.attributeTypeMap;
    };
    ValidationError.discriminator = undefined;
    ValidationError.mapping = undefined;
    ValidationError.attributeTypeMap = [
        {
            name: "loc",
            baseName: "loc",
            type: "Array<string>",
            format: ""
        },
        {
            name: "msg",
            baseName: "msg",
            type: "string",
            format: ""
        },
        {
            name: "type",
            baseName: "type",
            type: "string",
            format: ""
        },
    ];
    return ValidationError;
}());
exports.ValidationError = ValidationError;
var HTTPValidationError = /** @class */ (function () {
    function HTTPValidationError() {
    }
    HTTPValidationError.getAttributeTypeMap = function () {
        return HTTPValidationError.attributeTypeMap;
    };
    HTTPValidationError.discriminator = undefined;
    HTTPValidationError.mapping = undefined;
    HTTPValidationError.attributeTypeMap = [
        {
            name: "detail",
            baseName: "detail",
            type: "Array<ValidationError>",
            format: ""
        },
    ];
    return HTTPValidationError;
}());
exports.HTTPValidationError = HTTPValidationError;
