// svix/src/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
exports.__esModule = true;
exports.Svix = void 0;
// this file is @generated
var application_1 = require("./api/application");
var authentication_1 = require("./api/authentication");
var backgroundTask_1 = require("./api/backgroundTask");
var connector_1 = require("./api/connector");
var endpoint_1 = require("./api/endpoint");
var environment_1 = require("./api/environment");
var eventType_1 = require("./api/eventType");
var health_1 = require("./api/health");
var ingest_1 = require("./api/ingest");
var integration_1 = require("./api/integration");
var message_1 = require("./api/message");
var messageAttempt_1 = require("./api/messageAttempt");
var operationalWebhook_1 = require("./api/operationalWebhook");
var statistics_1 = require("./api/statistics");
var streaming_1 = require("./api/streaming");
var operationalWebhookEndpoint_1 = require("./api/operationalWebhookEndpoint");
var util_1 = require("./util");
__createBinding(exports, util_1, "type");
__createBinding(exports, util_1, "ApiException");
var HttpErrors_1 = require("./HttpErrors");
__createBinding(exports, HttpErrors_1, "HTTPValidationError");
__createBinding(exports, HttpErrors_1, "HttpErrorOut");
__createBinding(exports, HttpErrors_1, "ValidationError");
__exportStar(require("./webhook"), exports);
__exportStar(require("./models/index"), exports);
var message_2 = require("./api/message");
__createBinding(exports, message_2, "type");
__createBinding(exports, message_2, "messageInRaw");
var REGIONS = [
    { region: "us", url: "https://api.us.svix.com" },
    { region: "eu", url: "https://api.eu.svix.com" },
    { region: "in", url: "https://api.in.svix.com" },
    { region: "ca", url: "https://api.ca.svix.com" },
    { region: "au", url: "https://api.au.svix.com" },
];
var Svix = /** @class */ (function () {
    function Svix(token, options) {
        if (options === void 0) { options = {}; }
        var _a, _b, _c;
        var regionalUrl = (_a = REGIONS.find(function (x) { return x.region === token.split(".")[1]; })) === null || _a === void 0 ? void 0 : _a.url;
        var baseUrl = (_c = (_b = options.serverUrl) !== null && _b !== void 0 ? _b : regionalUrl) !== null && _c !== void 0 ? _c : "https://api.svix.com";
        if (options.retryScheduleInMs) {
            this.requestCtx = {
                baseUrl: baseUrl,
                token: token,
                timeout: options.requestTimeout,
                retryScheduleInMs: options.retryScheduleInMs,
                fetch: options.fetch
            };
            return;
        }
        if (options.numRetries) {
            this.requestCtx = {
                baseUrl: baseUrl,
                token: token,
                timeout: options.requestTimeout,
                numRetries: options.numRetries,
                fetch: options.fetch
            };
            return;
        }
        this.requestCtx = {
            baseUrl: baseUrl,
            token: token,
            timeout: options.requestTimeout,
            fetch: options.fetch
        };
    }
    Object.defineProperty(Svix.prototype, "application", {
        get: function () {
            return new application_1.Application(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "authentication", {
        get: function () {
            return new authentication_1.Authentication(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "backgroundTask", {
        get: function () {
            return new backgroundTask_1.BackgroundTask(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "connector", {
        get: function () {
            return new connector_1.Connector(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "endpoint", {
        get: function () {
            return new endpoint_1.Endpoint(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "environment", {
        get: function () {
            return new environment_1.Environment(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "eventType", {
        get: function () {
            return new eventType_1.EventType(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "health", {
        get: function () {
            return new health_1.Health(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "ingest", {
        get: function () {
            return new ingest_1.Ingest(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "integration", {
        get: function () {
            return new integration_1.Integration(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "message", {
        get: function () {
            return new message_1.Message(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "messageAttempt", {
        get: function () {
            return new messageAttempt_1.MessageAttempt(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "operationalWebhook", {
        get: function () {
            return new operationalWebhook_1.OperationalWebhook(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "statistics", {
        get: function () {
            return new statistics_1.Statistics(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "streaming", {
        get: function () {
            return new streaming_1.Streaming(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Svix.prototype, "operationalWebhookEndpoint", {
        get: function () {
            return new operationalWebhookEndpoint_1.OperationalWebhookEndpoint(this.requestCtx);
        },
        enumerable: false,
        configurable: true
    });
    return Svix;
}());
exports.Svix = Svix;
