// svix/src/dist/KitchenSink.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
var _1 = require(".");
var node_test_1 = require("node:test");
var strict_1 = require("node:assert/strict");
function getTestClient() {
    var token = process.env.SVIX_TOKEN;
    var serverUrl = process.env.SVIX_SERVER_URL;
    if (!token || !serverUrl) {
        console.warn("Unable to construct test client: `SVIX_TOKEN` or `SVIX_SERVER_URL` unset.");
        return null;
    }
    return new _1.Svix(token, { serverUrl: serverUrl });
}
var client = getTestClient();
// Auto-skip tests in this module if we don't have a test client to work with.
node_test_1.test("e2e tests", { skip: client === null }, function (t) { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, t.test("endpoint crud", function () { return __awaiter(void 0, void 0, void 0, function () {
                    var appOut, e_1, e_2, epOut, epPatched;
                    var _a, _b, _c;
                    return __generator(this, function (_d) {
                        switch (_d.label) {
                            case 0:
                                if (client === null || client === undefined) {
                                    throw new Error("unreachable");
                                }
                                return [4 /*yield*/, client.application.create({ name: "App" })];
                            case 1:
                                appOut = _d.sent();
                                _d.label = 2;
                            case 2:
                                _d.trys.push([2, 4, , 5]);
                                return [4 /*yield*/, client.eventType.create({
                                        name: "event.started",
                                        description: "Something started"
                                    })];
                            case 3:
                                _d.sent();
                                return [3 /*break*/, 5];
                            case 4:
                                e_1 = _d.sent();
                                // Conflicts are expected from test run to test run, but other statuses are not.
                                strict_1.strict.deepEqual(e_1.code, 409);
                                return [3 /*break*/, 5];
                            case 5:
                                _d.trys.push([5, 7, , 8]);
                                return [4 /*yield*/, client.eventType.create({
                                        name: "event.ended",
                                        description: "Something ended"
                                    })];
                            case 6:
                                _d.sent();
                                return [3 /*break*/, 8];
                            case 7:
                                e_2 = _d.sent();
                                // Conflicts are expected from test run to test run, but other statuses are not.
                                strict_1.strict.deepEqual(e_2.code, 409);
                                return [3 /*break*/, 8];
                            case 8: return [4 /*yield*/, client.endpoint.create(appOut.id, {
                                    url: "https://example.svix.com/",
                                    channels: ["ch0", "ch1"]
                                })];
                            case 9:
                                epOut = _d.sent();
                                strict_1.strict.deepEqual((_a = epOut.channels) === null || _a === void 0 ? void 0 : _a.sort(), ["ch0", "ch1"]);
                                strict_1.strict.deepEqual(epOut.filterTypes || [], []);
                                return [4 /*yield*/, client.endpoint.patch(appOut.id, epOut.id, {
                                        filterTypes: ["event.started", "event.ended"]
                                    })];
                            case 10:
                                epPatched = _d.sent();
                                strict_1.strict.deepEqual((_b = epPatched.channels) === null || _b === void 0 ? void 0 : _b.sort(), ["ch0", "ch1"]);
                                strict_1.strict.deepEqual((_c = epPatched.filterTypes) === null || _c === void 0 ? void 0 : _c.sort(), ["event.ended", "event.started"]);
                                // Should not throw an error while trying to deserialize the empty body.
                                return [4 /*yield*/, client.endpoint["delete"](appOut.id, epOut.id)];
                            case 11:
                                // Should not throw an error while trying to deserialize the empty body.
                                _d.sent();
                                return [2 /*return*/];
                        }
                    });
                }); })];
            case 1:
                _a.sent();
                return [2 /*return*/];
        }
    });
}); });
