// svix/src/dist/mockttp.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
var _1 = require(".");
var node_test_1 = require("node:test");
var strict_1 = require("node:assert/strict");
var mockttp = require("mockttp");
var util_1 = require("./util");
var ordering_1 = require("./models/ordering");
var request_1 = require("./request");
var ApplicationOut = "{\"uid\":\"unique-identifier\",\"name\":\"My first application\",\"rateLimit\":0,\"id\":\"app_1srOrx2ZWZBpBUvZwXKQmoEYga2\",\"createdAt\":\"2019-08-24T14:15:22Z\",\"updatedAt\":\"2019-08-24T14:15:22Z\",\"metadata\":{\"property1\":\"string\",\"property2\":\"string\"}}";
var ListResponseMessageOut = "{\"data\":[{\"eventId\":\"unique-identifier\",\"eventType\":\"user.signup\",\"payload\":{\"email\":\"test@example.com\",\"type\":\"user.created\",\"username\":\"test_user\"},\"channels\":[\"project_123\",\"group_2\"],\"id\":\"msg_1srOrx2ZWZBpBUvZwXKQmoEYga2\",\"timestamp\":\"2019-08-24T14:15:22Z\",\"tags\":[\"project_1337\"]}],\"iterator\":\"iterator\",\"prevIterator\":\"-iterator\",\"done\":true}";
var ListResponseApplicationOut = "{\"data\":[{\"uid\":\"unique-identifier\",\"name\":\"My first application\",\"rateLimit\":0,\"id\":\"app_1srOrx2ZWZBpBUvZwXKQmoEYga2\",\"createdAt\":\"2019-08-24T14:15:22Z\",\"updatedAt\":\"2019-08-24T14:15:22Z\",\"metadata\":{\"property1\":\"string\",\"property2\":\"string\"}}],\"iterator\":\"iterator\",\"prevIterator\":\"-iterator\",\"done\":true}";
var ListResponseMessageAttemptOut = "{\"data\":[{\"url\":\"https://example.com/webhook/\",\"response\":\"{}\",\"responseStatusCode\":200,\"responseDurationMs\":0,\"status\":0,\"triggerType\":0,\"msgId\":\"msg_1srOrx2ZWZBpBUvZwXKQmoEYga2\",\"endpointId\":\"ep_1srOrx2ZWZBpBUvZwXKQmoEYga2\",\"id\":\"atmpt_1srOrx2ZWZBpBUvZwXKQmoEYga2\",\"timestamp\":\"2025-02-16T21:38:21.977Z\",\"msg\":{\"eventId\":\"unique-identifier\",\"eventType\":\"user.signup\",\"payload\":{\"email\":\"test@example.com\",\"type\":\"user.created\",\"username\":\"test_user\"},\"channels\":[\"project_123\",\"group_2\"],\"id\":\"msg_1srOrx2ZWZBpBUvZwXKQmoEYga2\",\"timestamp\":\"2025-02-16T21:38:21.977Z\",\"tags\":[\"project_1337\"]}}],\"iterator\":\"iterator\",\"prevIterator\":\"-iterator\",\"done\":true}";
var ListResponseOperationalWebhookEndpointOut = "{\"data\":[{\"id\":\"ep_1srOrx2ZWZBpBUvZwXKQmoEYga2\",\"description\":\"string\",\"rateLimit\":0,\"uid\":\"unique-identifier\",\"url\":\"https://example.com/webhook/\",\"disabled\":false,\"filterTypes\":[\"message.attempt.failing\"],\"createdAt\":\"2019-08-24T14:15:22Z\",\"updatedAt\":\"2019-08-24T14:15:22Z\",\"metadata\":{\"property1\":\"string\",\"property2\":\"string\"}}],\"iterator\":\"iterator\",\"prevIterator\":\"-iterator\",\"done\":true}";
var ListResponseBackgroundTaskOut = "{\"data\":[{\"data\":{},\"id\":\"qtask_1srOrx2ZWZBpBUvZwXKQmoEYga2\",\"status\":\"running\",\"task\":\"endpoint.replay\",\"updatedAt\":\"2025-03-03T03:03:03.000000Z\"}],\"iterator\":\"iterator\",\"prevIterator\":\"-iterator\",\"done\":true}";
var EventTypeImportOpenApiOut = "{\"data\":{\"modified\":[\"user.signup\"],\"to_modify\":[{\"name\":\"user.signup\",\"description\":\"string\",\"schemas\":{},\"deprecated\":true,\"featureFlag\":\"cool-new-feature\",\"groupName\":\"user\"}]}}";
var ReplayOut = "{\"id\":\"qtask_1srOrx2ZWZBpBUvZwXKQmoEYga2\",\"status\":\"running\",\"task\":\"endpoint.replay\",\"updatedAt\":\"2025-03-03T03:03:03.000000Z\"}";
var EndpointOut = "{\"description\":\"string\",\"rateLimit\":0,\"uid\":\"unique-identifier\",\"url\":\"http://example.com\",\"version\":1,\"disabled\":true,\"filterTypes\":[\"user.signup\"],\"channels\":[\"project_1337\"],\"secret\":\"whsec_C2FVsBQIhrscChlQIMV+b5sSYspob7oD\",\"metadata\":{\"property1\":\"string\",\"property2\":\"string\"}}";
var ValidationErrorOut = "{\"detail\":[{\"loc\":[\"string\"],\"msg\":\"string\",\"type\":\"string\"}]}";
var IngestSourceOutCron = "{\"type\":\"cron\",\"config\":{\"schedule\":\"hello\",\"payload\":\"world\"},\"id\":\"src_2yZwUhtgs5Ai8T9yRQJXA\",\"uid\":\"unique-identifier\",\"name\":\"string\",\"ingestUrl\":\"http://example.com\",\"createdAt\":\"2019-08-24T14:15:22Z\",\"updatedAt\":\"2019-08-24T14:15:22Z\"}";
var IngestSourceOutGeneric = "{\"type\":\"generic-webhook\",\"config\":{},\"id\":\"src_2yZwUhtgs5Ai8T9yRQJXA\",\"uid\":\"unique-identifier\",\"name\":\"string\",\"ingestUrl\":\"http://example.com\",\"createdAt\":\"2019-08-24T14:15:22Z\",\"updatedAt\":\"2019-08-24T14:15:22Z\"}";
var mockServer = mockttp.getLocal();
node_test_1.test("mockttp tests", function (t) { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                t.beforeEach(function () { return __awaiter(void 0, void 0, void 0, function () { return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, mockServer.start(0)];
                        case 1: return [2 /*return*/, _a.sent()];
                    }
                }); }); });
                t.afterEach(function () { return __awaiter(void 0, void 0, void 0, function () { return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, mockServer.stop()];
                        case 1: return [2 /*return*/, _a.sent()];
                    }
                }); }); });
                return [4 /*yield*/, t.test("test Date in query param", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forGet(/\/api\/v1\/app\/app1\/attempt\/endpoint\/ep1.*/)
                                        .thenReply(200, ListResponseMessageAttemptOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.messageAttempt.listByEndpoint("app1", "ep1", {
                                            before: new Date(1739741901977)
                                        })];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    strict_1.strict(requests[0].url.endsWith("before=2025-02-16T21%3A38%3A21.977Z"));
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 1:
                _a.sent();
                return [4 /*yield*/, t.test("test Date request body", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests, _a, _b;
                        return __generator(this, function (_c) {
                            switch (_c.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forPost("/api/v1/app/app1/endpoint/ep1/replay-missing")
                                        .thenReply(200, ReplayOut)];
                                case 1:
                                    endpointMock = _c.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.endpoint.replayMissing("app1", "ep1", { since: new Date(1739741901977) })];
                                case 2:
                                    _c.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _c.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    _b = (_a = strict_1.strict).equal;
                                    return [4 /*yield*/, requests[0].body.getText()];
                                case 4:
                                    _b.apply(_a, [_c.sent(),
                                        "{\"since\":\"2025-02-16T21:38:21.977Z\"}"]);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 2:
                _a.sent();
                return [4 /*yield*/, t.test("test Date response body", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, res, requests;
                        var _a;
                        return __generator(this, function (_b) {
                            switch (_b.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forGet("/api/v1/app/app1/attempt/endpoint/ep1")
                                        .thenReply(200, ListResponseMessageAttemptOut)];
                                case 1:
                                    endpointMock = _b.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.messageAttempt.listByEndpoint("app1", "ep1")];
                                case 2:
                                    res = _b.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _b.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    strict_1.strict.equal((_a = res.data[0].timestamp) === null || _a === void 0 ? void 0 : _a.getTime(), 1739741901977);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 3:
                _a.sent();
                return [4 /*yield*/, t.test("test listResponseOut deserializes correctly", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, res, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forGet("/api/v1/app/app1/attempt/endpoint/ep1")
                                        .thenReply(200, "{\"data\":[],\"iterator\":null,\"prevIterator\":null,\"done\":true}")];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.messageAttempt.listByEndpoint("app1", "ep1")];
                                case 2:
                                    res = _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    strict_1.strict.equal(res.iterator, null);
                                    strict_1.strict.equal(res.prevIterator, null);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 4:
                _a.sent();
                return [4 /*yield*/, t.test("test enum as query param", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forGet(/\/api\/v1\/app.*/)
                                        .thenReply(200, ListResponseApplicationOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.application.list({ order: ordering_1.Ordering.Ascending })];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    strict_1.strict(requests[0].url.endsWith("order=ascending"));
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 5:
                _a.sent();
                return [4 /*yield*/, t.test("test list as query param", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forGet(/\/api\/v1\/app\/app1\/msg.*/)
                                        .thenReply(200, ListResponseMessageOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.message.list("app1", { eventTypes: ["val8", "val1", "val5"] })];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    strict_1.strict(requests[0].url.endsWith("/api/v1/app/app1/msg?event_types=val8%2Cval1%2Cval5"));
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 6:
                _a.sent();
                return [4 /*yield*/, t.test("test header param sent", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forPost(/\/api\/v1\/app.*/)
                                        .thenReply(200, ApplicationOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.application.create({ name: "test" }, { idempotencyKey: "random-key" })];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    strict_1.strict.equal(requests[0].headers["idempotency-key"], "random-key");
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 7:
                _a.sent();
                return [4 /*yield*/, t.test("test retry for status => 500", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var numRetries, endpointMock, svx, requests, req_id, i;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    numRetries = 5;
                                    return [4 /*yield*/, mockServer
                                            .forGet("/api/v1/app")
                                            .thenReply(500, "{\"code\":\"500\",\"detail\":\"asd\"}")];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", {
                                        serverUrl: mockServer.url,
                                        numRetries: numRetries
                                    });
                                    return [4 /*yield*/, strict_1.strict.rejects(svx.application.list(), util_1.ApiException)];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, numRetries + 1);
                                    req_id = requests[0].headers["svix-req-id"];
                                    strict_1.strict(req_id);
                                    strict_1.strict.equal(typeof req_id, "string");
                                    for (i = 0; i < requests.length; i++) {
                                        strict_1.strict.equal(requests[i].headers["svix-req-id"], req_id);
                                        if (i === 0) {
                                            // first request does not set svix-retry-count
                                            strict_1.strict.equal(requests[i].headers["svix-retry-count"], undefined);
                                        }
                                        else {
                                            strict_1.strict.equal(requests[i].headers["svix-retry-count"], i.toString());
                                        }
                                    }
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 8:
                _a.sent();
                return [4 /*yield*/, t.test("test retry schedule for status => 500", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var retryScheduleInMs, endpointMock, before, svx, requests, after;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    retryScheduleInMs = [60, 120, 240];
                                    return [4 /*yield*/, mockServer
                                            .forGet("/api/v1/app")
                                            .thenReply(500, "{\"code\":\"500\",\"detail\":\"asd\"}")];
                                case 1:
                                    endpointMock = _a.sent();
                                    before = Date.now();
                                    svx = new _1.Svix("token", {
                                        serverUrl: mockServer.url,
                                        retryScheduleInMs: retryScheduleInMs
                                    });
                                    return [4 /*yield*/, strict_1.strict.rejects(svx.application.list(), util_1.ApiException)];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, retryScheduleInMs.length + 1);
                                    after = Date.now();
                                    strict_1.strict(after - before >= retryScheduleInMs.reduce(function (prev, curr) { return prev + curr; }, 0));
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 9:
                _a.sent();
                return [4 /*yield*/, t.test("no body in response does not return anything", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forDelete("/api/v1/app/app1")
                                        .thenReply(204, "")];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.application["delete"]("app1")];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 10:
                _a.sent();
                return [4 /*yield*/, t.test("422 returns validation error", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, e_1, error, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forGet("/api/v1/app")
                                        .thenReply(422, ValidationErrorOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, strict_1.strict.rejects(svx.application.list(), util_1.ApiException())];
                                case 2:
                                    _a.sent();
                                    _a.label = 3;
                                case 3:
                                    _a.trys.push([3, 5, , 6]);
                                    return [4 /*yield*/, svx.application.list()];
                                case 4:
                                    _a.sent();
                                    return [3 /*break*/, 6];
                                case 5:
                                    e_1 = _a.sent();
                                    error = e_1;
                                    strict_1.strict.equal(error.code, 422);
                                    strict_1.strict.deepEqual(error.body, {
                                        detail: [{ loc: ["string"], msg: "string", type: "string" }]
                                    });
                                    return [3 /*break*/, 6];
                                case 6: return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 7:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 2);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 11:
                _a.sent();
                return [4 /*yield*/, t.test("400 returns ApiException", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, e_2, error, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forGet("/api/v1/app")
                                        .thenReply(400, "{\"code\":\"400\",\"detail\":\"text\"}")];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, strict_1.strict.rejects(svx.application.list(), util_1.ApiException())];
                                case 2:
                                    _a.sent();
                                    _a.label = 3;
                                case 3:
                                    _a.trys.push([3, 5, , 6]);
                                    return [4 /*yield*/, svx.application.list()];
                                case 4:
                                    _a.sent();
                                    return [3 /*break*/, 6];
                                case 5:
                                    e_2 = _a.sent();
                                    error = e_2;
                                    strict_1.strict.equal(error.code, 400);
                                    strict_1.strict.deepEqual(error.body, { code: "400", detail: "text" });
                                    return [3 /*break*/, 6];
                                case 6: return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 7:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 2);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 12:
                _a.sent();
                return [4 /*yield*/, t.test("sub-resource works", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forGet("/api/v1/operational-webhook/endpoint")
                                        .thenReply(200, ListResponseOperationalWebhookEndpointOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.operationalWebhookEndpoint.list()];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 13:
                _a.sent();
                return [4 /*yield*/, t.test("integer enum serialization", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, res, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forGet("/api/v1/app/app1/attempt/endpoint/endp1")
                                        .thenReply(200, ListResponseMessageAttemptOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.messageAttempt.listByEndpoint("app1", "endp1")];
                                case 2:
                                    res = _a.sent();
                                    strict_1.strict.equal(res.data[0].triggerType, _1.MessageAttemptTriggerType.Scheduled);
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 14:
                _a.sent();
                return [4 /*yield*/, t.test("string enum de/serialization", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, res, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forGet("/api/v1/background-task")
                                        .thenReply(200, ListResponseBackgroundTaskOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.backgroundTask.list({
                                            task: _1.BackgroundTaskType.EndpointReplay
                                        })];
                                case 2:
                                    res = _a.sent();
                                    strict_1.strict.equal(res.data[0].task, _1.BackgroundTaskType.EndpointReplay);
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    strict_1.strict(requests[0].url.endsWith("api/v1/background-task?task=endpoint.replay"));
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 15:
                _a.sent();
                return [4 /*yield*/, t.test("non-camelCase field name", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, res, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forPost("/api/v1/event-type/import/openapi")
                                        .thenReply(200, EventTypeImportOpenApiOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.eventType.importOpenapi({ dryRun: true })];
                                case 2:
                                    res = _a.sent();
                                    strict_1.strict.deepEqual(res.data.toModify, [
                                        {
                                            name: "user.signup",
                                            description: "string",
                                            schemas: {},
                                            deprecated: true,
                                            featureFlag: "cool-new-feature",
                                            featureFlags: undefined,
                                            groupName: "user"
                                        },
                                    ]);
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 16:
                _a.sent();
                return [4 /*yield*/, t.test("patch request body", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests, _a, _b, _c, _d, _e, _f;
                        return __generator(this, function (_g) {
                            switch (_g.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forPatch("/api/v1/app/app1/endpoint/endp1")
                                        .thenReply(200, EndpointOut)];
                                case 1:
                                    endpointMock = _g.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.endpoint.patch("app1", "endp1", { filterTypes: ["ty1"] })];
                                case 2:
                                    _g.sent();
                                    return [4 /*yield*/, svx.endpoint.patch("app1", "endp1", { filterTypes: null })];
                                case 3:
                                    _g.sent();
                                    return [4 /*yield*/, svx.endpoint.patch("app1", "endp1", { description: "text" })];
                                case 4:
                                    _g.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 5:
                                    requests = _g.sent();
                                    strict_1.strict.equal(requests.length, 3);
                                    // nullable field is sent
                                    _b = (_a = strict_1.strict).equal;
                                    return [4 /*yield*/, requests[0].body.getText()];
                                case 6:
                                    // nullable field is sent
                                    _b.apply(_a, [_g.sent(), "{\"filterTypes\":[\"ty1\"]}"]);
                                    // nullable field is null
                                    _d = (_c = strict_1.strict).equal;
                                    return [4 /*yield*/, requests[1].body.getText()];
                                case 7:
                                    // nullable field is null
                                    _d.apply(_c, [_g.sent(), "{\"filterTypes\":null}"]);
                                    // undefined field is omitted
                                    _f = (_e = strict_1.strict).equal;
                                    return [4 /*yield*/, requests[2].body.getText()];
                                case 8:
                                    // undefined field is omitted
                                    _f.apply(_e, [_g.sent(), "{\"description\":\"text\"}"]);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 17:
                _a.sent();
                return [4 /*yield*/, t.test("arbitrary json object body", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests, _a, _b;
                        return __generator(this, function (_c) {
                            switch (_c.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forPost("/api/v1/app/app1/msg")
                                        .thenReply(200, EndpointOut)];
                                case 1:
                                    endpointMock = _c.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.message.create("app1", {
                                            eventType: "asd",
                                            payload: { key1: "val", list: ["val1"], obj: { key: "val2" } }
                                        })];
                                case 2:
                                    _c.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _c.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    _b = (_a = strict_1.strict).equal;
                                    return [4 /*yield*/, requests[0].body.getText()];
                                case 4:
                                    _b.apply(_a, [_c.sent(),
                                        "{\"eventType\":\"asd\",\"payload\":{\"key1\":\"val\",\"list\":[\"val1\"],\"obj\":{\"key\":\"val2\"}}}"]);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 18:
                _a.sent();
                return [4 /*yield*/, t.test("token/user-agent is sent", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forDelete("/api/v1/app/app1")
                                        .thenReply(200, EndpointOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token.eu", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.application["delete"]("app1")];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    strict_1.strict.equal(requests[0].headers["authorization"], "Bearer token.eu");
                                    strict_1.strict.equal(requests[0].headers["user-agent"], "svix-libs/" + request_1.LIB_VERSION + "/javascript");
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 19:
                _a.sent();
                return [4 /*yield*/, t.test("MessageAttemptOut without msg", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var RES, endpointMock, svx, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    RES = "{\n      \"data\": [\n        {\n          \"url\": \"https://example.com/webhook/\",\n          \"response\": \"{}\",\n          \"responseStatusCode\": 200,\n          \"responseDurationMs\": 0,\n          \"status\": 0,\n          \"triggerType\": 0,\n          \"msgId\": \"msg_1srOrx2ZWZBpBUvZwXKQmoEYga2\",\n          \"endpointId\": \"ep_1srOrx2ZWZBpBUvZwXKQmoEYga2\",\n          \"id\": \"atmpt_1srOrx2ZWZBpBUvZwXKQmoEYga2\",\n          \"timestamp\": \"2019-08-24T14:15:22Z\"\n        }\n      ],\n      \"iterator\": \"iterator\",\n      \"prevIterator\": \"-iterator\",\n      \"done\": true\n    }";
                                    return [4 /*yield*/, mockServer
                                            .forGet("/api/v1/app/app/attempt/endpoint/edp")
                                            .thenReply(200, RES)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token.eu", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.messageAttempt.listByEndpoint("app", "edp")];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 20:
                _a.sent();
                return [4 /*yield*/, t.test("octothorpe in url query", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forGet("/api/v1/app/app1/msg")
                                        .thenReply(200, ListResponseMessageOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token.eu", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.message.list("app1", { tag: "test#test" })];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    strict_1.strict(requests[0].url.endsWith("api/v1/app/app1/msg?tag=test%23test"));
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 21:
                _a.sent();
                return [4 /*yield*/, t.test("content-type application/json is sent on request with body", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forPost(/\/api\/v1\/app.*/)
                                        .thenReply(200, ApplicationOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.application.create({ name: "test" })];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    strict_1.strict.equal(requests[0].headers["content-type"], "application/json");
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 22:
                _a.sent();
                return [4 /*yield*/, t.test("content type not sent on request without body", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forGet("/api/v1/app")
                                        .thenReply(200, ListResponseApplicationOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.application.list()];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    strict_1.strict.equal(requests[0].headers["content-type"], undefined);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 23:
                _a.sent();
                return [4 /*yield*/, t.test("struct enum with extra fields", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, res, requests, _a, _b;
                        return __generator(this, function (_c) {
                            switch (_c.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forPost("/ingest/api/v1/source")
                                        .thenReply(200, IngestSourceOutCron)];
                                case 1:
                                    endpointMock = _c.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.ingest.source.create({
                                            name: "crontab -r",
                                            type: "cron",
                                            config: { schedule: "hello", payload: "world" }
                                        })];
                                case 2:
                                    res = _c.sent();
                                    strict_1.strict.equal(res.type, "cron");
                                    // this will smart cast res.config
                                    if (res.type === "cron") {
                                        strict_1.strict.equal(res.config.schedule, "hello");
                                        strict_1.strict.equal(res.config.payload, "world");
                                    }
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _c.sent();
                                    _b = (_a = strict_1.strict).equal;
                                    return [4 /*yield*/, requests[0].body.getText()];
                                case 4:
                                    _b.apply(_a, [_c.sent(), '{"type":"cron","config":{"payload":"world","schedule":"hello"},"name":"crontab -r"}']);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 24:
                _a.sent();
                return [4 /*yield*/, t.test("struct enum with no extra fields", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, res, requests, _a, _b;
                        return __generator(this, function (_c) {
                            switch (_c.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forPost("/ingest/api/v1/source")
                                        .thenReply(200, IngestSourceOutGeneric)];
                                case 1:
                                    endpointMock = _c.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.ingest.source.create({
                                            name: "generic over <T>",
                                            type: "generic-webhook"
                                        })];
                                case 2:
                                    res = _c.sent();
                                    strict_1.strict.equal(res.type, "generic-webhook");
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _c.sent();
                                    // empty config object should be sent
                                    _b = (_a = strict_1.strict).equal;
                                    return [4 /*yield*/, requests[0].body.getText()];
                                case 4:
                                    // empty config object should be sent
                                    _b.apply(_a, [_c.sent(), '{"type":"generic-webhook","config":{},"name":"generic over <T>"}']);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 25:
                _a.sent();
                return [4 /*yield*/, t.test("test idempotency key is sent for create request", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests, idempotencyKey;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forPost("/api/v1/app")
                                        .thenReply(200, ApplicationOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.application.create({ name: "test app" })];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    idempotencyKey = requests[0].headers["idempotency-key"];
                                    strict_1.strict(idempotencyKey.startsWith("auto_"));
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 26:
                _a.sent();
                return [4 /*yield*/, t.test("test client provided idempotency key is not overridden", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, clientProvidedKey, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forPost("/api/v1/app")
                                        .thenReply(200, ApplicationOut)];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    clientProvidedKey = "test-key-123";
                                    return [4 /*yield*/, svx.application.create({ name: "test app" }, { idempotencyKey: clientProvidedKey })];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    strict_1.strict.equal(requests[0].headers["idempotency-key"], clientProvidedKey);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 27:
                _a.sent();
                return [4 /*yield*/, t.test("test unknown keys are ignored", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var endpointMock, svx, requests;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, mockServer
                                        .forGet("/api/v1/app")
                                        .thenReply(200, "{\"data\":[],\"done\":true,\"iterator\":null,\"prevIterator\":null,\"extra-key\":\"ignored\"}")];
                                case 1:
                                    endpointMock = _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url });
                                    return [4 /*yield*/, svx.application.list()];
                                case 2:
                                    _a.sent();
                                    return [4 /*yield*/, endpointMock.getSeenRequests()];
                                case 3:
                                    requests = _a.sent();
                                    strict_1.strict.equal(requests.length, 1);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 28:
                _a.sent();
                return [4 /*yield*/, t.test("should use custom fetch implementation", function () { return __awaiter(void 0, void 0, void 0, function () {
                        var customFetchCalled, mockFetch, svx;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    customFetchCalled = false;
                                    mockFetch = function (_input, _init) { return __awaiter(void 0, void 0, void 0, function () {
                                        return __generator(this, function (_a) {
                                            customFetchCalled = true;
                                            return [2 /*return*/, new Response(ListResponseApplicationOut, {
                                                    status: 200,
                                                    headers: { "Content-Type": "application/json" }
                                                })];
                                        });
                                    }); };
                                    return [4 /*yield*/, mockServer
                                            .forGet(/\/api\/v1\/app.*/)
                                            .thenReply(200, ListResponseApplicationOut)];
                                case 1:
                                    _a.sent();
                                    svx = new _1.Svix("token", { serverUrl: mockServer.url, fetch: mockFetch });
                                    return [4 /*yield*/, svx.application.list({ order: ordering_1.Ordering.Ascending })];
                                case 2:
                                    _a.sent();
                                    strict_1.strict(customFetchCalled);
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            case 29:
                _a.sent();
                return [2 /*return*/];
        }
    });
}); });
