// svix/src/dist/request.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.SvixRequest = exports.HttpMethod = exports.LIB_VERSION = void 0;
var util_1 = require("./util");
exports.LIB_VERSION = "1.92.2";
var USER_AGENT = "svix-libs/" + exports.LIB_VERSION + "/javascript";
var HttpMethod;
(function (HttpMethod) {
    HttpMethod["GET"] = "GET";
    HttpMethod["HEAD"] = "HEAD";
    HttpMethod["POST"] = "POST";
    HttpMethod["PUT"] = "PUT";
    HttpMethod["DELETE"] = "DELETE";
    HttpMethod["CONNECT"] = "CONNECT";
    HttpMethod["OPTIONS"] = "OPTIONS";
    HttpMethod["TRACE"] = "TRACE";
    HttpMethod["PATCH"] = "PATCH";
})(HttpMethod = exports.HttpMethod || (exports.HttpMethod = {}));
var SvixRequest = /** @class */ (function () {
    function SvixRequest(method, path) {
        this.method = method;
        this.path = path;
        this.queryParams = {};
        this.headerParams = {};
    }
    SvixRequest.prototype.setPathParam = function (name, value) {
        var newPath = this.path.replace("{" + name + "}", encodeURIComponent(value));
        if (this.path === newPath) {
            throw new Error("path parameter " + name + " not found");
        }
        this.path = newPath;
    };
    SvixRequest.prototype.setQueryParams = function (params) {
        for (var _i = 0, _a = Object.entries(params); _i < _a.length; _i++) {
            var _b = _a[_i], name = _b[0], value = _b[1];
            this.setQueryParam(name, value);
        }
    };
    SvixRequest.prototype.setQueryParam = function (name, value) {
        if (value === undefined || value === null) {
            return;
        }
        if (typeof value === "string") {
            this.queryParams[name] = value;
        }
        else if (typeof value === "boolean" || typeof value === "number") {
            this.queryParams[name] = value.toString();
        }
        else if (value instanceof Date) {
            this.queryParams[name] = value.toISOString();
        }
        else if (Array.isArray(value)) {
            if (value.length > 0) {
                this.queryParams[name] = value.join(",");
            }
        }
        else {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            var _assert_unreachable = value;
            throw new Error("query parameter " + name + " has unsupported type");
        }
    };
    SvixRequest.prototype.setHeaderParam = function (name, value) {
        if (value === undefined) {
            return;
        }
        this.headerParams[name] = value;
    };
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    SvixRequest.prototype.setBody = function (value) {
        this.body = JSON.stringify(value);
    };
    /**
     * Send this request, returning the request body as a caller-specified type.
     *
     * If the server returns a 422 error, an `ApiException<HTTPValidationError>` is thrown.
     * If the server returns another 4xx error, an `ApiException<HttpErrorOut>` is thrown.
     *
     * If the server returns a 5xx error, the request is retried up to two times with exponential backoff.
     * If retries are exhausted, an `ApiException<HttpErrorOut>` is thrown.
     */
    SvixRequest.prototype.send = function (ctx, 
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    parseResponseBody) {
        return __awaiter(this, void 0, Promise, function () {
            var response, responseBody;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.sendInner(ctx)];
                    case 1:
                        response = _a.sent();
                        if (response.status === 204) {
                            return [2 /*return*/, null];
                        }
                        return [4 /*yield*/, response.text()];
                    case 2:
                        responseBody = _a.sent();
                        return [2 /*return*/, parseResponseBody(JSON.parse(responseBody))];
                }
            });
        });
    };
    /** Same as `send`, but the response body is discarded, not parsed. */
    SvixRequest.prototype.sendNoResponseBody = function (ctx) {
        return __awaiter(this, void 0, Promise, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.sendInner(ctx)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    SvixRequest.prototype.sendInner = function (ctx) {
        var _a, _b;
        return __awaiter(this, void 0, Promise, function () {
            var url, _i, _c, _d, name, value, randomId, isCredentialsSupported, response;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        url = new URL(ctx.baseUrl + this.path);
                        for (_i = 0, _c = Object.entries(this.queryParams); _i < _c.length; _i++) {
                            _d = _c[_i], name = _d[0], value = _d[1];
                            url.searchParams.set(name, value);
                        }
                        if (this.headerParams["idempotency-key"] === undefined &&
                            this.method.toUpperCase() === "POST") {
                            this.headerParams["idempotency-key"] = "auto_" + crypto.randomUUID();
                        }
                        randomId = Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);
                        if (this.body != null) {
                            this.headerParams["content-type"] = "application/json";
                        }
                        isCredentialsSupported = "credentials" in Request.prototype;
                        return [4 /*yield*/, sendWithRetry(url, {
                                method: this.method.toString(),
                                body: this.body,
                                headers: __assign({ accept: "application/json, */*;q=0.8", authorization: "Bearer " + ctx.token, "user-agent": USER_AGENT, "svix-req-id": randomId.toString() }, this.headerParams),
                                credentials: isCredentialsSupported ? "same-origin" : undefined,
                                signal: ctx.timeout !== undefined ? AbortSignal.timeout(ctx.timeout) : undefined
                            }, ctx.retryScheduleInMs, (_a = ctx.retryScheduleInMs) === null || _a === void 0 ? void 0 : _a[0], ((_b = ctx.retryScheduleInMs) === null || _b === void 0 ? void 0 : _b.length) || ctx.numRetries, ctx.fetch)];
                    case 1:
                        response = _e.sent();
                        return [2 /*return*/, filterResponseForErrors(response)];
                }
            });
        });
    };
    return SvixRequest;
}());
exports.SvixRequest = SvixRequest;
function filterResponseForErrors(response) {
    return __awaiter(this, void 0, Promise, function () {
        var responseBody;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (response.status < 300) {
                        return [2 /*return*/, response];
                    }
                    return [4 /*yield*/, response.text()];
                case 1:
                    responseBody = _a.sent();
                    if (response.status === 422) {
                        throw new util_1.ApiException(response.status, JSON.parse(responseBody), response.headers);
                    }
                    if (response.status >= 400 && response.status <= 499) {
                        throw new util_1.ApiException(response.status, JSON.parse(responseBody), response.headers);
                    }
                    throw new util_1.ApiException(response.status, responseBody, response.headers);
            }
        });
    });
}
function sendWithRetry(url, init, retryScheduleInMs, nextInterval, triesLeft, fetchImpl, retryCount) {
    if (nextInterval === void 0) { nextInterval = 50; }
    if (triesLeft === void 0) { triesLeft = 2; }
    if (fetchImpl === void 0) { fetchImpl = fetch; }
    if (retryCount === void 0) { retryCount = 1; }
    return __awaiter(this, void 0, Promise, function () {
        var sleep, response, e_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    sleep = function (interval) {
                        return new Promise(function (resolve) { return setTimeout(resolve, interval); });
                    };
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, fetchImpl(url, init)];
                case 2:
                    response = _a.sent();
                    if (triesLeft <= 0 || response.status < 500) {
                        return [2 /*return*/, response];
                    }
                    return [3 /*break*/, 4];
                case 3:
                    e_1 = _a.sent();
                    if (triesLeft <= 0) {
                        throw e_1;
                    }
                    return [3 /*break*/, 4];
                case 4: return [4 /*yield*/, sleep(nextInterval)];
                case 5:
                    _a.sent();
                    init.headers["svix-retry-count"] = retryCount.toString();
                    nextInterval = (retryScheduleInMs === null || retryScheduleInMs === void 0 ? void 0 : retryScheduleInMs[retryCount]) || nextInterval * 2;
                    return [4 /*yield*/, sendWithRetry(url, init, retryScheduleInMs, nextInterval, --triesLeft, fetchImpl, ++retryCount)];
                case 6: return [2 /*return*/, _a.sent()];
            }
        });
    });
}
