// svix/src/dist/webhook.test.js
"use strict";
exports.__esModule = true;
var utf8 = require("@stablelib/utf8");
var base64 = require("@stablelib/base64");
var sha256 = require("fast-sha256");
var node_test_1 = require("node:test");
var strict_1 = require("node:assert/strict");
var index_1 = require("./index");
var defaultMsgID = "msg_p5jXN8AQM9LWM0D4loKWxJek";
var defaultPayload = "{\"test\": 2432232314}";
var defaultSecret = "MfKQ9r8GKYqrTwjUPD8ILPZIo2LaLaSw";
var tolerance_in_ms = 5 * 60 * 1000;
var TestPayload = /** @class */ (function () {
    function TestPayload(timestamp) {
        if (timestamp === void 0) { timestamp = Date.now(); }
        this.id = defaultMsgID;
        this.timestamp = Math.floor(timestamp / 1000);
        this.payload = defaultPayload;
        this.secret = defaultSecret;
        var toSign = utf8.encode(this.id + "." + this.timestamp + "." + this.payload);
        this.signature = base64.encode(sha256.hmac(base64.decode(this.secret), toSign));
        this.header = {
            "svix-id": this.id,
            "svix-signature": "v1," + this.signature,
            "svix-timestamp": this.timestamp.toString()
        };
    }
    return TestPayload;
}());
node_test_1.test("empty key raises error", function () {
    strict_1.strict.throws(function () {
        new index_1.Webhook("");
    }, Error);
    strict_1.strict.throws(function () {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        new index_1.Webhook(undefined);
    }, Error);
    strict_1.strict.throws(function () {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        new index_1.Webhook(null);
    }, Error);
});
node_test_1.test("missing id raises error", function () {
    var wh = new index_1.Webhook("MfKQ9r8GKYqrTwjUPD8ILPZIo2LaLaSw");
    var testPayload = new TestPayload();
    delete testPayload.header["svix-id"];
    strict_1.strict.throws(function () {
        wh.verify(testPayload.payload, testPayload.header);
    }, index_1.WebhookVerificationError);
});
node_test_1.test("missing timestamp raises error", function () {
    var wh = new index_1.Webhook("MfKQ9r8GKYqrTwjUPD8ILPZIo2LaLaSw");
    var testPayload = new TestPayload();
    delete testPayload.header["svix-timestamp"];
    strict_1.strict.throws(function () {
        wh.verify(testPayload.payload, testPayload.header);
    }, index_1.WebhookVerificationError);
});
node_test_1.test("invalid timestamp throws error", function () {
    var wh = new index_1.Webhook("MfKQ9r8GKYqrTwjUPD8ILPZIo2LaLaSw");
    var testPayload = new TestPayload();
    testPayload.header["svix-timestamp"] = "hello";
    strict_1.strict.throws(function () {
        wh.verify(testPayload.payload, testPayload.header);
    }, index_1.WebhookVerificationError);
});
node_test_1.test("missing signature raises error", function () {
    var wh = new index_1.Webhook("MfKQ9r8GKYqrTwjUPD8ILPZIo2LaLaSw");
    var testPayload = new TestPayload();
    delete testPayload.header["svix-signature"];
    strict_1.strict.throws(function () {
        wh.verify(testPayload.payload, testPayload.header);
    }, index_1.WebhookVerificationError);
});
node_test_1.test("invalid signature throws error", function () {
    var wh = new index_1.Webhook("MfKQ9r8GKYqrTwjUPD8ILPZIo2LaLaSw");
    var testPayload = new TestPayload();
    testPayload.header["svix-signature"] = "v1,dawfeoifkpqwoekfpqoekf";
    strict_1.strict.throws(function () {
        wh.verify(testPayload.payload, testPayload.header);
    }, index_1.WebhookVerificationError);
});
node_test_1.test("partial signature throws error", function () {
    var wh = new index_1.Webhook("MfKQ9r8GKYqrTwjUPD8ILPZIo2LaLaSw");
    var testPayload = new TestPayload();
    testPayload.header["svix-signature"] = testPayload.header["svix-signature"].slice(0, 8);
    strict_1.strict.throws(function () {
        wh.verify(testPayload.payload, testPayload.header);
    }, index_1.WebhookVerificationError);
    testPayload.header["svix-signature"] = "v1,";
    strict_1.strict.throws(function () {
        wh.verify(testPayload.payload, testPayload.header);
    }, index_1.WebhookVerificationError);
});
node_test_1.test("valid signature is valid and returns valid json", function () {
    var wh = new index_1.Webhook("MfKQ9r8GKYqrTwjUPD8ILPZIo2LaLaSw");
    var testPayload = new TestPayload();
    wh.verify(testPayload.payload, testPayload.header);
});
node_test_1.test("valid unbranded signature is valid and returns valid json", function () {
    var wh = new index_1.Webhook("MfKQ9r8GKYqrTwjUPD8ILPZIo2LaLaSw");
    var testPayload = new TestPayload();
    var unbrandedHeaders = {
        "webhook-id": testPayload.header["svix-id"],
        "webhook-signature": testPayload.header["svix-signature"],
        "webhook-timestamp": testPayload.header["svix-timestamp"]
    };
    testPayload.header = unbrandedHeaders;
    wh.verify(testPayload.payload, testPayload.header);
});
node_test_1.test("old timestamp fails", function () {
    var wh = new index_1.Webhook("MfKQ9r8GKYqrTwjUPD8ILPZIo2LaLaSw");
    var testPayload = new TestPayload(Date.now() - tolerance_in_ms - 1000);
    strict_1.strict.throws(function () {
        wh.verify(testPayload.payload, testPayload.header);
    }, index_1.WebhookVerificationError);
});
node_test_1.test("new timestamp fails", function () {
    var wh = new index_1.Webhook("MfKQ9r8GKYqrTwjUPD8ILPZIo2LaLaSw");
    var testPayload = new TestPayload(Date.now() + tolerance_in_ms + 1000);
    strict_1.strict.throws(function () {
        wh.verify(testPayload.payload, testPayload.header);
    }, index_1.WebhookVerificationError);
});
node_test_1.test("multi sig payload is valid", function () {
    var wh = new index_1.Webhook("MfKQ9r8GKYqrTwjUPD8ILPZIo2LaLaSw");
    var testPayload = new TestPayload();
    var sigs = [
        "v1,Ceo5qEr07ixe2NLpvHk3FH9bwy/WavXrAFQ/9tdO6mc=",
        "v2,Ceo5qEr07ixe2NLpvHk3FH9bwy/WavXrAFQ/9tdO6mc=",
        testPayload.header["svix-signature"],
        "v1,Ceo5qEr07ixe2NLpvHk3FH9bwy/WavXrAFQ/9tdO6mc=",
    ];
    testPayload.header["svix-signature"] = sigs.join(" ");
    wh.verify(testPayload.payload, testPayload.header);
});
node_test_1.test("verification works with and without signature prefix", function () {
    var testPayload = new TestPayload();
    var wh = new index_1.Webhook(defaultSecret);
    wh.verify(testPayload.payload, testPayload.header);
    wh = new index_1.Webhook("whsec_" + defaultSecret);
    wh.verify(testPayload.payload, testPayload.header);
});
node_test_1.test("sign function works", function () {
    var key = "whsec_MfKQ9r8GKYqrTwjUPD8ILPZIo2LaLaSw";
    var msgId = "msg_p5jXN8AQM9LWM0D4loKWxJek";
    var timestamp = new Date(1614265330 * 1000);
    var payload = '{"test": 2432232314}';
    var expected = "v1,g0hM9SsE+OTPJTGt/tmIKtSyZlE3uFJELVlNIOLJ1OE=";
    var wh = new index_1.Webhook(key);
    var signature = wh.sign(msgId, timestamp, payload);
    strict_1.strict.equal(signature, expected);
});
